{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5f989060-147b-4724-a728-42dc7aa8daa2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Mike IMC Pipeline v0.1 - 8th September 2022\n",
    "#### <font color='blue'>This is a work in progress, and should be treated as such. Expect errors and issues! The start should hopefully be fairly 'one size fits all', but as it progresses it gets more and more complex, and will require an increasing knowledge of Python or more customisation to run: </font>\n",
    "\n",
    "### <font color='green'>Green</font> - Should run with only minor or no changes at all\n",
    "### <font color='orange'>Orange</font> - Requires a fair amount of customisation, and usually adding in external files (e.g. images or extra .csv files)\n",
    "### <font color='red'>Red</font> - Currently not well annotated and/or very complex code\n",
    "\n",
    "Developed by Michael Haley (michael.haley@manchester.ac.uk), **please do not share publicly!!**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "200ae033-b55b-4ef3-9de0-2ab57162038b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <font color='red'>Setting up Python</font>\n",
    "\n",
    "### 1. Setting up Anaconda\n",
    "If you haven't yet setup Anaconda on your system, and are viewing this on the Github, there first step is to install and setup **Anaconda** (https://www.anaconda.com/products/distribution)\n",
    "\n",
    "\n",
    "### 2. Setting up the environment\n",
    "Once Anaconda is installed, we need to open the Anaconda Powershell Prompt, and create a new environment which we will call IMC (you can call it anything) using Conda. \n",
    ">**conda create --name IMC**</br>\n",
    "\n",
    "We then want to add some channels to Conda where packages will be downloaded from\n",
    ">**conda config --add channels conda-forge bioconda**</br>\n",
    "\n",
    "We now activate our new environment, and then start installing the packages we will need. This could take a while!\n",
    ">**conda install squidpy scanpy pandas anndata seaborn yaml scikit-image yellowbrick dill umap-learn napari-imc colorcet scikit-learn jupyterlab shapely tqdm git**\n",
    "</br>\n",
    "\n",
    "Some packages aren't available on Conda, and therefore need to be installed via Pip\n",
    ">**pip install ai4scr-spatial-omics ai4scr-athena phenograph**</br>\n",
    "\n",
    "### 3. Starting Jupyter\n",
    "Download this notebook and the **'mikeimc_v2.py'** place them somewhere sensible (e.g. 'C:/Programming', or wherever)\n",
    "\n",
    "Open Anaconda Powershell and change directories to where you have put the downloaded files. In the above example...\n",
    ">**cd c:/programming**\n",
    "\n",
    "Activate the IMC environment. You can tell what environment you are currently in by looking at the start of the command line by default it will be *(base)*, after this is should be *(IMC)*\n",
    "\n",
    ">**conda activate IMC**\n",
    "\n",
    "Start Jupyter notebook\n",
    ">**jupyter notebook**\n",
    "\n",
    "Open up the Jupyter notebook, and contine below..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d633d09-99e8-4234-8519-88272beb7bee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## <a id=\"celltable\"><font color='red'>Cleaning up the cell table</font></a>\n",
    "\n",
    "This pipeline assumes you have already segmented your data to get single-cell data, and have a **cell table** for your data (see example in image below). This is simply a spreadsheet (.csv format) in which each cell is one row, and information about that cell in columns - usually marker expression and X and Y location in the image, but could also be other information such as cell size.\n",
    "\n",
    "You will want to clean up your cell table first in Excel. The output of most segmentation pipelines give you a *huge* cell table, with loads of extra information you probably dont need.\n",
    "\n",
    "The minimum columns you need are:\n",
    "\n",
    "- 1 column called **'ROI'** that identifies the image where the cells came from\n",
    "\n",
    "- **1 column *per* marker**, which should be the mean intensity of that marker, with an easily identifiable name\n",
    "\n",
    "- 1 column for **X location** of cell\n",
    "\n",
    "- 1 column for **Y location** of cell\n",
    "\n",
    "## <font color='red'>Creating the accompanying files</font>\n",
    "Create (or modify using the templates) **columns.csv** and **dictionary.csv** files as detailed in the below schematic.\n",
    "\n",
    "Save them in the save folder as the Jupyter notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f92fa316-6b92-4f58-b2f9-948dc5bfd1ac",
   "metadata": {},
   "source": [
    "<img src=\"celltable.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87dfcc48-96d2-43b7-a068-24b827dc5606",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## <font color='red'>Important notes on using this pipeline</font>\n",
    "When we do analyses in Python, it is important to understand *where* and *how* our data is being stored. Simply, there are 3 different places that our analyses may exist, potentially all at the same time:\n",
    "\n",
    "*Data and analyes may be stored in....*\n",
    "\n",
    "#### The Python kernel\n",
    "This is the instance of Python that you are currrently working in. It has all the variables and objects you have currently defined stored in the memory, allowing us to access and work with them. However, if we close the kernel, or it crashes, **we lose all our analyses!** Usually we can just re-run all the analyses again without an issue, but it's time consuming and annoying. If the browser crashes, that is not an issue - just reopen the browser,  go back to the Jupyter web page (look in your browser 'history' to find the address), and you should be able to pick up where you left off. However, in order to keep Jupyter and the kernel running, do not close the Anaconda prompt you used to start Jupyter, or turn off your computer\n",
    "\n",
    "#### The Jupyter notebook file (.ipynb)\n",
    "This stores all your **markdown, code and the outputs** of any analyses you have run - essentially, **if you can see it visually on screen it will be saved in the Jupyter notebook file!**. Jupyter will automaticlaly save periodically. However, if you rerun cells, you lose any previous outputs, unless you have saved them to disk...\n",
    "\n",
    "#### Exported / saved figures\n",
    "The safest way to store your analyses and export them out of Python is to store the resulting figures to disk as images, or export any tabular data as spread sheets. \n",
    "\n",
    "#### Adata object saved on disk\n",
    "You can save AnnData (*adata*) objects to disk, and reload them later. This is the a very convenient way to come back to later analyses. Unfortunately, there is no good way to save spatial-omics *so* objects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5209e003-5737-40b7-9143-6d10b8bc32fb",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "Do not close the Anaconda Powershell you used to start Jupyter, turn off your computer entirely, or manually restart the kernel <b>unless</b> you have saved your analyses somehow, or don't mind rerunning your analyses (which in some cases may take several hours!)\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ffb709a-5a9c-49eb-acb7-d35501849d98",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Import libraries and setup </font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8e9b796-5cee-4c21-8062-821598d59dcf",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Read the package documentation!</b> I'm not going to give an extended explanation of how the various packages work. Instead, go and read the documentation and papers associated with the main packages I have linked here. This will explain how the different functions work, give worked examples, and explain how you can customise the code. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d2e1e80-3639-4d09-bc0a-2ebe1c0b7d6e",
   "metadata": {
    "tags": []
   },
   "source": [
    "This will import the libraries that we need, including a module I've made called **'mikeimc'** which has some useful functions for processing IMC data - *mikeimc_v2.py* should the in the same directory as this notebook. The other libraries should already be installed if you have setup the environment correctly, if not they won't import here and you will get an error.\n",
    "\n",
    "### Packages:\n",
    "\n",
    "##### Scanpy\n",
    "Single-cell analysis developed for single-cell sequencing, but also works great for IMC.\n",
    "https://scanpy.readthedocs.io/en/stable/\n",
    "\n",
    "##### Squidpy\n",
    "Built on scanpy, expands to do several different spatial analyses\n",
    "https://squidpy.readthedocs.io/en/stable/\n",
    "https://www.biorxiv.org/content/10.1101/2021.02.19.431994v1\n",
    "\n",
    "##### AnnData\n",
    "Data object used to store all our single cell analyses (in the *adata* object)\n",
    "https://anndata.readthedocs.io/en/latest/\n",
    "\n",
    "##### Pandas\n",
    "Tools for working with tables of data, which is pretty much all the data we will be working with.\n",
    "\n",
    "##### BBKNN:\n",
    "Algorithm implemented by Scanpy to do batch correction\n",
    "https://academic.oup.com/bioinformatics/article/36/3/964/5545955\n",
    "\n",
    "##### ATHENA and 'spatial omics' object\n",
    "Developed by a former member of the Bodemiller lab, has a lot more functionality to do in-depth spatial analyses. The problem is that it isn't build directly on an *adata* object, instead using an *so* (spatial omics) object. Therefore, we have to transer all the data from the *adata* to the *so* half way through.\n",
    "\n",
    "https://github.com/AI4SCR/ATHENA\n",
    "https://github.com/AI4SCR/spatial-omics\n",
    "https://histocartography.github.io/athena/index.html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9369db5-5866-43bd-94cd-332743664767",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>CaseViewer/3DHistech Issue</b> If you are having an error where Jupyter crashes and you have CaseViewer or some other 3D Histech software installed, it could be because of the following issue:\n",
    "https://github.com/lmcinnes/umap/issues/702\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c0557a9-e68f-48ba-b09f-38c5bc476646",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Standard libraries for data analysis, most of these should be automatically installed\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "\n",
    "# Pandas for working with dataframes and tabular data\n",
    "import pandas as pd\n",
    "\n",
    "# AnnData for working with anndatas (annotated datas)\n",
    "import anndata as ad\n",
    "\n",
    "# Matplotlib and seaborn for plotting\n",
    "import matplotlib\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "from matplotlib import cm\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap, Normalize\n",
    "import seaborn as sb\n",
    "\n",
    "# For access to good colour maps\n",
    "import colorcet as cc\n",
    "\n",
    "# OS and pathlib for opening/deleting files\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "# Scanpy \n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "\n",
    "# Squidpy\n",
    "import squidpy as sq\n",
    "\n",
    "#Packages to do Elbow analysis\n",
    "from sklearn.cluster import KMeans\n",
    "from yellowbrick.cluster import KElbowVisualizer\n",
    "\n",
    "# Import UMAP\n",
    "# This can cause a crash if CaseViewer is installed! See comment above.\n",
    "import umap \n",
    "\n",
    "#My module for importing IMC data into AnnData format\n",
    "import mikeimc_v2\n",
    "import mikeimc_v2 as mimc\n",
    "\n",
    "#To display multiple dataframes in same output\n",
    "from IPython.display import display\n",
    "\n",
    "#ATHENA\n",
    "from spatialOmics import SpatialOmics\n",
    "import athena as sh\n",
    "\n",
    "#Loading bar functionality for long tasks\n",
    "from tqdm import tqdm\n",
    "\n",
    "#For multiprocessing\n",
    "from functools import partial\n",
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b688dc-1c7a-48ac-8bab-f6348770feb2",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Reload mikeimc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83f79262-16f8-41f5-82e1-1bf457c73e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mikeimc_v2\n",
    "import imp\n",
    "imp.reload(mikeimc_v2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea939523-828c-4c37-9b9f-ad0f51950ff3",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Setup or reset default parameters for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a4fadac-e017-431b-9deb-baa1eefb12f9",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "    <b>Tip</b> Re-run this cell if your plots start looking weird and you're not sure why! \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2b53a9c-57cd-45c3-82d4-ee9d75655434",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reset plotting parameters to their defaults\n",
    "matplotlib.pyplot.rcParams.update(matplotlib.rcParamsDefault)\n",
    "\n",
    "# Set up output figure settings\n",
    "plt.rcParams['figure.figsize']=(5,5) #rescale figures, increase sizehere\n",
    "\n",
    "# Set up scanpy settings\n",
    "sc.settings.verbosity = 3\n",
    "sc.set_figure_params(dpi=100, dpi_save=200) #Increase DPI for better resolution figures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2858e49-345a-4aa1-8d9d-04dcb7ed7372",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Saving / loading adata objects </font>\n",
    "These don't need to be run initially, but will be useful to save and load analyses, or export data to a spreadsheet for external analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2b09b5-b038-441e-8a7f-78d0c5907231",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Be careful using save and load!</b> There will be no warning if you overwrite the wrong file, or load in an empty adata and therefore lose the one you're working on!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6790405-f4a3-4080-b66b-2a527fad5b3d",
   "metadata": {},
   "source": [
    "#### Save adata object\n",
    "This will save the adata object in the current directory as a file called 'adata'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd74f57f-2df1-46d0-b102-942dbec99348",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('adata')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "012b2876-33ec-43b7-9cff-e2dfe7aa330a",
   "metadata": {},
   "source": [
    "#### Export cell table\n",
    "This will export the currently adata.obs, which will give you a table format of all your cells, what populations and results of several key analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f751c59-5c94-4169-a8ad-a0c24dca66b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.to_csv('adata_obs.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "683e3b53-3b29-4ff0-ab53-7fabc798a291",
   "metadata": {},
   "source": [
    "#### Load adata\n",
    "This will open up a stored adata object (from the 'raw' directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e84328f-0cf6-420b-8c1d-1c6058fcb2c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad('adata')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "991f689b-66cc-4775-a9ea-b259d748c066",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Importing raw data (*cell table*, *column* and *dictionary* .csv files) into the adata object</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcc083a8-3af4-456b-9482-497ece009862",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Import cell data from .csv files\n",
    "[See above for an explanation of how data should be layed out for importing](#celltable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "725a0c2a-5988-49a1-8402-5b754e358b29",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_fullimport, df_misc = mimc.celltable_to_adata(column_properties='raw/columns.csv', #Contents define how the columns in the cell table should be handled\n",
    "                                                cell_table='raw/cells.csv', #The full cell table as exported by whatever cell segmentation pipeline\n",
    "                                                dictionary='raw/dictionary.csv', #If this is specified, it will add extra observations based upon the ROI, e.g. what group/patient/treatment it came from\n",
    "                                                marker_normalisation='99th', #Marker normalisation, by default will be None\n",
    "                                                misc_table=True) #Whether or not a second dataframe of 'misc' findings will be imported"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cf2a9cf-f646-4378-9e71-5e0ffa21aff8",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Check AnnData import was worked as expected\n",
    "Check that all the variables (markers) are listed here, as well as any obserations under 'obs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a4d423c-244f-44e1-b939-b54ff04e4243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print a list of the variables imported\n",
    "print(adata_fullimport.var_names)\n",
    "\n",
    "# Print the structure of the AnnData\n",
    "print(adata_fullimport)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c317775-e718-4a2f-9bed-326c445d216d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Removing markers and non-primary samples to make the working AnnData"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3aec822-a8d7-4d77-a8ea-64a4467bdb9e",
   "metadata": {},
   "source": [
    "Use this code to remove any markers or samples which you don't want included in your final adata that you'll be using for analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bffe80a9-2d85-4bde-b458-d61308b2a2ab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = mimc.filter_anndata(adata_source=adata_fullimport,\n",
    "                            markers_to_remove=['DNA1', 'DNA3','PanCytokeratin','iNOS'], #List of markers to remove, e.g, DNA markers\n",
    "                            obs_to_filter_on='ROI', #The .obs category which you want to filter on\n",
    "                            obs_to_remove=['Tonsil_Test','BrainCtrl','2A3_Test','Tonsil']) #The types of .obs you want removing, e.g. test ROIs\n",
    "\n",
    "# Only work with primary tumours\n",
    "adata = adata[adata.obs['Type']=='Primary'].copy()\n",
    "\n",
    "# Print a list of the variables that appear in the working anndata\n",
    "print(adata.var_names)\n",
    "\n",
    "# Print the structure of the working anndata\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f67e3b0b-1d57-4903-8f1a-5eaee5b1534b",
   "metadata": {},
   "source": [
    "#### Convert some columns to category\n",
    "This will save headaches in later stages - makes sure that all categorical variables in obs are stored as categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fc3295e-1e9a-42cc-a2a1-14fa238bfd11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will list out the columns you may want to now convert...\n",
    "adata.obs.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3daac3-f751-44a2-90d8-3383df23c0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alter this with the categorical obs from above\n",
    "for i in ['ROI','Case','Type','HEClass','HEClass_2class','HEClass_3class','HEClass_niches']:\n",
    "    adata.obs[i]=adata.obs[i].astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeffab90-b5a4-426c-ae59-c90974d5cc43",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Very quick QC check of data\n",
    "Some simple QC checks - look at expression of all the markers in each ROI in a heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc50faca-1b69-4d93-9d3f-af2913895815",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The title of  the column in which defines each image/region of interest\n",
    "image_var = 'ROI'\n",
    "\n",
    "# Plot a heat of all the markers\n",
    "sc.pl.matrixplot(adata, adata.var_names, groupby=image_var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927401ad-72b3-4a49-877e-e2d7dad64c3e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Batch integration using BBKNN </font>\n",
    "This will batch correct using BBKNN. There are other options batch correction, such as Harmony, but I found this the easiest to implement. Read more here: https://bodenmillergroup.github.io/IMCDataAnalysis/batch-effects.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8379879c-29ed-4c01-bbe1-57401a952293",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Run PCA and BBKNN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d42b85-d284-4b72-a341-8611a585d5ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define the number of PCA dimensions to work with - one less than number of markers. Without this, it usually defaults to 50, which we don't have enough markers for.\n",
    "n_for_pca = len(adata.var_names)-1\n",
    "\n",
    "# Define the 'obs' which defines the different cases\n",
    "batch_correction_obs = 'Case'\n",
    "\n",
    "# Calculate PCA, this must be done before BBKNN\n",
    "sc.tl.pca(adata, n_comps=n_for_pca)\n",
    "\n",
    "# BBKNN - it is used in place of the scanpy 'neighbors' command that calculates nearest neighbours in the feature space\n",
    "sc.external.pp.bbknn(adata, batch_key='Case', n_pcs=n_for_pca)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "151a56eb-b8f3-4420-b356-e32f593cc371",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> UMAPS</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e4c8377-78f8-42fd-8c61-505e301c4e28",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Calculate UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1e3bd5-0127-470f-9d81-6d5bc8682882",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d9bc00-a9bf-4a45-893d-aa12fdf0fc59",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Plot UMAPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339fc527-43ef-400f-a4ef-776d2cc20034",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Define a list of .obs to colour the UMAP by\n",
    "UMAP_groups = ['Case','ROI', 'HEClass_2class', 'HEClass_3class', 'HEClass_niches'] \n",
    "\n",
    "# Plot UMAPs coloured by list above\n",
    "sc.pl.umap(adata, color=UMAP_groups, ncols=1, size=3)\n",
    "\n",
    "# This will plot a UMAP for each of the individual markers\n",
    "sc.pl.umap(adata, color=adata.var_names.tolist(), color_map='plasma', ncols=4, size=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fea9870-3343-4d94-9b42-0e5836c2b9c4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Population identification</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed607813-cec3-4619-a545-7c0c8d995809",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Initial population identification using Leiden clustering\n",
    "This is the same algorithm used by PhenoGraph, and is more or less the same"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27fe8339-fa7a-41c1-a9c1-cb1b98cc3cf0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Initial Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6caa8c74-68bb-480b-b799-b9390ee5fddc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Change this to alter the size of the clusters - small resolution results in bigger clusters (therefore, less clusters overall, but each with more cells)\n",
    "resolution = 1\n",
    "leiden_obs_name = 'leiden_'+str(resolution)\n",
    "\n",
    "# This will perform the clustering, then add an 'obs' with name specified above, e.g leiden_0.35\n",
    "sc.tl.leiden(adata, resolution=resolution, key_added = leiden_obs_name) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e476007f-641e-4ec0-b41d-131794c45901",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plot initial clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a8cb082-8b3a-4517-bb04-6174315b88df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "image_var = 'ROI' # The adata.obs that defines each image or ROI\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))\n",
    "sc.pl.umap(adata, color=leiden_obs_name,size=2)\n",
    "\n",
    "# Plot a heatmap\n",
    "sc.pl.matrixplot(adata,\n",
    "                 adata.var_names.tolist(), \n",
    "                 groupby=leiden_obs_name, \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True)\n",
    "\n",
    "# Alter this list with adata.obs variables that you want to plot against the resulting leiden population\n",
    "for i in ['Case']:\n",
    "\n",
    "    mikeimc_v2.grouped_graph(adata,\n",
    "                             ROI_id=image_var,\n",
    "                             group_by_obs=i,\n",
    "                             x_axis=leiden_obs_name,\n",
    "                             fig_size=(10,3),\n",
    "                             log_scale=True,\n",
    "                            display_tables=False) #If you change display_tables to True, will also do stats on the groups\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "669bbd2d-adb6-4761-8889-56086322ba84",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Sub clustering example\n",
    "Once you've done the initital subclustering, which could be with a fairly low resolution, you can then do subclustering on specific clusters from previous Leiden analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db411f4-f072-4f9d-9bcf-2f0137a03983",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Change this to alter the size of the clusters in the subclustering\n",
    "resolution = 0.2\n",
    "leiden_obs_name = 'leiden_'+str(resolution)\n",
    "\n",
    "# In the following example, it will do a leiden clustering based upon the the results of 'leiden_1', selecting population 2\n",
    "sc.tl.leiden(adata, resolution=resolution, key_added = leiden_obs_name, restrict_to=('leiden_1',['2'])) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de741b91-4d11-4c40-8eef-c5ca618d43b7",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plot final clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63cc9c9b-fb0e-4632-8b18-184d4ea31fd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "image_var = 'ROI' # The adata.obs that defines each image or ROI\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))\n",
    "sc.pl.umap(adata, color=leiden_obs_name,size=2)\n",
    "\n",
    "# Plot a heatmap\n",
    "sc.pl.matrixplot(adata,\n",
    "                 adata.var_names.tolist(), \n",
    "                 groupby=leiden_obs_name, \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True)\n",
    "\n",
    "# Alter this list with adata.obs variables that you want to plot against the resulting leiden population\n",
    "for i in ['Case']:\n",
    "\n",
    "    mikeimc_v2.grouped_graph(adata,\n",
    "                             ROI_id=image_var,\n",
    "                             group_by_obs=i,\n",
    "                             x_axis=leiden_obs_name,\n",
    "                             fig_size=(10,3),\n",
    "                             log_scale=True,\n",
    "                            display_tables=False) #If you change display_tables to True, will also do stats on the groups\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23d5802f-0633-4cea-abd6-ce608f42b751",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Label final populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "658fbc55-b1ac-4b00-94db-b065cef8bcd0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Remapping populations\n",
    "Here we are creating 'dictionaries' that will remap the population numbers from the Leiden clustering, to actual names.\n",
    "\n",
    "I've usually been working with 3 different labels.  \n",
    "**populations** - This is the exact population name for that Leiden population. If you give two numbers the same name, they will be combined (e.g. 10 and 14 below).\n",
    "\n",
    "**hierarchy** - This gives a more broad hierarchy for the populations, allowing us to specicfy analyses only be done on certain cell types.\n",
    "\n",
    "**populations_broad** - This is a combination of populations and hierarchy, and what I've mostly used in subsequent analyses. For my analyses that mostly focused on myeloid cells, I decided to combine all the different tumour types into one label ('Tumour'), but still keep all the myeloid populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee9b79c9-4faa-4ea5-8269-51d0a968db42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "leiden_obs_name = 'leiden_final'\n",
    "\n",
    "populations={'0':'Tumour, Sox2+Nestin+',\n",
    "'1':'Tumour,  Cd44+S100B+',\n",
    "'2':'TAM/Micro intermediate',\n",
    "'3':'Tumour, Sox2+Olig2+',\n",
    "'4':'Tumour, Cd44+',\n",
    "'5':'Tumour, Sox2+',\n",
    "'6':'Tumour, Cd44+GFAP+Vimentin+',\n",
    "'7':'Tumour, GLUT1+',\n",
    "'8':'Tumour, pERK+',\n",
    "'9':'Stroma',\n",
    "'10':'Stroma',\n",
    "'11':'Astrocytes',\n",
    "'12':'Microglia, resting',\n",
    "'13':'Vascular',\n",
    "'14':'Stroma',\n",
    "'15':'Tumour, Sox2+',\n",
    "'16,0':'TAMs, Cd206+',\n",
    "'16,1':'TAMs, Cd74+VISTA+',\n",
    "'17':'Tumour, Sox2+Nestin+Ki67+',\n",
    "'18,0':'Microglia, Cd16+VISTA+',\n",
    "'18,1':'Microglia',\n",
    "'19':'Vascular',\n",
    "'20':'Tumour, Cd44+',\n",
    "'21':'Myeloid, Cd68+',\n",
    "'22':'Immune, HIF1a+',\n",
    "'23':'Cd8 lymphocytes'\n",
    "}\n",
    "\n",
    "hierarchy={'0':'Tumour',\n",
    "'1':'Tumour',\n",
    "'2':'Myeloid',\n",
    "'3':'Tumour',\n",
    "'4':'Tumour',\n",
    "'5':'Tumour',\n",
    "'6':'Tumour',\n",
    "'7':'Tumour',\n",
    "'8':'Tumour',\n",
    "'9':'Stroma',\n",
    "'10':'Stroma',\n",
    "'11':'Stroma',\n",
    "'12':'Myeloid',\n",
    "'13':'Vascular',\n",
    "'14':'Stroma',\n",
    "'15':'Tumour',\n",
    "'16,0':'Myeloid',\n",
    "'16,1':'Myeloid',\n",
    "'17':'Tumour',\n",
    "'18,0':'Myeloid',\n",
    "'18,1':'Myeloid',\n",
    "'19':'Vascular',\n",
    "'20':'Tumour',\n",
    "'21':'Myeloid',\n",
    "'22':'Immune',\n",
    "'23':'Lymphocytes'\n",
    "}\n",
    "\n",
    "population_broad={\n",
    "'0':'Tumour',\n",
    "'1':'Tumour',\n",
    "'2':'TAM/Micro intermediate',\n",
    "'3':'Tumour',\n",
    "'4':'Tumour',\n",
    "'5':'Tumour',\n",
    "'6':'Tumour',\n",
    "'7':'Tumour',\n",
    "'8':'Tumour',\n",
    "'9':'Stroma',\n",
    "'10':'Stroma',\n",
    "'11':'Stroma',\n",
    "'12':'Microglia, resting',\n",
    "'13':'Vascular',\n",
    "'14':'Stroma',\n",
    "'15':'Tumour',\n",
    "'16,0':'TAMs, Cd206+',\n",
    "'16,1':'TAMs, Cd74+VISTA+',\n",
    "'17':'Tumour',\n",
    "'18,0':'Microglia, Cd16+VISTA+',\n",
    "'18,1':'Microglia',\n",
    "'19':'Vascular',\n",
    "'20':'Tumour',\n",
    "'21':'Myeloid, Cd68+',\n",
    "'22':'Immune, HIF1a+',\n",
    "'23':'Cd8 lymphocytes'\n",
    "}    \n",
    "    \n",
    "adata.obs['population']=adata.obs[leiden_obs_name].map(populations).astype('category')\n",
    "adata.obs['population_broad']=adata.obs[leiden_obs_name].map(population_broad).astype('category')\n",
    "adata.obs['hierarchy']=adata.obs[leiden_obs_name].map(hierarchy).astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "738835b9-9641-448e-97e6-4e6e2f38059d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plotting final populations with new names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d55ec6-f33b-4244-8ed6-cb5c02f85387",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in ['population','population_broad','hierarchy']:\n",
    "    sc.set_figure_params(figsize=(5, 5), dpi=200)\n",
    "    sc.pl.umap(adata, color=i,size=2)\n",
    "        \n",
    "    sc.pl.matrixplot(adata,\n",
    "                     adata.var_names, \n",
    "                     groupby=i, \n",
    "                    var_group_rotation = 0,\n",
    "                    vmax=0.7,\n",
    "                    dendrogram=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbefaa6-9f47-4f3f-9608-8c0e6ffe1a5f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Spatial mapping of cells</font>\n",
    "Cell populations can be spatially plotted using a couple of different ways. In this example, we just express cells as dots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436b54e8-0d97-4360-bfe5-cca20f3c010e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sb.set_style(\"whitegrid\", {'axes.grid' : False})\n",
    "\n",
    "graph = sb.lmplot(data = adata.obs,\n",
    "                  x = 'X_loc',\n",
    "                  y='Y_loc',\n",
    "                  hue='population_broad', #This is the adata.obs that you want to plot\n",
    "                  height = 4,\n",
    "                  col = 'ROI',\n",
    "                  col_wrap = 3,\n",
    "                  fit_reg = False,\n",
    "                  aspect=1,\n",
    "                  scatter_kws={\"s\": 5})\n",
    "#graph.savefig(\"figures/MappedPlots\"+a+\".png\")    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebda1c5b-064c-42e8-a548-6b04f9408768",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='orange'> ATHENA setup</font>\n",
    "**<font color='red'>This section will need editing depending how your data is layed out, carefully read and alter as appropriate, and will need you to create some folders and files to use masks</font>**\n",
    "\n",
    "This will transfer the Adata into ATHENA for more complex spatial analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8bd0fe3-e355-490d-98b5-5c4cd2100bb0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Transfer data into ATHENA format ('so' object)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a03159b0-966f-481f-81fd-4450e6744fe4",
   "metadata": {},
   "source": [
    "Make a copy of the adata to import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a89e780-0bc0-4dca-b10a-5851abdb86b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = adata.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f66b887e-b031-40e9-ac8a-c17b5147c763",
   "metadata": {},
   "source": [
    "Add a cell_index - numbers each cell in each sample (starting again at 0 for each sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6742e5af-d258-428e-9581-f198a1f30650",
   "metadata": {},
   "outputs": [],
   "source": [
    "for roi in ad.obs.ROI.cat.categories:\n",
    "    ad.obs.loc[ad.obs.ROI==roi,'cell_id'] = ad.obs.loc[ad.obs.ROI==roi,'Master_Index'] - ad.obs.loc[ad.obs.ROI==roi,'Master_Index'].min() + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d94c3756-7147-4674-a8d1-086cf1d08dd6",
   "metadata": {},
   "source": [
    "Convert the group names into ids - numbers for each group. This is required for some of athenas plotting functions. Here, we just convert each group name into a number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c71ad2e9-c101-4a98-85f1-8d69dd55392c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs['hierarchy_id'] = ad.obs.groupby('hierarchy').ngroup()\n",
    "ad.obs['population_id'] = ad.obs.groupby('population').ngroup()\n",
    "ad.obs['population_broad_id'] = ad.obs.groupby('population_broad').ngroup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83714f1f-866a-4313-94e4-5e148c515df9",
   "metadata": {},
   "source": [
    "Make sure the columns are set to 'category' type, to enable correct plotting. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60da323d-dc1e-41fe-b2a3-2a6fbf94d8f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a2dce9-677a-4f61-a5d7-1f80994b332b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in ['ROI', 'Case', 'hierarchy_id', 'population_id','population_broad_id']:\n",
    "    ad.obs.loc[:, i] = ad.obs[i].astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b872acc9-24bd-4ca4-b073-04a8c1ad2953",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Construct the SpatialOmics object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97dca0d3-c097-41d9-aaa7-3bbd783bccdb",
   "metadata": {},
   "source": [
    "We need to create a new data frame that containts the 'sample level' meta data, these are currently stored in the adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf1dbd2e-2a0f-4d23-80e3-67f1b17934f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "spl = ad.obs[['Case', 'ROI', 'Type']]  #These are the sample level obs in the adata\n",
    "spl = spl[~spl.duplicated()]\n",
    "spl.set_index('ROI', inplace=True) #Set the index as ROI, which is the unique ID for each region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c839a81-0034-459c-aa79-184db65b8bbb",
   "metadata": {},
   "source": [
    "If you use the Bodenmiller pipeline, the masks are created in the 'CP_Output' folder. Unfortunately they all have quite confusing names that won't match with your ROI names. You can find out which masks match to which ROI by looking at the Image.csv file in the same folder as the outputted masks. There is a row per image analysed (per ROI), the columns you are looking for are 'Metadata_description' and 'Metadata_acname' to work out which ROI each line coresponds to, then look for 'URL_cellmask' to get the file name. I then created a subfolder called 'masks', and copied the masks into there. If you look at the 'mask_dict.csv' example here, you can see how I then layed out the dictionary file that will point each ROI to it's relevant mask file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df5c4ecd-8e6f-4be5-8a6b-9790ca266839",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Open the mask dictionary\n",
    "map_file = pd.read_csv('mask_dict.csv').set_index('ROI').to_dict()['mask_path']\n",
    "\n",
    "#Map into the spl dataframe\n",
    "spl.loc[:, 'cell_mask_file'] = spl.index.map(map_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3551075e-f6f2-4534-b3b6-6208406e73df",
   "metadata": {},
   "source": [
    "Create the SpatialOmics instance and add in the sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4028ccbd-c698-4d21-92d6-1e3485821e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "so = SpatialOmics()\n",
    "so.spl = spl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f076b4b9-3f2f-4f7a-bfd4-ee4dde1c6c15",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Add in cell masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d8bde46-c45d-4623-82ee-fbffb36f29dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "for r in so.spl.index:\n",
    "    mask = ad.obs.ROI == r\n",
    "    so.X[r] = pd.DataFrame(ad.X[mask], columns=ad.var.index)\n",
    "    so.obs[r] = ad.obs[mask]\n",
    "    so.obs[r].set_index('cell_id', inplace=True)\n",
    "    so.X[r].index = so.obs[r].index\n",
    "\n",
    "    # this is how you can add masks to the spatial omics instance\n",
    "    # please use `to_store=False` as this prevents writing the file to disk which is still experimental\n",
    "    cell_mask_file = spl.loc[r].cell_mask_file\n",
    "\n",
    "    # first argument is the sample name\n",
    "    # second argument is the KEY in so.masks[KEY] under which the mask is stored\n",
    "    # third argument the file name\n",
    "    so.add_mask(r, 'cellmasks', cell_mask_file, to_store=False)\n",
    "    so.masks[r]['cellmasks'] = so.masks[r]['cellmasks'].astype(int)  # should be int\n",
    "\n",
    "    # process segmentation masks and remove masks that do not represent a cell\n",
    "    existing_cells = set(so.obs[r].index)\n",
    "    segmentation_ids = set(np.unique(so.masks[r]['cellmasks']))\n",
    "    idsToDelete = segmentation_ids - existing_cells\n",
    "    for i in idsToDelete:\n",
    "        cm = so.masks[r]['cellmasks']\n",
    "        cm[cm == i] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9acc47a-13bb-43a1-8205-7da86a8a659f",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Check that the number of cell IDs matches the number of cell masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b073dba2-9f22-4a82-9a5f-7ba181e128a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for spl in so.spl.index:\n",
    "   ids = np.unique(so.masks[spl]['cellmasks'])\n",
    "   ids = ids[ids != 0]  # drop the background identifier\n",
    "   if len(so.obs[spl]) == len(ids):\n",
    "       print(f'Sample: {spl} has {len(ids)} cell identifiers in cellmasks but {len(so.obs[spl])} cells in so.obs[spl]')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88150012-f882-4027-b2ae-c48cbf61a532",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Colour maps</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb63b6e-732b-4acc-9cf2-8d41c032e97c",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Create colour maps dictionaries and labels\n",
    "This will create dictionaries that pair up the id numbers to the actual names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2403c07-c279-494d-9f4c-f3e0f0b54690",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_dict = dict(zip(ad.obs['population_id'].cat.categories, ad.obs['population'].cat.categories))\n",
    "hierarchy_dict = dict(zip(ad.obs['hierarchy_id'].cat.categories, ad.obs['hierarchy'].cat.categories))\n",
    "population_broad_dict = dict(zip(ad.obs['population_broad_id'].cat.categories, ad.obs['population_broad'].cat.categories))\n",
    "\n",
    "so.uns['cmap_labels'].update({'population_id': population_dict})\n",
    "so.uns['cmap_labels'].update({'hierarchy_id': hierarchy_dict})\n",
    "so.uns['cmap_labels'].update({'population_broad_id': population_broad_dict})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05c69b71-4dbf-4817-a6d5-bfd1bf46d84a",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Assign colour map\n",
    "This will assign colours in the order they appear in the dictionaries created above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86362e23-3f86-4254-83eb-0db8c9d7037b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify different colour maps here\n",
    "colourmap = cc.glasbey_category10\n",
    "\n",
    "\n",
    "for i in ['hierarchy_id','population_id','population_broad_id']:\n",
    "    length = len(so.obs[so.spl.index[0]][i].cat.categories)\n",
    "    cmap = colourmap[:length]\n",
    "    cmap = ListedColormap(cmap)\n",
    "    print(i)\n",
    "    display(cmap)\n",
    "    so.uns['cmaps'].update({i: cmap})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "724b3a98-3015-4025-b2e7-0124a22bb285",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Duplicate colour maps in format friendly for matplotlib and scanpy etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a306d169-c6fd-484e-8aca-dd45df6aad7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "items = list(population_dict.values())\n",
    "colours = so.uns['cmaps']['population_id'].colors\n",
    "population_colours = {items[i]: colours[i] for i in range(len(items))}\n",
    "\n",
    "items = list(population_broad_dict.values())\n",
    "colours = so.uns['cmaps']['population_broad_id'].colors\n",
    "population_broad_colours = {items[i]: colours[i] for i in range(len(items))}\n",
    "\n",
    "items = list(hierarchy_dict.values())\n",
    "colours = so.uns['cmaps']['hierarchy_id'].colors\n",
    "hierarchy_colours = {items[i]: colours[i] for i in range(len(items))}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb23029-6a08-4075-8d6f-cd20eda6afe5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> ATHENA calculations</font>\n",
    "Most of this section you should just be able to run 'as is', but it can take a **long time!** Best to leave running over night!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52858523-46bf-4b59-9a5e-35deda09fb16",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Graph construction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95afd06-605b-45d4-a617-a4d35f878287",
   "metadata": {},
   "source": [
    "This will construct the different spatial connectivity graphs, these define which cells are considered neighbours for later analyses.\n",
    "\n",
    "use your extracted cell locations to build a {radius, knn} graph (it does not use the cellmasks)\n",
    "\n",
    "you need to define the 'r' parameter to a specific ROI, otherwise the code runs only for the last ROI in the SO object. Or the original code does not work."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bef1d0f-524a-4ab5-a0df-260dbbdbf797",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82c0d1e8-c769-46b6-a2a2-d36f4c12a8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This select which samples to use - all by default\n",
    "samples = so.spl.index \n",
    "\n",
    "# Import the default graph settings\n",
    "from athena.graph_builder.constants import GRAPH_BUILDER_DEFAULT_PARAMS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3421d537-877c-4088-b736-09eceb3124f1",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Extract centroids from masks\n",
    "We can define where cells are in two ways - either from the centre of the mask or from the cell locations extracted from the cell profiler pipeline. In theory, they should be almost identical, and in previous tests they are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0435d738-81f9-49c4-8c97-7e4aa69c9e1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Extract cell centroids across all samples\n",
    "for r in tqdm(so.obs.keys()):\n",
    "    sh.pp.extract_centroids(so, r, mask_key='cellmasks')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d0a884-cdcf-4020-b9e9-79908f598eb0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### KNN graph\n",
    "k nearest neighbors - change 'number_neighbours' to define number of nearest neighbors to identify per cell. \n",
    "\n",
    "The graph key will be knn_[number_neighbours]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a528a6-0ddb-4f25-be39-0ad01d3dcf30",
   "metadata": {},
   "outputs": [],
   "source": [
    "number_neighbours = 6\n",
    "\n",
    "#kNN graph, without cellmasks, using predetermined X, Y locs\n",
    "config = GRAPH_BUILDER_DEFAULT_PARAMS['knn']\n",
    "config['builder_params']['n_neighbors'] = number_neighbours # set parameter k\n",
    "for spl in tqdm(samples):\n",
    "             sh.graph.build_graph(so, spl,\n",
    "                     key_added='knn_'+str(number_neighbours),\n",
    "                     builder_type='knn', mask_key=None, coordinate_keys=('X_loc', 'Y_loc'), config=config) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac9bc5f-921f-4b5d-baee-4d4f75ad94b2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Radius graph\n",
    "Defines a cells neighbours using a radius (in um) from the cell\n",
    "\n",
    "The graph key will be radius_[radius]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c75395-9df5-4770-86da-657075bc72f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "radius = 20\n",
    "\n",
    "# radius graph without cellmasks, using my predetermined X, Y locs\n",
    "config = GRAPH_BUILDER_DEFAULT_PARAMS['radius']\n",
    "config['builder_params']['radius'] = radius\n",
    "for spl in tqdm(samples):\n",
    "             sh.graph.build_graph(so, spl,\n",
    "                     key_added='radius_'+str(radius),\n",
    "                     builder_type='radius', mask_key=None, coordinate_keys=('X_loc', 'Y_loc'), config=config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d65c5c9b-bb6c-4855-b83c-ddaad4b3dac3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Contact graph (with multiprocessing)\n",
    "Defines neighbours using direct cell-cell contact using cell mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68838a0-9558-4a01-acf6-5c896d558421",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# This select which samples to use - all by default\n",
    "samples = so.spl.index.tolist() #so.spl.index \n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Run the paralell processing\n",
    "if parallel==True:\n",
    "    \n",
    "    contact_graph_partial = partial(mikeimc_v2.contact_graph, so)\n",
    "    \n",
    "    with Pool(processes = num_cores) as pool:      \n",
    "        data = pool.map(contact_graph_partial, tqdm(samples))\n",
    "        print('Complete')\n",
    "    \n",
    "    for spl,grph in zip(samples,data):\n",
    "\n",
    "        if spl in so.G:\n",
    "            so.G[spl].update({'contact': grph})\n",
    "        else:\n",
    "            so.G[spl] = {'contact':grph}\n",
    "    \n",
    "    print('Added back into original SO!')\n",
    "    \n",
    "elif parallel==False:\n",
    "    \n",
    "    for spl in samples:\n",
    "        try:\n",
    "            sh.graph.build_graph(so, spl, builder_type='contact', mask_key='cellmasks')\n",
    "        except KeyboardInterrupt:\n",
    "            pass\n",
    "        except BaseException as err:\n",
    "            print(\"An exception occurred in calculating contact graph for \" + spl)\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e285a2d-44e1-4adf-819e-54375db6a14c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell-level scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d208e360-cdd6-4874-ab7d-3577b54619ef",
   "metadata": {},
   "source": [
    "In their original denition, all entropic scores described above do not take the spatial component into account when calculated on a whole tumor level. For this reason, ATHENA implements two avors of these scores: a global avor, in which the metrics are computed at a whole sample level using only the phenotype distribution, and a local avor, in which the scores are computed at a single-cell level, using also the graph structure. Specically, when computing local scores, ATHENA iterates over all cells, and for each cell, computes the local entropy within its neighborhood. In this way, highly diverse regions where cells from multiple diferent phenotypes coexist can be highlighted, and, instead of computing a single entropy value as for the global avor, a distribution of entropic values is returned.\n",
    "\n",
    "Cell-level scores quantify heterogeneity in a spatial manner, accounting for local effects, and return a value per single cell, saved in so.obs. To apply these scores to the data we use again .metrics but this time with local=True. Since these scores heavily depend on the tumor topology, the graph type and occasionally additional parameters also need to be provided."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34d150c1-da7a-40d1-91b4-371d0c466189",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Calculate richness, shannon and quadratic entropy (with multiprocessing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "510ea38e-1188-495c-bac7-0c067988d54b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "# Population identifiers\n",
    "obs = ['population_id','population_broad_id']\n",
    "\n",
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# Get a list of all samples\n",
    "all_samples = so.spl.index.values\n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Get a list of graphs\n",
    "graph_keys = ['contact']#list(so.G[all_samples[0]].keys())\n",
    "\n",
    "# Setting up lists for either parallel or normal processing\n",
    "s_list=[]\n",
    "o_list=[]\n",
    "g_list=[]\n",
    "\n",
    "for g in graph_keys:\n",
    "    for s in all_samples:\n",
    "        for o in obs:    \n",
    "            o_list.append(o)\n",
    "            s_list.append(s)\n",
    "            g_list.append(g)\n",
    "\n",
    "pool_func_input=list(zip(s_list,o_list,g_list))\n",
    "\n",
    "# To keep the same 'so' object passed every time\n",
    "cell_metrics_partial = partial(mikeimc_v2.cell_metrics, so)\n",
    "\n",
    "# Parallel processing option\n",
    "if parallel==True:\n",
    "       \n",
    "    with Pool(processes = num_cores) as pool:\n",
    "        data = pool.starmap(cell_metrics_partial, tqdm(pool_func_input, total=len(pool_func_input)))\n",
    "        print('Complete')\n",
    "\n",
    "    data_iter = iter(data)\n",
    "    for s, _, _ in pool_func_input:        \n",
    "        so.obs[s] = pd.concat((so.obs[s], next(data_iter)), axis=1)\n",
    "\n",
    "    print('Added back into original SO!')\n",
    "        \n",
    "        \n",
    "# Non-parallel processing option        \n",
    "elif parallel==False:\n",
    "\n",
    "    for s, o, g in tqdm(func_input):\n",
    "\n",
    "        try:\n",
    "            sh.metrics.richness(so, s, o, local=True, graph_key=g)\n",
    "            sh.metrics.shannon(so, s, o, local=True, graph_key=g)\n",
    "            sh.metrics.quadratic_entropy(so, s, o, local=True, graph_key=g, metric='cosine')              \n",
    "        except KeyboardInterrupt:\n",
    "            pass  \n",
    "        except BaseException as err:\n",
    "            print(f'Error caclculating sample:{s}, graph:{g}, observation:{o}')\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")  \n",
    "\n",
    "# This will delete duplicate columns in the .obs\n",
    "#for i in all_samples:\n",
    "#    so.obs[i]=so.obs[i].loc[:,~so.obs[i].columns.duplicated()].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af24d55-9912-4539-8063-e0eef991b467",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell type interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0fb883b-8049-4f93-8650-678345a2780e",
   "metadata": {},
   "source": [
    "mode: it can be classical, histocat or proportion\n",
    "\n",
    "Classic: In the classic avor of the neighborhood analysis score, the average number of interactions between cells of the same phenotype is computed.\n",
    "\n",
    "HistoCAT: In the histoCAT avor, the global average of interaction score is only computed across cells that actually show this interaction.\n",
    "\n",
    "Proportional: Finally, the proportional avor uses interaction frequencies instead of counts, i.e., divides the counts of pairwise interactions by all interactions a given cell has. We propose this approach as an alternative that normalizes the scores with respect to varying cell density. In contrast to the classic or the histoCAT avor, the proportional avor is not inuenced by the number of cells in a sample and the score is bounded in the range of [0,1].\n",
    "\n",
    "graph_key can be knn, radius or contact\n",
    "\n",
    "prediction_type can be observation, pvalue or diff\n",
    "In addition to a p-value, we propose to compute the dierence between the observed proportion of interactions obs ij between cell type i ! j and the randomised proportion of interactions rand ij . This dierence can be asymmetrical (i ! j = 6 j ! i), is bounded in the range of [-1,1], and might be more suited as input for certain machine learning models."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "682a2e63-fa98-49d4-98df-e63839cd1bb5",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Calculate all potential interaction metrics (with multiprocessing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f29d9c7-daa4-48b9-b1bb-80bad99c3294",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Get a list of all samples\n",
    "all_samples = so.spl.index.values\n",
    "\n",
    "# Modes to run the analyses in\n",
    "calc_modes = ['proportion']#['proportion', 'histoCAT', 'classic']\n",
    "\n",
    "# Population identifiers\n",
    "pop_ids = ['population_id','population_broad_id', 'hierarchy_id']\n",
    "\n",
    "# Get a list of all graphs\n",
    "graph_keys = ['contact']#list(so.G[all_samples[0]].keys())\n",
    "\n",
    "\n",
    "import logging\n",
    "logging.getLogger().setLevel(logging.ERROR)  # set logger to logging.INFO if you want to see more progress information\n",
    "\n",
    "# Create blank lists\n",
    "m_list=[]\n",
    "o_list=[]\n",
    "s_list=[]\n",
    "g_list=[]\n",
    "\n",
    "# Creates several parallel lists with the order of inputs we will need to call in the pool function\n",
    "for m in calc_modes:\n",
    "    for o in pop_ids:\n",
    "        for s in all_samples:\n",
    "            for g in graph_keys:\n",
    "                m_list.append(m)\n",
    "                o_list.append(o)\n",
    "                s_list.append(s)\n",
    "                g_list.append(g)\n",
    "\n",
    "neigh_int_partial = partial(mikeimc_v2.neigh_int, so)\n",
    "\n",
    "# Create a 'list of lists' for the pool.starmap\n",
    "pool_func_input=list(zip(m_list,o_list,s_list,g_list))\n",
    "\n",
    "\n",
    "if parallel==True:\n",
    "          \n",
    "    with Pool(processes = num_cores) as pool:\n",
    "        data = pool.starmap(neigh_int_partial, tqdm(pool_func_input, total=len(pool_func_input)))\n",
    "        print('Multiprocessing complete')\n",
    "                      \n",
    "    data_iter = iter(data)\n",
    "    for m, o, s, g in pool_func_input:\n",
    "        key = f'{o}_{m}_diff_{g}'\n",
    "        sh.neighborhood.estimators.add2uns(so, next(data_iter), s, 'interactions', key) \n",
    "\n",
    "    print('Added back into original SO!')\n",
    "            \n",
    "elif parallel==False:\n",
    "    \n",
    "    for m, o, s, g in tqdm(func_input):\n",
    "        try:\n",
    "            sh.neigh.interactions(so, s, o, mode=m, prediction_type='diff', graph_key=g)    \n",
    "        except KeyboardInterrupt:\n",
    "            pass  \n",
    "        except BaseException as err:\n",
    "            print(f'Error caclculating sample:{s}, graph:{g}, observation:{o}, mode:{m}')\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")\n",
    "            \n",
    "logging.getLogger().setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "464b2752-8de5-43bb-8cbb-a401917a3172",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='orange'> Cell environment analysis </font>\n",
    "This requires the markers of interest to be in a folder named after the marker, containing an image for each ROI that is named after the image, e.g.\n",
    "\n",
    "GFAP\n",
    "- ROI1.png\n",
    "- ROI2.png\n",
    "- ROI3.png"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41e35daf-3b83-41c3-9491-fbd530782522",
   "metadata": {},
   "source": [
    "This will extract information about the marker expression in a circle around the celll. This will require that you have the images in a folder called the marker name, with each image in that directory named after its ROI."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c837e138-5762-4f1e-aa56-b43869a95a56",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Multi-threading version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d60bf41f-202d-4b53-aaf4-2b5b75b74cc9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "from analyse import analyse_cell\n",
    "import numpy as np\n",
    "from multiprocessing import Pool\n",
    "import itertools\n",
    "from tqdm import tqdm\n",
    "from skimage.draw import disk, ellipse\n",
    "from skimage.measure import label, regionprops, regionprops_table\n",
    "import skimage.io as skio\n",
    "\n",
    "# Setup\n",
    "radius=20 #Define the radius of the circle to measure\n",
    "marker_list=['GLUT1','pERK','GFAP', 'CD44'] #The markers to analyse\n",
    "x_loc = 'X_loc'\n",
    "y_loc = 'Y_loc'\n",
    "cell_index_id = 'Master_Index'\n",
    "num_cores = 4\n",
    "images = adata.obs.ROI.unique()\n",
    "\n",
    "# Make blank lists which will be re-used to make dictionaries\n",
    "cell_id=[]\n",
    "intensity=[]\n",
    "\n",
    "for marker in marker_list:\n",
    "    print('Marker: ' +marker)\n",
    "\n",
    "    for i in images:\n",
    "        print('Image: '+i)\n",
    "\n",
    "        adata_roi = adata.obs[adata.obs.ROI==i]\n",
    "        number_of_cells = range(len(adata_roi))\n",
    "\n",
    "        roi_cells_index = adata_roi[cell_index_id]            \n",
    "      \n",
    "        raw_image  = skio.imread(marker+\"/\"+i+\".tiff\")\n",
    "        \n",
    "        raw_image_list = [raw_image for i in number_of_cells]\n",
    "        size0 = [len(raw_image[0]) for i in number_of_cells]\n",
    "        size1 = [len(raw_image[1]) for i in number_of_cells]                     \n",
    "        radius_list = [radius for i in number_of_cells]\n",
    "        roi_cells_x = adata_roi[x_loc]\n",
    "        roi_cells_y = adata_roi[y_loc]\n",
    "                     \n",
    "        analyse_cells_inputs=list(zip(raw_image_list,size0,size1,radius_list,roi_cells_x,roi_cells_y))\n",
    "                 \n",
    "        #analyse_cell(raw_image, size0, size1, radius, cell_axis0, cell_axis1)\n",
    "        \n",
    "        with Pool(processes = num_cores) as pool:\n",
    "            intensity_data = pool.starmap(analyse_cell, tqdm(analyse_cells_inputs, total=len(raw_image_list)))\n",
    "\n",
    "        cell_id.extend(roi_cells_index)\n",
    "        intensity.extend(intensity_data)\n",
    "    \n",
    "    # Create a mapping dictionary for all the cells\n",
    "    mapping_dict = dict(zip(cell_id,intensity))\n",
    "\n",
    "    # Map into a new obs in the original adata\n",
    "    adata.obs[(marker+\"_\"+str(radius))]=adata.obs[cell_index_id].map(mapping_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1522ca76-249e-469e-977e-cf1b9ae5259e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Cell environment plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6fc13f3-b244-49a0-9ca4-e953b2e19054",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plotting_data = adata.obs.copy()\n",
    "\n",
    "for i in ['pERK_20','GLUT1_20','GFAP_20','CD44_20']: #This is the list of markers        \n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(6, 3), dpi=150)\n",
    "    fig.tight_layout()    \n",
    "    \n",
    "    #Plotting\n",
    "    sb.barplot(data = plotting_data, \n",
    "               y = i, \n",
    "               x='population',\n",
    "               ci=68,\n",
    "               ax=ax)\n",
    "\n",
    "    #Plotting settings\n",
    "    ax.tick_params(axis='x', labelrotation = 90)\n",
    "    #plt.savefig((f'{i}.png'), bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb240346-2168-44b8-80d3-4ef9777a98e9",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plot cell populations over markers\n",
    "Use this to plot the cell populations over the markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b422a25-8b99-4ac2-bfd0-94fb4d42fb82",
   "metadata": {},
   "outputs": [],
   "source": [
    "from analyse import analyse_cell\n",
    "import numpy as np\n",
    "#from multiprocessing import Process\n",
    "from multiprocessing import Pool\n",
    "import itertools\n",
    "from tqdm import tqdm\n",
    "from skimage.draw import disk, ellipse\n",
    "from skimage.measure import label, regionprops, regionprops_table\n",
    "import skimage.io as skio\n",
    "\n",
    "population= 'population_broad_id'\n",
    "figsize=(5,5)\n",
    "norm_percentile=0.95\n",
    "cmap='gist_gray'\n",
    "\n",
    "for marker in ['GLUT1','pERK','GFAP', 'CD44']:\n",
    "    \n",
    "    # First work out the percentile of all the images in the channel, so that we're normalising intensity over all the images\n",
    "    \n",
    "    raw_image_list=[]\n",
    "    \n",
    "    # Load all the images into a list\n",
    "    for i in so.spl.index:\n",
    "        raw_image_list.append(skio.imread(f'{marker}/{i}.tiff'))    \n",
    "        \n",
    "    # Convert to a numpy array and then get the quantile of all the pixels in the images (now stored in a numpy array)\n",
    "    raw_image_list  = np.array(raw_image_list)\n",
    "    norm_value = np.quantile(raw_image_list, q=norm_percentile)\n",
    "        \n",
    "        \n",
    "    for i in so.spl.index:\n",
    "        \n",
    "        # We only want to look at myeloid cells\n",
    "        so_c =so.copy()\n",
    "        so_c.obs[i]=so_c.obs[i][so_c.obs[i].hierarchy=='Myeloid']\n",
    "\n",
    "        # Open the raw image then normalise to the 90th percentile\n",
    "        raw_image  = skio.imread(f'{marker}/{i}.tiff')\n",
    "        norm_image = np.clip((raw_image / norm_value),0,1)\n",
    "        \n",
    "        fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "        # Show cell populations and then the marker image on the same axis, and then save\n",
    "        sh.pl.spatial(so_c, i, population, ax=ax, node_size=3) \n",
    "        ax.imshow(norm_image, vmin=0, vmax = 1, alpha=1, cmap=cmap)\n",
    "        \n",
    "        fig.savefig((f'{marker}/{i}.svg'), bbox_inches='tight')\n",
    "        fig.savefig((f'{marker}/{i}.png'), bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c83f7f-a267-4e60-be94-8d968bc55558",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Non-spatial graphs</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1b272d1-884f-4b9a-bd74-4edd0d14a2f8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Categorise samples using UMAP for cell abundance\n",
    "This will use the abundance of the different populations to create a UMAP, then colour it by an adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cb786e5-94a3-4048-bd8b-83767c6de4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.cellabundance_UMAP(adata,\n",
    "                              ROI_id='ROI',\n",
    "                              population='population',\n",
    "                              colour_by='HEClass_2class',\n",
    "                              annotate=True,\n",
    "                              save='roi_abundance_umap.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b766ccf3-6f18-42df-b361-654882371130",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Abundance graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6565eaab-7915-4b2b-b27f-06aaf6fb2fe9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Example breakdown of populations per case\n",
    "Here, break down the myeloid and tumour populations per case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09405008-9ea7-4784-a7ba-3f2379a7890d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_m = adata[adata.obs.hierarchy=='Myeloid']\n",
    "adata_t = adata[adata.obs.hierarchy=='Tumour']\n",
    "\n",
    "fig, axs = plt.subplots(1,3,figsize=(11, 6),constrained_layout=True)\n",
    "\n",
    "tmp = pd.crosstab(adata_m.obs['Case'],adata.obs['population'], normalize='index')\n",
    "tmp.plot.bar(color=population_colours,ax=axs[0],stacked=True).legend(bbox_to_anchor=(0.75, -0.2))#.legend(bbox_to_anchor=(1.1, 1))\n",
    "\n",
    "tmp = pd.crosstab(adata_t.obs['Case'],adata.obs['population'], normalize='index')\n",
    "tmp.plot.bar(color=population_colours,ax=axs[1],stacked=True).legend(bbox_to_anchor=(0.75, -0.2))#.legend(bbox_to_anchor=(1.1, 1))\n",
    "\n",
    "tmp = pd.crosstab(adata_t.obs['Case'],adata.obs['population_broad'], normalize='index')\n",
    "tmp.plot.bar(color=population_broad_colours,ax=axs[2],stacked=True).legend(bbox_to_anchor=(0.75, -0.2))#.legend(bbox_to_anchor=(1.1, 1))\n",
    "\n",
    "fig.savefig('pop_breakdowns.eps')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f2f98d4-aa89-4293-bb4e-590af78e29fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Abundances graphs with stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c534e636-49b0-4850-b9c4-5a445e95d7f7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Example of case averaging abundance stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e0cdec4-e7e7-435c-8206-49d2767905ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.pop_stats(adata,\n",
    "          groups='HEClass_2class', #The adata.obs we want to use to compare between\n",
    "          Case_id='Case', #The adata.obs that defines case\n",
    "          ROI_id='ROI',\n",
    "          x_axis='population',\n",
    "          display_tables=True,\n",
    "          fig_size=(8,3),\n",
    "        #save='myeloid_casegroup.png',\n",
    "           log_scale=True\n",
    "         )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b4ebd1f-9924-4e77-aeb2-4b1d8d574295",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### ROI-averaged abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99a1833-7f2f-4da1-b5fd-73640e65df27",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.grouped_graph(adata,\n",
    "          group_by_obs='HEClass_2class',\n",
    "          ROI_id='ROI',\n",
    "          x_axis='population',\n",
    "          display_tables=True,\n",
    "          fig_size=(3,3),\n",
    "        #save='graph.svg',\n",
    "           log_scale=True\n",
    "         )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da0e7340-eb3e-4dce-ada3-7b5e2fa8aefe",
   "metadata": {
    "tags": []
   },
   "source": [
    "## UMAPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5450eb43-cd2b-4bdf-9654-4614dc32210e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Categorical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b67f145-91be-466b-aba9-2cc7980f06f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in ['population','population_broad','hierarchy']:\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(5,5))\n",
    "    sc.pl.umap(adata, color=col, size=3,ax=ax, save=+col+'.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc215c5-b426-48cf-8d6a-18e33777476b",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Individual markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1bfc6cd-b163-440f-96cb-54958145f09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.figsize']=(5,5) #rescale figures, increase sizehere\n",
    "sc.set_figure_params(dpi=100, dpi_save=200) #Increase DPI for better resolution figures\n",
    "\n",
    "sc.pl.umap(adata, color=adata.var_names, size=3, ncols=4, color_map='viridis', save='umap_allmarkers.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6a7361e-e214-4a33-92fc-274d5bf11b4e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Heatmaps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55f4e404-e06e-474f-8a4b-8afb6fcc7e87",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Marker lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fb538a3-17f4-49ea-93ee-9c5d390077e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example marker list\n",
    "myeloid_markers = ['CD11b','CD11c','CD14','CD163','CD16','CD206','CD45','CD66b','CD68','Cd74','HLADR','P2RY12','TMEM119','VISTA','iba1','HIF1a','Ki67']\n",
    "\n",
    "marker_dict = {'Lymphocyte': ['CD3','Cd8a','GranzymeB'],\n",
    "                'Myeloid': ['iba1','P2RY12','TMEM119','CD68','CD11c','CD14','CD16','CD163','CD206','CD66b','VISTA','Cd74','HLADR'],\n",
    "                'Pan-immune': ['CD11b','CD45'],\n",
    "                'Neoplastic': ['MHCI','CD44','Cd109','Met','S100B','Ki67','Nestin','Olig2','Sox2','pERK','GLUT1','GFAP','HIF1a','Vimentin'],\n",
    "                'Vascular': ['CD31','Sma','Fibrinogen']}\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0129488e-172a-4ee5-b96e-75bcd8457d8b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Example heatmap using simple marker list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e3b27dd-ef59-49d3-ac62-5a376403b259",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata,\n",
    "                 myeloid_markers, #Specify the markers\n",
    "                 groupby='population', \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True,\n",
    "                #save='heatmap_nonmyeloid_population.svg'\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c4a180b-98ad-4e0f-bb52-857200506554",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Example heatmap using simple marker list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b465f062-f956-4476-b9cc-dbabf790d13e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata,\n",
    "                 marker_dict, #Specify the markers\n",
    "                 groupby='population', \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac5b24f-726d-4bc6-bf3e-59ce74cec7a7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='red'>ATHENA graphs</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02cd1aa-85c7-451a-9f1d-a712b7aa3a4a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Save spatial mapping of pops using cell masks (without displaying)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da406c1a-f1ac-43e7-a283-76684ce009e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for obs in ['hierarchy_id', 'population_id', 'population_broad_id']:\n",
    "    for bgd in ['black','white']: \n",
    "        for i in so.spl.index:    \n",
    "            sh.pl.spatial(so, i, obs, mode='mask', background_color=bgd, save=f'{bgd}_{obs}__{i}.svg', show=False) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d44077d-fc29-404c-a6df-71ceb3f1aa0f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Interaction graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98260d2c-97de-4509-bce2-c58d196fc13a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Interaction heatmap example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d100d75a-ee50-4f76-8639-5004e9525fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.interactions_summary(so,\n",
    "                                all_samples,\n",
    "                                'population_broad_id_proportion_diff_contact',\n",
    "                                population_dictionary=so.uns['cmap_labels']['population_broad_id'],\n",
    "                                calc_ttest_p_value=0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21b6b477-4fec-4565-b611-740478668d6f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "####  Interaction bar graphs - Cell type broad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6078a92b-6446-46ad-96f9-23ba4cf38668",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = list(so.spl.index)\n",
    "\n",
    "\n",
    "myeloid_pops=['Microglia, resting',\n",
    "            'Microglia',\n",
    "            'Microglia, Cd16+VISTA+',\n",
    "            'TAM/Micro intermediate',\n",
    "            'Myeloid, Cd68+',\n",
    "            'TAMs, Cd74+VISTA+',\n",
    "            'TAMs, Cd206+']\n",
    "\n",
    "remap_dict={'Microglia, resting':'Myeloid',\n",
    "            'Microglia':'Myeloid',\n",
    "            'Microglia, Cd16+VISTA+':'Myeloid',\n",
    "            'TAM/Micro intermediate':'Myeloid',\n",
    "            'Myeloid, Cd68+':'Myeloid',\n",
    "            'TAMs, Cd74+VISTA+':'Myeloid',\n",
    "            'TAMs, Cd206+':'Myeloid',\n",
    "           'Cd8 lymphocytes':'Lymphocytes'}                    \n",
    "\n",
    "titles = ['Contact, Broad', \n",
    "          'Knn, Broad', \n",
    "          'Radius, Broad']\n",
    "\n",
    "graphs = ['population_broad_id_proportion_diff_contact',\n",
    "          'population_broad_id_proportion_diff_knn_6',          \n",
    "          'population_broad_id_proportion_diff_radius_20']\n",
    "\n",
    "\n",
    "for b,k in zip(titles, graphs):    \n",
    "    \n",
    "    for t in ['diff','score']:    \n",
    "        variable=t\n",
    "\n",
    "        summary = mikeimc_v2.interactions_table(so,\n",
    "                            samples_list=samples,\n",
    "                            interaction_reference=k,\n",
    "                            var=variable,\n",
    "                            population_dictionary=so.uns['cmap_labels']['population_broad_id'],\n",
    "                            mode='individual',\n",
    "                            remap=remap_dict,\n",
    "                            remap_agg='sum')\n",
    "\n",
    "\n",
    "\n",
    "        fig, axs = plt.subplots(1, 5, figsize=(8, 3), dpi=300)\n",
    "        fig.suptitle(b, fontsize=16)\n",
    "        fig.tight_layout()\n",
    "        plt.subplots_adjust(wspace = 1)\n",
    "        \n",
    "        for count,i in enumerate(['Vascular','Lymphocytes','Myeloid','Tumour','Stroma']):\n",
    "\n",
    "            data = summary.loc[summary.target_label==i,:]\n",
    "            data = data.loc[np.where(data.source_label.isin(myeloid_pops),True,False),:]\n",
    "            try:\n",
    "                data.source_label.cat.remove_unused_categories(inplace=True)\n",
    "                data.target_label.cat.remove_unused_categories(inplace=True)\n",
    "\n",
    "            except:\n",
    "                'None'\n",
    "\n",
    "\n",
    "            sb.barplot(data = data, \n",
    "                       x = \"source_label\", \n",
    "                       y = variable, \n",
    "                       ci=68,\n",
    "                       ax=axs[count],\n",
    "                       palette=population_broad_colours,\n",
    "                      order=data.groupby(['source_label','target_label']).mean().sort_values(variable).reset_index()['source_label']\n",
    "                      )\n",
    "            #axs[count].set_yscale(\"log\")\n",
    "            axs[count].set_title(i)\n",
    "            axs[count].tick_params(axis='x', labelrotation = 90, labelsize=8)\n",
    "            axs[count].tick_params(axis='y', labelsize=8)\n",
    "            plt.xticks(rotation=90)\n",
    "            axs[count].set_ylabel(t, fontsize=12)\n",
    "        \n",
    "        fig.savefig(('figures/interaction_bargraphs/'+t+'_'+k+'.svg'), bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42362f7a-ee59-4314-8b69-efb51ee80d36",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### UMAP clustering of interactions - \"cell_type_broad_id_proportion_diff_contact\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5aeb62-7273-4dee-8a83-f45ff6fbc8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = list(so.spl.index)\n",
    "#remove_samples=['3A3']\n",
    "#samples = [x for x in samples if x not in remove_samples]\n",
    "\n",
    "\n",
    "mikeimc_v2.interactions_summary_UMAP(so,samples,'population_broad_id_proportion_diff_contact',var='score',category_columns=['ROI','HEClass_2class','HEClass_3class','HEClass_niches'],annotate='ROI')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91749a66-19a9-41c3-832c-aaa92b379ba8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell-cell entropy graphs - e.g. Quadratic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a9d1b5e-75a5-412c-81ec-f63a064ba3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df = pd.DataFrame\n",
    "\n",
    "for count,samp in enumerate(so.spl.index):\n",
    "    \n",
    "    working_obs = so.obs[samp][so.obs[samp]['hierarchy']=='Myeloid']\n",
    "    working_obs.hierarchy.cat.remove_unused_categories(inplace=True)\n",
    "    working_obs.population.cat.remove_unused_categories(inplace=True)\n",
    "    \n",
    "    results = working_obs.groupby('population',observed=True).mean().reset_index()\n",
    "    results\n",
    "    results['ROI']=samp\n",
    "    results['HEClass']=so.obs[samp]['HEClass_2class'][1]\n",
    "    \n",
    "    if count==0:\n",
    "        results_df = pd.concat([results])\n",
    "    else:\n",
    "        results_df = pd.concat([results_df,results])\n",
    "        \n",
    "final_df = results_df.groupby(['population','HEClass','ROI'],observed=True).mean().reset_index()\n",
    "final_df.population = final_df.population.astype('category')\n",
    "\n",
    "for i in final_df.columns[12:]:\n",
    "    plt.figure(figsize = (2,3))\n",
    "    ax = sb.barplot(data = final_df, y = i, x = 'population', \n",
    "                    palette = population_broad_colours,\n",
    "                    order=myeloid_order,\n",
    "                    ci=68)\n",
    "   \n",
    "\n",
    "    ax.set_xticklabels(ax.get_xticklabels(),rotation = 90, fontsize = 10)\n",
    "    #ax.legend(bbox_to_anchor=(1.01, 1))\n",
    "    fig = ax.get_figure()\n",
    "    fig.savefig('figures/cell_metrics/'+i+'.eps', bbox_inches='tight',dpi=400)\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d947e86-38ca-4ce4-9793-081698bb6cdd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='red'>Ripley scores</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00a1d7f6-a9ce-4ff1-939b-4e0182d96cd9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Ripleys on myeloid cells - averaged over ROIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6c152fe-53b4-4bb7-9f6b-b6d9d7914601",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cluster_key = 'population_broad'\n",
    "\n",
    "ripley_results_df = pd.DataFrame(columns=['bins',cluster_key,'stats','ROI'])\n",
    "sims_results_df = pd.DataFrame(columns=['bins','simulations','stats','ROI'])\n",
    "\n",
    "for roi in adata.obs.ROI.cat.categories:\n",
    "    a1 = adata[adata.obs.ROI==roi]\n",
    "    a2 = a1[a1.obs.hierarchy=='Myeloid'].copy()\n",
    "    \n",
    "    sq.gr.ripley(a2, cluster_key=cluster_key, mode='L', max_dist=40)\n",
    "    #sq.pl.ripley(a2, cluster_key=\"phen_cluster\", mode='L')\n",
    "    #plt.show()\n",
    "    \n",
    "    ripley_results = a2.uns[f'{cluster_key}_ripley_L']['L_stat']\n",
    "    ripley_results['ROI'] = roi\n",
    "    \n",
    "    sims_results = a2.uns[f'{cluster_key}_ripley_L']['sims_stat']\n",
    "    sims_results['ROI'] = roi\n",
    "    \n",
    "    ripley_results_df = pd.concat([ripley_results_df, ripley_results])\n",
    "    sims_results_df = pd.concat([sims_results_df, sims_results])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69ac093d-15fb-490e-a645-ec69c968de3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.figsize']=(4,4)\n",
    "\n",
    "rip_summary = ripley_results_df.groupby(['bins',cluster_key]).mean()\n",
    "sims_summary = sims_results_df.groupby(['bins']).mean().reset_index()\n",
    "\n",
    "ax = sb.lineplot(\n",
    "    y=\"stats\",\n",
    "    x=\"bins\",\n",
    "    ci=\"sd\",\n",
    "    hue=cluster_key,\n",
    "    palette = population_broad_colours,\n",
    "    data=rip_summary,\n",
    ")\n",
    "\n",
    "# This plots the simulated data\n",
    "sb.lineplot(y=\"stats\", x=\"bins\", ci=\"sd\", alpha=0.1, color=\"gray\", data=sims_summary)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "#ripley_results_df.to_csv('ripley.csv') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40aaae57-f12b-48a6-909b-3fb4add7a374",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='red'>ATHENA myeloid - 6 kNN for myeloid cells, with neighborhood analyses</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d7d6eab-a7de-43a4-8047-ece89ba1c3f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Set up myeloid SO object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d452efa8-0255-4b3e-b559-cfcac443692a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a copy, as we're going to be deleting cells when making the spatial graphs\n",
    "so_myeloid = so.copy()\n",
    "\n",
    "# Remove non-myeloid clels\n",
    "for spl in samples:\n",
    "    so_myeloid.X[spl] = so_myeloid.X[spl][so_myeloid.obs[spl].hierarchy=='Myeloid']\n",
    "    so_myeloid.obs[spl] = so_myeloid.obs[spl][so_myeloid.obs[spl].hierarchy=='Myeloid']\n",
    "    so_myeloid.obs[spl].population_id.cat.remove_unused_categories(inplace=True)\n",
    "    so_myeloid.obs[spl].population_broad_id.cat.remove_unused_categories(inplace=True)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c46ea2f-0bd9-4bb0-8cf6-464088df4a7e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Build the kNN graph with 6 neighbours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c00fb6-0b03-48ec-87ba-ab776d3f0b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the list of samples for which you want to build the graph\n",
    "samples = so_myeloid.spl.index # this way you compute the graphs for all samples in your spatialOmics instance\n",
    "\n",
    "# import default graph builder parameters\n",
    "from athena.graph_builder.constants import GRAPH_BUILDER_DEFAULT_PARAMS\n",
    "\n",
    "#kNN graph, without cellmasks, using my predetermined X, Y locs\n",
    "config = GRAPH_BUILDER_DEFAULT_PARAMS['knn']\n",
    "config['builder_params']['n_neighbors'] = 6 # set parameter k\n",
    "for spl in samples:\n",
    "             sh.graph.build_graph(so_myeloid, spl,\n",
    "                     key_added='knn_6_myeloid',\n",
    "                     builder_type='knn', mask_key=None, coordinate_keys=('X_loc', 'Y_loc'), config=config) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834b6b1c-f855-4380-9d9a-0874f556ad20",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Calculate interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6cf1f4a-2cbd-4a9b-b8d7-074612e6b8e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get a list of all samples\n",
    "all_samples = so_myeloid.spl.index.values\n",
    "\n",
    "# Modes to run the analyses in\n",
    "calc_modes = ['proportion']\n",
    "\n",
    "# Population identifiers\n",
    "pop_ids = ['population_id', 'population_broad_id']\n",
    "\n",
    "# Get a list of all graphs\n",
    "graph_keys = ['knn_6_myeloid']\n",
    "\n",
    "import logging\n",
    "logging.getLogger().setLevel(logging.ERROR)  # set logger to logging.INFO if you want to see more progress information\n",
    "\n",
    "# Create blank lists\n",
    "m_list=[]\n",
    "o_list=[]\n",
    "s_list=[]\n",
    "g_list=[]\n",
    "\n",
    "# Creates several parallel lists with the order of inputs we will need to call in the pool function\n",
    "for m in calc_modes:\n",
    "    for o in pop_ids:\n",
    "        for s in all_samples:\n",
    "            for g in graph_keys:\n",
    "                m_list.append(m)\n",
    "                o_list.append(o)\n",
    "                s_list.append(s)\n",
    "                g_list.append(g)\n",
    "\n",
    "func_input=list(zip(m_list,o_list,s_list,g_list))\n",
    "\n",
    "for m, o, s, g in tqdm(func_input):\n",
    "    try:\n",
    "        sh.neigh.interactions(so_myeloid, s, o, mode=m, prediction_type='diff', graph_key=g)    \n",
    "    except KeyboardInterrupt:\n",
    "        pass  \n",
    "    except BaseException as err:\n",
    "        print(f'Error caclculating sample:{s}, graph:{g}, observation:{o}, mode:{m}')\n",
    "        #print(f\"Unexpected {err=}, {type(err)=}\")\n",
    "\n",
    "logging.getLogger().setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef8ecd51-45aa-4cb8-af64-0eaf5fb1e9ca",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Myeloid cell interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c63be6ee-0bb0-47c2-a393-7395049b79f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "myeloid_order=['Myeloid, Cd68+',\n",
    "'Microglia, resting',\n",
    "'Microglia',\n",
    "'Microglia, Cd16+VISTA+',\n",
    "'TAM/Micro intermediate',\n",
    "'TAMs, Cd74+VISTA+',\n",
    "'TAMs, Cd206+']\n",
    "\n",
    "mikeimc_v2.interactions_summary(so=so_myeloid, \n",
    "                    samples_list=all_samples, \n",
    "                    interaction_reference='population_id_proportion_diff_knn_6_myeloid',\n",
    "                    var='diff', #The variable we want to return from the interactions table\n",
    "                    population_dictionary=so_myeloid.uns['cmap_labels']['population_id'],\n",
    "                    aggregate_function='mean',\n",
    "                    calc_ttest_p_value=0.05,\n",
    "                    title=False,\n",
    "                    cmap=cc.coolwarm,\n",
    "                    save='figures/myeloid_alone/myeloid_interactions.svg',\n",
    "                    reindex=myeloid_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a49a8ac7-280d-43f8-82e0-6dfe5a52ee87",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Save network graph images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "991e9d2b-ac2c-4cba-98a1-760c2d3634f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in so_myeloid.spl.index:\n",
    "    sh.pl.spatial(so_myeloid, i, 'population_broad_id', edges=True, graph_key='knn_6_myeloid', cbar=True, node_size=5, save=f'figures/myeloid_alone/networks/{i}.svg') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4295826-50fb-4d87-a4d0-f2764f53448c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Cell-mask myeloid populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2533cc26-8a15-41e0-995e-9ebd681b49cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_c = so.copy()\n",
    "\n",
    "# Get the existing colour map, add on a new colour at the end (grey), which we will use for a new population (12) which will be for non-myeloid\n",
    "colourlist= so_c.uns['cmaps']['population_broad_id'].colors\n",
    "colourlist.append('#E9E8EA')\n",
    "colourlist=ListedColormap(colourlist)\n",
    "so_c.uns['cmaps'].update({'population_broad_id': colourlist})\n",
    "so_c.uns['cmap_labels']['population_broad_id'].update({12:'Non-myeloid'})\n",
    "\n",
    "for bgd in ['black','white']:\n",
    "    for i in so_myeloid.spl.index:       \n",
    "        so_c.obs[i]['population_broad_id'] = np.where(so_c.obs[i].hierarchy=='Myeloid',so_c.obs[i].population_broad_id,12)\n",
    "        sh.pl.spatial(so_c, i, 'population_broad_id', mode='mask', background_color=bgd, save=f'figures/myeloid_alone/pop_maps/{bgd}_{i}.svg') \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d377ff7-c8c6-494f-814f-ce1cc2dbbe8e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### UMAPs based on myeloid interactions and abundances"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc79e01-9aab-4190-8027-58c8b65eb08f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c73726d3-0d67-4ca1-b571-9755a768884a",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_samples1 = all_samples.copy()\n",
    "all_samples1 = list(all_samples1)\n",
    "#all_samples1.remove('4A2')\n",
    "#all_samples1.remove('1A1')\n",
    "\n",
    "mikeimc_v2.interactions_summary_UMAP(so_myeloid,all_samples1,'population_id_proportion_diff_knn_6_myeloid',var='score',category_columns=['ROI','HEClass_2class','HEClass_3class','HEClass_niches'], annotate='ROI')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb8d559-6087-4b1b-bfe6-c6f7cc421e09",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f208078-24a8-4a21-ad07-b400c5837f06",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.cellabundance_UMAP(adata_m,ROI_id='ROI',population='population',colour_by='HEClass_2class',annotate=True,save='roi_myeloid_abundance_umap.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b58a5a-7b14-47c2-9937-c85cd921ee4a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Squidpy spatial graph analyses</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b95b2d-6908-4466-a118-aab684ac9ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_key = \"population_broad\"\n",
    "\n",
    "##########\n",
    "adata_copy=adata.copy()\n",
    "sq.gr.spatial_neighbors(adata_copy,library_key='ROI',coord_type='generic', n_neighs=6)\n",
    "sq.gr.centrality_scores(adata_copy, cluster_key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d81ec3d9-fc3c-4e6b-b48a-e40851720404",
   "metadata": {},
   "source": [
    "Closeness_centrality - how close are the cells to other cells? A high value would suggest this population is 'well connected' to other populations, because it is not far away (in terms of number of connections) from other populations. Closeness Centrality measures how many \"hops\" it would take to reach every other node in a network (taking the shortest path). It can be informally thought as 'average distance' to all other nodes.\n",
    "\n",
    "degree_centrality - the fraction of non-group members connected to group members."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72bf1f4e-b3c7-46c8-977a-c2bc787e49c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = adata_copy.uns[f'{cluster_key}_centrality_scores'].reset_index()\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, figsize=(8, 3), dpi=300)\n",
    "fig.suptitle(b, fontsize=16)\n",
    "fig.tight_layout()\n",
    "plt.subplots_adjust(wspace = 0.4)\n",
    "\n",
    "for count, i in enumerate(['degree_centrality','average_clustering','closeness_centrality']):\n",
    "\n",
    "    sb.barplot(data = data, \n",
    "               x = \"index\", \n",
    "               y = i, \n",
    "               ci=68,\n",
    "               ax=axs[count],\n",
    "               palette=population_broad_colours,\n",
    "              order=data.sort_values(i)['index']\n",
    "              )\n",
    "    \n",
    "    #if i=='average_clustering':\n",
    "        #axs[count].set(ylim=(0.48, 0.52))\n",
    "    \n",
    "    axs[count].set_title(i)\n",
    "    axs[count].tick_params(axis='x', labelrotation = 90)\n",
    "    axs[count].tick_params(axis='y', labelsize=10)\n",
    "    plt.xticks(rotation=90)\n",
    "    #axs[count].set_ylabel('Difference', fontsize=12)\n",
    "\n",
    "fig.savefig('network_centrality_scores.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcca26f0-5b46-46c5-8660-114a533c0060",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>PAGA/Trajectory Analysis</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4def298c-ee53-473f-918c-c71b00223694",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### PAGA on ALL cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1612cb95-9b78-467b-8849-9cc9dee3570a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Calculate PAGA\n",
    "sc.tl.paga(adata, groups='population_broad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3e9d7d-831e-45e3-85f9-fea956e969d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Plot PAGA\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "sc.pl.paga(adata, \n",
    "           node_size_scale=3, \n",
    "           edge_width_scale=1, \n",
    "           save='all_cells_paga.svg', \n",
    "           threshold=0.1,\n",
    "           ax=ax,\n",
    "           node_size_power=1,\n",
    "          fontoutline=1.25\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f5f2511-f6b4-492a-9d8a-e65eb243161f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot PAGA\n",
    "#fig, axs = plt.subplots(1, 2, figsize=(8,8))\n",
    "axs = sc.pl.paga_compare(adata,save='test.svg',**{'threshold':0.1})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1003ef1f-9c64-4b6c-9db7-5e783942e139",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Confusion matrix\n",
    "Compare distribution between two categorical labels of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8811ad94-9972-4a60-9b19-e8272b7634d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "cmtx = sc.metrics.confusion_matrix(\"population\", \"population_broad\", adata.obs)\n",
    "sb.heatmap(cmtx, square=True)\n",
    "\n",
    "fig.savefig('figures/other/confusion.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "052ace4a-4c58-4ff5-8ea5-c4f2b0e558e4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='orange'>LISA Clustering</font>\n",
    "LISA clustering\n",
    "Paper: https://www.biorxiv.org/content/10.1101/2021.08.16.456469v1\n",
    "\n",
    "Web app: https://shiny.maths.usyd.edu.au/lisaClust/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c046c39-e02e-4a53-b450-7566842f9ea9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Exporting data for LISA analysis\n",
    "Export a simplified version of the .obs dataframe with the information we need to do the LISA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab916bfd-1f49-4a66-aa4b-c1be6d1b971a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Specify the cluster that has population information in the adata.obs\n",
    "cluster_id='population_broad'\n",
    "\n",
    "adata.obs[['ROI',cluster_id,'X_loc','Y_loc']].to_csv('lisaclust_export.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bc653ee-1070-45d0-b83a-0c9c1a654a45",
   "metadata": {},
   "source": [
    "#### <font color='red'>Upload the 'lisaclust_export.csv' to the web app above, and then download the results as 'LISAclust.csv' in this folder</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c498bc-d9e8-4699-bbb9-2cc8547220a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Importing data from LISA analysis from web app\n",
    "This will add in the LISA results to the adata.obs as a new column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea2cb210-19e0-4bab-a9d2-a5ff8a36ecc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.lisa_import(adata,LISA_file = 'LISAclust.csv',LISA_col_title = 'LISAclust')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c31e9d72-42cb-4202-8b7e-60447ae68491",
   "metadata": {},
   "source": [
    "#### Stacked bar graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e199fb2-e517-448e-ba11-7c58f9bec7a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in ['LISAclust']: \n",
    "\n",
    "    for x in ['population','population_broad','hierarchy']:\n",
    "        \n",
    "        tmp = pd.crosstab(adata.obs[i],adata.obs[x], normalize='index')\n",
    "        tmp.plot.bar(stacked=True, figsize=(3, 3)).legend(bbox_to_anchor=(1.02, 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37f06ea-e75e-4bfc-91a0-b19c8e9087c5",
   "metadata": {},
   "source": [
    "#### Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dee3a2c-2a35-43df-b0b4-f40afa4ed636",
   "metadata": {},
   "outputs": [],
   "source": [
    "LISA_clust_obs='LISAclust'\n",
    "population_id='population_broad'\n",
    "\n",
    "vmax=0.30\n",
    "vmin=0\n",
    "figsize=(5,5)\n",
    "\n",
    "fig, axs = plt.subplots(figsize=figsize)\n",
    "tmp = pd.crosstab(adata.obs[population_id], adata.obs[LISA_clust_obs], normalize='index')\n",
    "sb.heatmap(data=tmp, robust=True,linewidths=.5,square=True,cmap='viridis',vmax=vmax, vmin=vmin, ax=axs)\n",
    "fig.savefig('LISAclust_heatmap.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cc4420-7904-4b6a-919e-ba0a3fca23c8",
   "metadata": {},
   "source": [
    "#### UMAP based on cluster proportion in an ROI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0f7a01-d3e9-45f1-a910-496e69a64dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.cellabundance_UMAP(adata, 'ROI',population='LISAclust', colour_by='HEClass_2class', normalize='index', dim_red='UMAP')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "041168ca-08b1-4cb9-b35c-6e9180b1ebdf",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Edge vs core of lisa clusters\n",
    "Normalising over ROI, so it becomes proportion of cells in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b787f1d-1c5b-43b5-a07a-3450966cf400",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.grouped_graph(adata,\n",
    "                         ROI_id='ROI',\n",
    "                         group_by_obs='HEClass_2class',\n",
    "                         x_axis='LISAclust',\n",
    "                         fig_size=(3,3),\n",
    "                         log_scale=False,\n",
    "                        display_tables=True,\n",
    "                         crosstab_norm='index'\n",
    "                        ) #If you change display_tables to True, will also do stats on the groups\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d980e13-3cbf-49db-9a3b-97f018ad2570",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Draw all voronoi and save them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f012a30-b6f9-4212-a3f7-1233ab6b8955",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from mikeimc_v2 import draw_voronoi_scatter\n",
    "\n",
    "for i in adata.obs['ROI'].unique().tolist():\n",
    "\n",
    "    spot = adata.obs[adata.obs['ROI']==i]\n",
    "\n",
    "    _ = draw_voronoi_scatter(spot=spot,\n",
    "                             c=[],\n",
    "                             voronoi_palette = sc.pl.palettes.vega_20_scanpy,\n",
    "                             X='X_loc',\n",
    "                             Y='Y_loc',\n",
    "                             voronoi_hue='LISAclust')\n",
    "    plt.savefig(str(i)+'_voronoi.svg')\n",
    "    plt.savefig(str(i)+'_voronoi.png')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5cfc15-dc7a-430e-8cd6-81b1e4da9ba5",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Voronoi, with a specific cell population drawn ontop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14fab5e-3bf9-475b-9cdb-84b0377a4aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define the ROI to look at\n",
    "ROI = adata.obs[adata.obs['ROI']=='1A1']\n",
    "\n",
    "#Specify which cell populations will be overlayed\n",
    "specific_cells=ROI[ROI.hierarchy=='Myeloid'].copy()\n",
    "\n",
    "#Make a new column with a number per category, will be used to colour cells that are added\n",
    "specific_cells['colour']=specific_cells['population_broad'].cat.codes\n",
    "\n",
    "_ = draw_voronoi_scatter(spot=ROI,\n",
    "                         c=specific_cells,\n",
    "                         voronoi_palette = sc.pl.palettes.vega_20_scanpy,\n",
    "                         X='X_loc',\n",
    "                         Y='Y_loc',\n",
    "                         voronoi_hue='LISAclust',\n",
    "                         scatter_hue='colour',\n",
    "                         scatter_palette=sc.pl.palettes.vega_20_scanpy,\n",
    "                         scatter_kwargs={'s':10},\n",
    "                         figsize=(5,5))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
