{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5f989060-147b-4724-a728-42dc7aa8daa2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mike IMC Pipeline v2.0 - 14th Feb 2023\n",
    "#### <font color='blue'>This is a work in progress, and should be treated as such. Expect errors and issues! The start should hopefully be fairly 'one size fits all', but as it progresses it gets more and more complex, and will require an increasing knowledge of Python or more customisation to run: </font>\n",
    "\n",
    "### <font color='green'>Green</font> - Should run with only minor or no changes at all\n",
    "### <font color='orange'>Orange</font> - Requires a fair amount of customisation, and usually adding in external files (e.g. images or extra .csv files)\n",
    "\n",
    "Developed by Michael Haley (michael.haley@manchester.ac.uk), **please do not share publicly!!**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ce37d08-437e-4a1b-b4ba-82b53a441156",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <font color='blue'>Before starting...</font>\n",
    "This guide assumes you have already segmented your data, likely using the Bodenmiller pipeline, but really any pipeline which produces a 'cell table' will work nicely. A really good explanation on the overall IMC analyis strategy can be found here: https://bodenmillergroup.github.io/IMCDataAnalysis/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da97bfb7-270b-4247-b561-9e82a5568351",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(\"segmentation.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "200ae033-b55b-4ef3-9de0-2ab57162038b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <font color='blue'>Setting up Python</font>\n",
    "\n",
    "### 1. Setting up Anaconda\n",
    "If you haven't yet setup Anaconda on your system, and are viewing this on the Github, there first step is to install and setup **Anaconda** (https://www.anaconda.com/products/distribution)\n",
    "\n",
    "\n",
    "### 2. Setting up the environment\n",
    "Once Anaconda is installed, we need to open the Anaconda Powershell Prompt, and create a new environment which we will call IMC (you can call it anything) using Conda. \n",
    ">**conda create --name IMC**</br>\n",
    "\n",
    "We then want to add some channels to Conda where packages will be downloaded from\n",
    ">**conda config --add channels conda-forge bioconda**</br>\n",
    "\n",
    "\n",
    "\n",
    "We now activate our new environment, and then start installing the packages we will need. This could take a while!\n",
    "\n",
    ">**conda activate IMC**\n",
    "\n",
    ">**conda install squidpy scanpy pandas anndata seaborn yaml scikit-image yellowbrick dill umap-learn napari-imc colorcet scikit-learn jupyterlab shapely tqdm git**\n",
    "\n",
    "Some packages aren't available on Conda, and therefore need to be installed via Pip\n",
    ">**pip install phenograph**</br>\n",
    "\n",
    "Athena must also be installed via Pip...\n",
    ">**pip install ai4scr-spatial-omics ai4scr-athena**</br>\n",
    "\n",
    "<font color='red'>If you get an error in the above command, try this instead...</font>\n",
    ">**pip install ai4scr-spatial-omics ai4scr-athena --no-deps**</br>\n",
    ">**pip install astropy**</br>\n",
    "\n",
    "Install bbknn, which we use for batch correction\n",
    ">**pip install bbknn**</br>\n",
    "\n",
    "<font color='red'>If you got an error in the above mentioning 'Microsoft Visual C++ 14.0 is required', you can download it here, though you will need admin rights:</font>\n",
    "https://visualstudio.microsoft.com/visual-cpp-build-tools/\n",
    "\n",
    "\n",
    "### 3. Starting Jupyter\n",
    "Download this notebook and the **'mikeimc_v2.py'** place them somewhere sensible (e.g. 'C:/Programming', or wherever)\n",
    "\n",
    "Open Anaconda Powershell and change directories to where you have put the downloaded files. In the above example...\n",
    ">**cd c:/programming**\n",
    "\n",
    "Activate the IMC environment. You can tell what environment you are currently in by looking at the start of the command line by default it will be *(base)*, after this is should be *(IMC)*\n",
    "\n",
    ">**conda activate IMC**\n",
    "\n",
    "Start Jupyter notebook\n",
    ">**jupyter notebook**\n",
    "\n",
    "Open up the Jupyter notebook, and contine below..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d633d09-99e8-4234-8519-88272beb7bee",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <a id=\"celltable\"><font color='blue'>Required files and setup</font></a>\n",
    "\n",
    "This pieline assumes you have just finished the final step of the Bodenmiller Pipeline, the last step of which creates a folder called **cpout** in the **analysis** folder.\n",
    "\n",
    "Copy the **cpout** folder into the same directory as this notebook.\n",
    "\n",
    "Within the **cpout** folder, there should the following files and folders\n",
    "\n",
    "- **Image.csv** - Contains metadata for each region of interest\n",
    "- **cell.csv** - The cell table with all the extracted single cell data\n",
    "- **panel.csv** - The panel file used in the Bodenmiller pipeline setup\n",
    "- **masks** - The cell masks\n",
    "- **images** - The image stacks for the channels\n",
    "\n",
    "In older versions of the pipeline, you may also have.\n",
    "\n",
    "- **acquisition_metadata.csv** - Similar to image.csv, contains metadata for each region of interest\n",
    "\n",
    "## <font color='blue'>Creating the accompanying dictionary file</font>\n",
    "Create (or modify using the templates) a **dictionary.csv** file as detailed in the below schematic.\n",
    "\n",
    "Save them in the save folder as this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a3bba03-ee5a-42cb-bf69-62e396221428",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(\"celltable.jpg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87dfcc48-96d2-43b7-a068-24b827dc5606",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <font color='blue'>Important notes on using this pipeline</font>\n",
    "When we do analyses in Python, it is important to understand *where* and *how* our data is being stored. Simply, there are 3 different places that our analyses may exist, potentially all at the same time:\n",
    "\n",
    "*Data and analyes may be stored in....*\n",
    "\n",
    "#### The Python kernel\n",
    "This is the instance of Python that you are currrently working in. It has all the variables and objects you have currently defined stored in the memory, allowing us to access and work with them. However, if we close the kernel, or it crashes, **we lose all our analyses!** Usually we can just re-run all the analyses again without an issue, but it's time consuming and annoying. If the browser crashes, that is not an issue - just reopen the browser,  go back to the Jupyter web page (look in your browser 'history' to find the address), and you should be able to pick up where you left off. However, in order to keep Jupyter and the kernel running, do not close the Anaconda prompt you used to start Jupyter, or turn off your computer\n",
    "\n",
    "#### The Jupyter notebook file (.ipynb)\n",
    "This stores all your **markdown, code and the outputs** of any analyses you have run - essentially, **if you can see it visually on screen it will be saved in the Jupyter notebook file!**. Jupyter will automaticlaly save periodically. However, if you rerun cells, you lose any previous outputs, unless you have saved them to disk...\n",
    "\n",
    "#### Exported / saved figures\n",
    "The safest way to store your analyses and export them out of Python is to store the resulting figures to disk as images, or export any tabular data as spread sheets. \n",
    "\n",
    "#### Adata object saved on disk\n",
    "You can save AnnData (*adata*) objects to disk, and reload them later. This is the a very convenient way to come back to later analyses. Unfortunately, there is no good way to save spatial-omics *so* objects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5209e003-5737-40b7-9143-6d10b8bc32fb",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "Do not close the Anaconda Powershell you used to start Jupyter, turn off your computer entirely, or manually restart the kernel <b>unless</b> you have saved your analyses somehow, or don't mind rerunning your analyses (which in some cases may take several hours!)\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ffb709a-5a9c-49eb-acb7-d35501849d98",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'> Importing required packages</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8e9b796-5cee-4c21-8062-821598d59dcf",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Read the package documentation!</b> I'm not going to give an extended explanation of how the various packages work. Instead, go and read the documentation and papers associated with the main packages I have linked here. This will explain how the different functions work, give worked examples, and explain how you can customise the code. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d2e1e80-3639-4d09-bc0a-2ebe1c0b7d6e",
   "metadata": {
    "tags": []
   },
   "source": [
    "This will import the libraries that we need, including a module I've made called **'mikeimc'** which has some useful functions for processing IMC data - *mikeimc_v2.py* should the in the same directory as this notebook. The other libraries should already be installed if you have setup the environment correctly, if not they won't import here and you will get an error.\n",
    "\n",
    "### Packages:\n",
    "\n",
    "##### Scanpy\n",
    "Single-cell analysis developed for single-cell sequencing, but also works great for IMC.\n",
    "https://scanpy.readthedocs.io/en/stable/\n",
    "\n",
    "##### Squidpy\n",
    "Built on scanpy, expands to do several different spatial analyses\n",
    "https://squidpy.readthedocs.io/en/stable/\n",
    "https://www.biorxiv.org/content/10.1101/2021.02.19.431994v1\n",
    "\n",
    "##### AnnData\n",
    "Data object used to store all our single cell analyses (in the *adata* object)\n",
    "https://anndata.readthedocs.io/en/latest/\n",
    "\n",
    "##### Pandas\n",
    "Tools for working with tables of data, which is pretty much all the data we will be working with.\n",
    "\n",
    "##### BBKNN:\n",
    "Algorithm implemented by Scanpy to do batch correction\n",
    "https://academic.oup.com/bioinformatics/article/36/3/964/5545955\n",
    "\n",
    "##### ATHENA and 'spatial omics' object\n",
    "Developed by a former member of the Bodemiller lab, has a lot more functionality to do in-depth spatial analyses. The problem is that it isn't build directly on an *adata* object, instead using an *so* (spatial omics) object. Therefore, we have to transer all the data from the *adata* to the *so* half way through.\n",
    "\n",
    "https://github.com/AI4SCR/ATHENA\n",
    "https://github.com/AI4SCR/spatial-omics\n",
    "https://histocartography.github.io/athena/index.html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9369db5-5866-43bd-94cd-332743664767",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>CaseViewer/3DHistech Issue</b> If you are having an error where Jupyter crashes and you have CaseViewer or some other 3D Histech software installed, it could be because of the following issue:\n",
    "https://github.com/lmcinnes/umap/issues/702\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c0557a9-e68f-48ba-b09f-38c5bc476646",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Standard libraries for data analysis, most of these should be automatically installed\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "\n",
    "# Pandas for working with dataframes and tabular data\n",
    "import pandas as pd\n",
    "\n",
    "# AnnData for working with anndatas (annotated datas)\n",
    "import anndata as ad\n",
    "\n",
    "# Matplotlib and seaborn for plotting\n",
    "import matplotlib\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "from matplotlib import cm\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap, Normalize\n",
    "import seaborn as sb\n",
    "\n",
    "# For access to good colour maps\n",
    "import colorcet as cc\n",
    "\n",
    "# OS and pathlib for opening/deleting files\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "# Scanpy \n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "\n",
    "# Squidpy\n",
    "import squidpy as sq\n",
    "\n",
    "#Packages to do Elbow analysis\n",
    "from sklearn.cluster import KMeans\n",
    "from yellowbrick.cluster import KElbowVisualizer\n",
    "\n",
    "# Import UMAP\n",
    "# This can cause a crash if CaseViewer is installed! See comment above.\n",
    "import umap \n",
    "\n",
    "#My module for importing IMC data into AnnData format\n",
    "import mikeimc_v2\n",
    "import mikeimc_v2 as mimc\n",
    "\n",
    "#To display multiple dataframes in same output\n",
    "from IPython.display import display\n",
    "\n",
    "#ATHENA\n",
    "from spatialOmics import SpatialOmics\n",
    "import athena as sh\n",
    "\n",
    "#Loading bar functionality for long tasks\n",
    "from tqdm import tqdm\n",
    "\n",
    "#For multiprocessing\n",
    "from functools import partial\n",
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b688dc-1c7a-48ac-8bab-f6348770feb2",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Reload mikeimc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83f79262-16f8-41f5-82e1-1bf457c73e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mikeimc_v2\n",
    "import imp\n",
    "imp.reload(mikeimc_v2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea939523-828c-4c37-9b9f-ad0f51950ff3",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Setup or reset default parameters for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a4fadac-e017-431b-9deb-baa1eefb12f9",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "    <b>Tip</b> Re-run this cell if your plots start looking weird and you're not sure why! \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2b53a9c-57cd-45c3-82d4-ee9d75655434",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reset plotting parameters to their defaults\n",
    "matplotlib.pyplot.rcParams.update(matplotlib.rcParamsDefault)\n",
    "\n",
    "# Set up output figure settings\n",
    "plt.rcParams['figure.figsize']=(5,5) #rescale figures, increase sizehere\n",
    "\n",
    "# Set up scanpy settings\n",
    "sc.settings.verbosity = 3\n",
    "sc.set_figure_params(dpi=100, dpi_save=200) #Increase DPI for better resolution figures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2858e49-345a-4aa1-8d9d-04dcb7ed7372",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'> Saving / loading AnnData objects </font>\n",
    "These don't need to be run initially, but will be useful to save and load analyses, or export data to a spreadsheet for external analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2b09b5-b038-441e-8a7f-78d0c5907231",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Be careful using save and load!</b> There will be no warning if you overwrite the wrong file, or load in an empty adata and therefore lose the one you're working on!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6790405-f4a3-4080-b66b-2a527fad5b3d",
   "metadata": {},
   "source": [
    "#### Save adata object\n",
    "This will save the adata object in the current directory as a file called 'adata'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd74f57f-2df1-46d0-b102-942dbec99348",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('adata')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "012b2876-33ec-43b7-9cff-e2dfe7aa330a",
   "metadata": {},
   "source": [
    "#### Export cell table\n",
    "This will export the currently adata.obs, which will give you a table format of all your cells, what populations and results of several key analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f751c59-5c94-4169-a8ad-a0c24dca66b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.to_csv('adata_obs.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "683e3b53-3b29-4ff0-ab53-7fabc798a291",
   "metadata": {},
   "source": [
    "#### Load adata\n",
    "This will open up a stored adata object (from the 'raw' directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e84328f-0cf6-420b-8c1d-1c6058fcb2c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad('adata')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "991f689b-66cc-4775-a9ea-b259d748c066",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>Setting up AnnData object</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8280c2c2-8b5a-4cf1-8919-e4c4687b13c8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import raw\n",
    "If you have followed the instructions above, and the right files have been found, this should run without issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23530d03-d6e9-47e1-be00-f4d259984e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = 'cpout'\n",
    "panel_file='panel.csv'\n",
    "cell_file='cell.csv'\n",
    "image_file='image.csv'\n",
    "masks_dir='masks'\n",
    "images_dir='images'\n",
    "acquisition_metadata='acquisition_metadata.csv'\n",
    "\n",
    "#######################\n",
    "\n",
    "# Check the files exist\n",
    "files = [panel_file, cell_file, image_file, panel_file, masks_dir, images_dir]\n",
    "\n",
    "for file in files:\n",
    "    file_path = Path(directory, file)\n",
    "    if not os.path.exists(file_path):\n",
    "        raise TypeError(f\"Error: {file} does not exist in {directory}\")\n",
    "    \n",
    "# Import the files\n",
    "panel_df = pd.read_csv(Path(directory,panel_file), low_memory=False)\n",
    "image_df = pd.read_csv(Path(directory,image_file), low_memory=False)\n",
    "cell_df = pd.read_csv(Path(directory,cell_file), low_memory=False)\n",
    "\n",
    "# acquisition_metadata is output by some older versions of Bodeniller pipeline        \n",
    "try:\n",
    "    metadata_df = pd.read_csv(os.path.join(directory,acquisition_metadata), low_memory=False)\n",
    "except:\n",
    "    print('No metadata file found')\n",
    "\n",
    "masks_dir = Path(directory, masks_dir)\n",
    "images_dir = Path(directory, images_dir)\n",
    "\n",
    "# Filter down the panel to only those in FullStack, and then drop columns we don't need\n",
    "panel_df = panel_df.loc[panel_df['full']==True,:].drop(columns=['full','ilastik'])\n",
    "\n",
    "# Create a column that maps the channels with the column names all they will appear in the CellProfiler cell table\n",
    "panel_df['cell_table_channel'] = [('Intensity_MeanIntensity_FullStack_c'+str(i+1)) for i, _ in enumerate(panel_df.index)]\n",
    "\n",
    "panel_df.index = np.arange(1, len(panel_df) + 1)\n",
    "\n",
    "# Extract only the intensities from the cell table\n",
    "cell_df_intensities = cell_df[[col for col in cell_df.columns if 'Intensity_MeanIntensity_FullStack' in col]].copy()\n",
    "\n",
    "# Remap the column names to the proper channel names\n",
    "try:\n",
    "    panel_df.rename(columns={'Label':'Target'}, inplace=True)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "mapping = dict(zip(panel_df['cell_table_channel'],panel_df['Target']))\n",
    "    \n",
    "cell_df_intensities.rename(columns=mapping, inplace=True)\n",
    "\n",
    "# Will first try and use the image_df, which should work with the newest version of the Bodenmiller pipeline\n",
    "try:\n",
    "    # Setup a dataframe with ROI-level information\n",
    "    sample_df=image_df.loc[:,['FileName_FullStack', 'FileName_cellmask','Metadata_acname', 'Metadata_description', 'Width_FullStack', 'Height_FullStack', 'ImageNumber']]\n",
    "    sample_df=sample_df.rename(columns={'Metadata_acname':'MCD_File', 'Metadata_description':'ROI', 'Metadata_max_x':'Size_x', 'Metadata_max_y':'Size_y'}).set_index('ROI')\n",
    "    sample_df['AcSession']=[x[:-10] for x in sample_df['FileName_FullStack']]\n",
    "    \n",
    "# Alternatively, will use acquisition_metadata.csv from older versions of the pipeline\n",
    "except:\n",
    "    print('Couldnt find info in image_df, using acquisition_metadata instead...')\n",
    "    # Setup a dataframe with ROI-level information\n",
    "    sample_df=metadata_df.loc[:,['AcSession', 'description', 'max_x', 'max_y']]\n",
    "    sample_df=sample_df.rename(columns={'description':'ROI', 'max_x':'Size_x', 'max_y':'Size_y'}).set_index('ROI')\n",
    "    sample_df['FileName_FullStack']=sample_df['AcSession']+'_full.tiff'\n",
    "    sample_df['FileName_cellmask']=sample_df['AcSession']+'_ilastik_s2_Probabilities_mask.tiff'\n",
    "    sample_df['ImageNumber']=sample_df['FileName_FullStack'].map(dict(zip(image_df['FileName_FullStack'], image_df['ImageNumber'])))\n",
    "    \n",
    "sample_df['mm2']=(sample_df['Size_x']/1000)*(sample_df['Size_y']/1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd709f25-9853-4f6e-982a-e6c46195bf9f",
   "metadata": {},
   "source": [
    "## Setup imaging folders\n",
    "\n",
    "This will create **mask** and **image** folders in the current directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25d2e981-a2d0-4422-a770-5802548781b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.stacks_to_imagefolders(input_folder=images_dir,\n",
    "                           masks_folder=masks_dir,\n",
    "                            panel_df=panel_df,\n",
    "                            sample_df=sample_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8240af5e-110c-48cc-b6fe-84265c0b16cc",
   "metadata": {},
   "source": [
    "# <font color='purple'>Normalisation method optimisation</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "925e46dc-1688-4afc-9927-b226cc5a92f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.normalisation_optimisation(cell_dfs=cell_df,\n",
    "                              sample_df=sample_df,\n",
    "                              panel_df=panel_df, \n",
    "                              image_df=image_df,\n",
    "                              clustering_resolution=0.35,\n",
    "                              marker_normalisation_list=[['nonorm'],['q99.9'],['q99'],['arcsinh5'], ['arcsinh5', 'q99.9'],['arcsinh5', 'q99']])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02c7430-833a-4564-b4ec-1371f08ef544",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Create AnnData object\n",
    "\n",
    "This will create the anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00069564-d15d-408a-b767-6dfe150b53c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_fullimport = mikeimc_v2.setup_anndata(cell_df=cell_df,\n",
    "                                          sample_df=sample_df,\n",
    "                                          panel_df=panel_df, \n",
    "                                          image_df=image_df,\n",
    "                                          cell_df_ROIcol='ImageNumber',\n",
    "                                           marker_normalisation='q99.9')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c317775-e718-4a2f-9bed-326c445d216d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Removing unused markers and ROIs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3aec822-a8d7-4d77-a8ea-64a4467bdb9e",
   "metadata": {},
   "source": [
    "Use this code to remove any markers or samples which you don't want included in your final adata that you'll be using for analyses. For example, any test regions, or markers that you know for certain haven't worked at all.\n",
    "\n",
    "<font color='blue'>This will need changing depending on what ROIs (**obs_to_remove**) or markers (**markers_to_remove**) you want to remove!</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bffe80a9-2d85-4bde-b458-d61308b2a2ab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = mikeimc_v2.filter_anndata(adata_source=adata_fullimport,\n",
    "                            markers_to_remove=['DNA1', 'DNA3'], #List of markers to remove, e.g, DNA markers, but also any \n",
    "                            obs_to_filter_on='ROI', #The .obs category which you want to filter on\n",
    "                            obs_to_remove=['Test1', 'Test2']) #The types of .obs you want removing - in this example we're removing 'Test1' and 'Test2'.\n",
    "\n",
    "# Print a list of the variables that appear in the working anndata\n",
    "print(adata.var_names)\n",
    "\n",
    "# Print the structure of the working anndata\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f67e3b0b-1d57-4903-8f1a-5eaee5b1534b",
   "metadata": {},
   "source": [
    "## Convert some columns to category\n",
    "This will save headaches in later stages - makes sure that all categorical variables in obs are stored as categories.\n",
    "\n",
    "<font color='blue'>This will need changing depending on what extra columns you added from your **'observations' in the cell table, or information added in the dictionary.csv**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fc3295e-1e9a-42cc-a2a1-14fa238bfd11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will list out the columns you may want to now convert...\n",
    "adata.obs.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07492d6-f2dc-4ad2-a26a-e7d5636a1c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "categorical_obs=['ROI','Case', 'Disease_Severity', 'Growth_Rate','Histopathology','Abnormal_Vasculature','Matrix_Deposition', 'Treatment_Status', 'Primary_recurrent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3daac3-f751-44a2-90d8-3383df23c0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alter this with the categorical obs from above\n",
    "for i in categorical_obs:\n",
    "    adata.obs[i]=adata.obs[i].astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeffab90-b5a4-426c-ae59-c90974d5cc43",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Very quick QC check of data\n",
    "Some simple QC checks - look at expression of all the markers in each ROI in a heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc50faca-1b69-4d93-9d3f-af2913895815",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The title of  the column in which defines each image/region of interest\n",
    "image_var = 'ROI'\n",
    "\n",
    "# Plot a heat of all the markers\n",
    "sc.pl.matrixplot(adata, adata.var_names, groupby=image_var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301a9e7c-9d06-4068-bcb6-51b100ba2e51",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>Image maker</font>\n",
    "This function will create RGB images using up to 7 channels, similar to MCD Viewer or ImageJ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d764e770-9878-4453-a906-be282a493f18",
   "metadata": {},
   "source": [
    "### Args:\n",
    "\n",
    "**image_folder**:\n",
    "    The folder with the source images. In this folder, each ROI should have its own subfolder, with the folder named after the ROI.\n",
    "\n",
    "**samples_list**:\n",
    "    List of samples to use from the image_folder. Only these samples will be used for auto-exposure if using quatiles (which is the default)\n",
    "\n",
    "**output_folder**:\n",
    "    Output folder for images\n",
    "\n",
    "**name_prefix**:\n",
    "    A prefix that will be put at the front of the file names\n",
    "\n",
    "**minimum**:\n",
    "    The default minimum value (see 'range' below')\n",
    "\n",
    "**max_quantile**:\n",
    "    The default max staining, which is the 97th quantile\n",
    "\n",
    "**{colour}**:\n",
    "    The name of the marker to use for that colour\n",
    "\n",
    "**{colour_range}**:\n",
    "    You can specify the range that should be the maximum and minimum, using the format (minimum, maximum). Both are specified by raw counts by default. However, if you put in a 'q' before a number, it will use that quantile range instead. e.g. 'q0.99' will use the 99th percentile as the maximum.\n",
    "\n",
    "### Returns:\n",
    "Saves a .png per roi in the specified output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32c879a8-de22-4fff-b92a-d197b7cab228",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.make_images(image_folder='images', \n",
    "                samples_list=adata.obs.ROI.unique().tolist(), #This will do all ROIs\n",
    "                output_folder='images_output',\n",
    "                name_prefix='',\n",
    "                minimum=0.2,\n",
    "                max_quantile='q0.97',\n",
    "                red='iba1',\n",
    "                red_range=None,\n",
    "                green=None,\n",
    "                green_range=None,\n",
    "                blue='DNA1', \n",
    "                blue_range=None,\n",
    "                magenta=None,\n",
    "                magenta_range=None,\n",
    "                cyan=None,\n",
    "                cyan_range=None,\n",
    "                yellow=None,\n",
    "                yellow_range=None,\n",
    "                white=None,\n",
    "                white_range=None,\n",
    "                roi_folder_save=False,\n",
    "                simple_file_names=False,\n",
    "                save_subfolder='')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927401ad-72b3-4a49-877e-e2d7dad64c3e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'> Batch integration using BBKNN </font>\n",
    "This will batch correct using BBKNN. There are other options batch correction, such as Harmony, but I found this the easiest to implement. Read more here: https://bodenmillergroup.github.io/IMCDataAnalysis/batch-effects.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8379879c-29ed-4c01-bbe1-57401a952293",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Run PCA and BBKNN\n",
    "\n",
    "<font color='blue'>**batch_correction_obs** -This defines which .obs should be used to identify the different batches, in the example here it is 'Case'</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d42b85-d284-4b72-a341-8611a585d5ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define the number of PCA dimensions to work with - one less than number of markers. Without this, it usually defaults to 50, which we don't have enough markers for.\n",
    "n_for_pca = len(adata.var_names)-1\n",
    "\n",
    "# Define the 'obs' which defines the different cases\n",
    "batch_correction_obs = 'Case'\n",
    "\n",
    "# Calculate PCA, this must be done before BBKNN\n",
    "sc.tl.pca(adata, n_comps=n_for_pca)\n",
    "\n",
    "# BBKNN - it is used in place of the scanpy 'neighbors' command that calculates nearest neighbours in the feature space\n",
    "sc.external.pp.bbknn(adata, batch_key='Case', n_pcs=n_for_pca)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "151a56eb-b8f3-4420-b356-e32f593cc371",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>UMAPs</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e4c8377-78f8-42fd-8c61-505e301c4e28",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Calculate UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1e3bd5-0127-470f-9d81-6d5bc8682882",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d9bc00-a9bf-4a45-893d-aa12fdf0fc59",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plot UMAPs\n",
    "<font color='blue'>You can add extra .obs to **UMAP_groups** to colour the UMAPs by, e.g. treatment</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339fc527-43ef-400f-a4ef-776d2cc20034",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "figure_dir=Path('Figures','UMAPs')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "#Define a list of .obs to colour the UMAP by\n",
    "UMAP_groups = categorical_obs#['ROI'] \n",
    "\n",
    "# Plot UMAPs coloured by list above\n",
    "fig = sc.pl.umap(adata, color=UMAP_groups, ncols=1, size=10, return_fig=True)\n",
    "fig.savefig(Path(figure_dir, 'Categorical_UMAPS.png'), bbox_inches='tight', dpi=300)\n",
    "\n",
    "# This will plot a UMAP for each of the individual markers\n",
    "fig = sc.pl.umap(adata, color=adata.var_names.tolist(), color_map='plasma', ncols=4, size=10, return_fig=True)\n",
    "fig.savefig(Path(figure_dir, 'Marker_UMAPS.png'), bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fea9870-3343-4d94-9b42-0e5836c2b9c4",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>Population identification</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed607813-cec3-4619-a545-7c0c8d995809",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Initial population identification using Leiden clustering\n",
    "This is the same algorithm used by PhenoGraph, and is more or less the same"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27fe8339-fa7a-41c1-a9c1-cb1b98cc3cf0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Initial Leiden clustering\n",
    "\n",
    "<font color='blue'>**resolution** - Change this to alter the size of the clusters - small resolution results in bigger clusters (therefore, less clusters overall, but each with more cells). </font>\n",
    "\n",
    "Feel free to re-run this with several different resolutions, to see which looks best!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6caa8c74-68bb-480b-b799-b9390ee5fddc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Change this to alter the size of the clusters - small resolution results in bigger clusters (therefore, less clusters overall, but each with more cells)\n",
    "resolution = 0.35\n",
    "leiden_obs_name = 'leiden_'+str(resolution)\n",
    "\n",
    "# This will perform the clustering, then add an 'obs' with name specified above, e.g leiden_0.35\n",
    "sc.tl.leiden(adata, resolution=resolution, key_added = leiden_obs_name) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e476007f-641e-4ec0-b41d-131794c45901",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plot initial clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a8cb082-8b3a-4517-bb04-6174315b88df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "image_var = 'ROI' # The adata.obs that defines each image or ROI\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))\n",
    "sc.pl.umap(adata, color=leiden_obs_name,size=2)\n",
    "\n",
    "# Delete old dendrograms\n",
    "try:\n",
    "    del adata.uns[f'dendrogram_{leiden_obs_name}']\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# Plot a heatmap\n",
    "sc.pl.matrixplot(adata,\n",
    "                 adata.var_names.tolist(), \n",
    "                 groupby=leiden_obs_name, \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True)\n",
    "\n",
    "# Alter this list with adata.obs variables that you want to plot against the resulting leiden population\n",
    "for i in ['Case']:\n",
    "\n",
    "    mikeimc_v2.grouped_graph(adata,\n",
    "                             ROI_id=image_var,\n",
    "                             group_by_obs=i,\n",
    "                             x_axis=leiden_obs_name,\n",
    "                             fig_size=(10,3),\n",
    "                             log_scale=True,\n",
    "                            display_tables=False) #If you change display_tables to True, will also do stats on the groups\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "669bbd2d-adb6-4761-8889-56086322ba84",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sub clustering example using the same markers as initial analysis\n",
    "Once you've done the initital subclustering, which could be with a fairly low resolution, you can then do subclustering on specific clusters from previous Leiden analyses\n",
    "\n",
    "<font color='blue'>**resolution** - As above. Change this to alter the size of the subclusters of the previous analysis </font>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db411f4-f072-4f9d-9bcf-2f0137a03983",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Change this to alter the size of the clusters in the subclustering\n",
    "resolution = 0.2\n",
    "leiden_obs_name = 'leiden_'+str(resolution)\n",
    "\n",
    "# In the following example, it will do a leiden clustering based upon the the results of 'leiden_1', selecting population 2\n",
    "sc.tl.leiden(adata, resolution=resolution, key_added = leiden_obs_name, restrict_to=('leiden_1',['2'])) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61c08237-ab94-48fb-96af-6dc194569d54",
   "metadata": {},
   "source": [
    "## <font color='purple'>Subclustering with different markers</font>\n",
    "This is optional extra that will see if you can subcluster using a different set of markers, as the subclustering options above use the same markers. The only drawback is that it creates a new AnnData object which you'll have to analyse separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad040700-9459-464d-8293-39c29c05aa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "myeloid = ['Iba1','CD14','CD16','CD44','CD11c','CD206','CX3CR1','HLA-DR', 'CD163', 'CD68']\n",
    "tcell = ['CD8a','CD45','CD45RO','CD45RA','CD4','CD3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "070aea69-0045-44ea-bf0b-6d1c3d255bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tcell = mikeimc_v2.adata_subclustering(adata,\n",
    "                        population_obs='leiden_0.35',\n",
    "                        populations=['8'],\n",
    "                        marker_list=tcell,\n",
    "                        clustering_resolutions=[0.1, 0.25, 0.4, 0.7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9018f5e-ca16-4fa5-8e8e-142dfc5ead25",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_myeloid = mikeimc_v2.adata_subclustering(adata,\n",
    "                        population_obs='leiden_0.35',\n",
    "                        populations=['4'],\n",
    "                        marker_list=myeloid,\n",
    "                        clustering_resolutions=[0.1, 0.25, 0.4, 0.7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd4d0736-f4cd-41c2-be39-75a09093c782",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.transfer_populations(adata_source=adata_tcell,\n",
    "                                 adata_source_populations_obs='leiden_0.1',\n",
    "                                 adata_target=adata,\n",
    "                                 adata_target_populations_obs='leiden_0.35',\n",
    "                                 pop_prefix='NEW')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de741b91-4d11-4c40-8eef-c5ca618d43b7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plot final clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63cc9c9b-fb0e-4632-8b18-184d4ea31fd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "image_var = 'ROI' # The adata.obs that defines each image or ROI\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))\n",
    "sc.pl.umap(adata, color=leiden_obs_name,size=2)\n",
    "\n",
    "# Delete old dendrograms\n",
    "try:\n",
    "    del adata.uns[f'dendrogram_{leiden_obs_name}']\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# Plot a heatmap\n",
    "sc.pl.matrixplot(adata,\n",
    "                 adata.var_names.tolist(), \n",
    "                 groupby=leiden_obs_name, \n",
    "                var_group_rotation = 0,\n",
    "                vmax=0.6,\n",
    "                dendrogram=True)\n",
    "\n",
    "# Alter this list with adata.obs variables that you want to plot against the resulting leiden population\n",
    "for i in ['Case']:\n",
    "\n",
    "    mikeimc_v2.grouped_graph(adata,\n",
    "                             ROI_id=image_var,\n",
    "                             group_by_obs=i,\n",
    "                             x_axis=leiden_obs_name,\n",
    "                             fig_size=(10,3),\n",
    "                             log_scale=True,\n",
    "                            display_tables=False) #If you change display_tables to True, will also do stats on the groups\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23d5802f-0633-4cea-abd6-ce608f42b751",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Label final populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "658fbc55-b1ac-4b00-94db-b065cef8bcd0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Remapping populations\n",
    "Here we are creating 'dictionaries' that will remap the population numbers from the Leiden clustering, to actual names.\n",
    "\n",
    "<font color='blue'>**Change these lists to match the Leiden population numbers, to actual populations that you identfy based upon marker expression**</font>\n",
    "\n",
    "I've usually been working with 3 different labels.  \n",
    "**populations** - This is the exact population name for that Leiden population. If you give two numbers the same name, they will be combined (e.g. 10 and 14 below).\n",
    "\n",
    "**hierarchy** - This gives a more broad hierarchy for the populations, allowing us to specicfy analyses only be done on certain cell types.\n",
    "\n",
    "**populations_broad** - This is a combination of populations and hierarchy, and what I've mostly used in subsequent analyses. For my analyses that mostly focused on myeloid cells, I decided to combine all the different tumour types into one label ('Tumour'), but still keep all the myeloid populations.\n",
    "\n",
    "<font color='blue'>**leiden_obs_name** - The name of the final leiden clustering that you want to use to assign populations</font>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee9b79c9-4faa-4ea5-8269-51d0a968db42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "leiden_obs_name = 'leiden_0.2' #This will depend on what subclustering you did\n",
    "\n",
    "populations={'0':'Pop0',\n",
    "'1':'Pop1',\n",
    "'2':'PanCyto+Tim3+',\n",
    "'3':'MHCI',\n",
    "'4':'Vascular',\n",
    "'5':'CAIX',\n",
    "'6':'CD206'}\n",
    "    \n",
    "adata.obs['population']=adata.obs[leiden_obs_name].map(populations).astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "738835b9-9641-448e-97e6-4e6e2f38059d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Plotting final populations with new names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d55ec6-f33b-4244-8ed6-cb5c02f85387",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in ['population']:\n",
    "    \n",
    "    # Delete old dendrograms\n",
    "    try:\n",
    "        del adata.uns[f'dendrogram_{i}']\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    sc.set_figure_params(figsize=(5, 5), dpi=200)\n",
    "    sc.pl.umap(adata, color=i,size=2)\n",
    "        \n",
    "    sc.pl.matrixplot(adata,\n",
    "                     adata.var_names, \n",
    "                     groupby=i, \n",
    "                    var_group_rotation = 0,\n",
    "                    vmax=0.7,\n",
    "                    dendrogram=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aade7d0f-a15d-492a-ad3d-96e4eff56d15",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>Create list of final populations</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9fa5256-3c0a-448a-9c9c-fb946e01d1ed",
   "metadata": {},
   "source": [
    "Create a list of the final populations we want to take forward for further analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1177d222-5e4e-4ebc-8a0a-4fe653b1083b",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs_list = ['population']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09687535-600a-4053-9798-a566483ca6a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'>Backgating assessment</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ec1a88-8a34-4644-920e-bfadab0b5ac6",
   "metadata": {
    "tags": []
   },
   "source": [
    "This will do a backgating assessment. This identifies a specific number of cells from each subpopulation of a population obs (ie, leiden cluster), then plots each population mask over a RGB image with the top 3 highest channels for each population. There is a fair amount of flexibility in this function.\n",
    "\n",
    "#### Args:\n",
    "**adata**: adata object\n",
    "\n",
    "**image_folder**: Folder with raw or denoised images. Each ROI has it's own folder.\n",
    "\n",
    "**pop_obs**: adata.obs that identifies populations\n",
    "\n",
    "**cells_per_group**: Cells per group to plot\n",
    "\n",
    "**radius**: Size of square over cell to visualise\n",
    "\n",
    "**roi_obs**: adata.obs identifying ROIs\n",
    "\n",
    "**minimum**: Minimum pixel value for images\n",
    "\n",
    "**max_quantile**: Quantile of signal for max intensity of images\n",
    "\n",
    "**markers_exclude**: List of markers to exclude when calculating highest expression of markers in group\n",
    "\n",
    "**only_use_markers**: Supply a list of markers to use when calculating highest expression of markers in group\n",
    "\n",
    "**number_top_markers**: Number of highest expression of markers, between 1 and 3\n",
    "\n",
    "**mode**:\n",
    "- 'full': Will calculate top markers, then immediately backgate\n",
    "- 'save_markers': Will only calculate markers then save in a file, which can be reviewed and modified\n",
    "- 'load_markers': Loads a saved marker file, then runs backgating           \n",
    "\n",
    "**specify_red**:Specify a channel which all pops will use for red\n",
    "\n",
    "**specify_green**:Specify a channel which all pops will use for green\n",
    "\n",
    "**specify_blue**:Specify a channel which all pops will use for blue\n",
    "\n",
    "#### Output:\n",
    "Will save everthing to output, with each population having it's own subfolder   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f7bfa8a-8e1c-4c7c-afb7-a02334c06495",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify th population to backgate\n",
    "population_obs='leiden_1'\n",
    "\n",
    "######\n",
    "\n",
    "mikeimc_v2.backgating_assessment(adata=adata,                          \n",
    "                      image_folder='images',\n",
    "                      pop_obs=population_obs,\n",
    "                      cells_per_group=50,\n",
    "                      radius=15, \n",
    "                      minimum=0.6,\n",
    "                      max_quantile='q0.99',\n",
    "                      mode='full',\n",
    "                      number_top_markers=2,\n",
    "                      markers_exclude=[],\n",
    "                      specify_blue='DNA1',\n",
    "                      output_folder='Backgating'\n",
    "                     )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebda1c5b-064c-42e8-a548-6b04f9408768",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='green'> ATHENA setup</font>\n",
    "This will transfer the Adata into ATHENA for more complex spatial analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd8cb594-9dc1-4484-ab3c-3ee0881e72f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eacabb40-4d6a-4a72-bd70-092eb46aa31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_obs(adata, so, ROI_obs='ROI', cell_index='Master_Index'):\n",
    "    \n",
    "    so_samples=so.spl.index()\n",
    "    adata_samples=adata.obs[ROI_obs].cat.categories.tolist()\n",
    "    \n",
    "    assert [x for x in so_samples if x not in adata_samples] == [x for x in adata_samples if x not in so_samples], 'ROIs in SO and AnnData do not match'\n",
    "    \n",
    "    so_cols = [so.obs[s].columns.tolist() for s in so_samples]\n",
    "    assert all(element == so_cols[0] for element in so_cols), 'SO obs tables do not have identical columns'\n",
    "    \n",
    "    adata_cols = adata.obs.columns.tolist()\n",
    "    \n",
    "    # Work out columns that arent in the other object\n",
    "    adata_to_so_cols = [x for x in adata_cols if x not in so_cols]\n",
    "    so_to_adata_cols = [x for x in so_cols if x not in adata_cols]\n",
    "    \n",
    "    for c in adata_to_so_cols:\n",
    "        \n",
    "        print(f'Mapping {c} from AnnData to SO....')\n",
    "        \n",
    "        mapping = dict(zip(adata.obs[cell_index], adata.obs[c]))\n",
    "        \n",
    "        for s in so_samples:\n",
    "            so.obs[s][c]=so.obs[s].index.map(mapping)\n",
    "            \n",
    "    for c in so_to_adata_cols:\n",
    "        \n",
    "        print(f'Mapping {c} from AnnData to SO....')\n",
    "        \n",
    "        mapping = dict(zip(adata.obs[cell_index], adata.obs[c]))\n",
    "        \n",
    "        for s in so_samples:\n",
    "            so.obs[s][c]=so.obs[s].index.map(mapping)\n",
    "                \n",
    "            \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c61b4ec6-3139-4ce9-ae34-3d1e3e33b8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "dictionary = pd.read_csv('dictionary.csv')\n",
    "\n",
    "so, colour_maps = mikeimc_v2.athena_from_adata(adata,\n",
    "                              sample_df=sample_df,\n",
    "                                sample_level_obs=dictionary.columns.tolist(),\n",
    "                             population_obs_list=population_obs_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ab80621-6ad2-4bb0-9670-c3509602ccd3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <font color='red'>Colour picker</font>\n",
    "This code occassionally causes the Kernel to crash, so probably avoid using for now"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50f775a1-ec14-464a-bf3e-f1c78d87ba5b",
   "metadata": {},
   "source": [
    "#mikeimc_v2.recolour_population(so=so, \n",
    "                               population_obs='population', \n",
    "                               colour_maps=colour_maps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb23029-6a08-4075-8d6f-cd20eda6afe5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> ATHENA calculations</font>\n",
    "Most of this section you should just be able to run 'as is', but it can take a **long time!** Best to leave running over night!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52858523-46bf-4b59-9a5e-35deda09fb16",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Graph construction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95afd06-605b-45d4-a617-a4d35f878287",
   "metadata": {},
   "source": [
    "This will construct the different spatial connectivity graphs, these define which cells are considered neighbours for later analyses.\n",
    "\n",
    "use your extracted cell locations to build a {radius, knn} graph (it does not use the cellmasks)\n",
    "\n",
    "you need to define the 'r' parameter to a specific ROI, otherwise the code runs only for the last ROI in the SO object. Or the original code does not work."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bef1d0f-524a-4ab5-a0df-260dbbdbf797",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82c0d1e8-c769-46b6-a2a2-d36f4c12a8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This select which samples to use - all by default\n",
    "samples = so.spl.index \n",
    "\n",
    "# Import the default graph settings\n",
    "from athena.graph_builder.constants import GRAPH_BUILDER_DEFAULT_PARAMS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3421d537-877c-4088-b736-09eceb3124f1",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Extract centroids from masks\n",
    "We can define where cells are in two ways - either from the centre of the mask or from the cell locations extracted from the cell profiler pipeline. In theory, they should be almost identical, and in previous tests they are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0435d738-81f9-49c4-8c97-7e4aa69c9e1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Extract cell centroids across all samples\n",
    "for r in tqdm(so.obs.keys()):\n",
    "    sh.pp.extract_centroids(so, r, mask_key='cellmasks')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d0a884-cdcf-4020-b9e9-79908f598eb0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### KNN graph\n",
    "<font color='blue'>k nearest neighbors - change **'number_neighbours'** to define number of nearest neighbors to identify per cell. Default is 6 </font>\n",
    "\n",
    "\n",
    "The graph key will be knn_[number_neighbours]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a528a6-0ddb-4f25-be39-0ad01d3dcf30",
   "metadata": {},
   "outputs": [],
   "source": [
    "number_neighbours = 6\n",
    "\n",
    "#kNN graph, without cellmasks, using predetermined X, Y locs\n",
    "config = GRAPH_BUILDER_DEFAULT_PARAMS['knn']\n",
    "config['builder_params']['n_neighbors'] = number_neighbours # set parameter k\n",
    "for spl in tqdm(samples):\n",
    "             sh.graph.build_graph(so, spl,\n",
    "                     key_added='knn_'+str(number_neighbours),\n",
    "                     builder_type='knn', mask_key=None, coordinate_keys=('X_loc', 'Y_loc'), config=config) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac9bc5f-921f-4b5d-baee-4d4f75ad94b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Radius graph\n",
    "\n",
    "<font color='blue'>Defines a cells neighbours using a radius (in um) from the cell. Default is 20</font>\n",
    "\n",
    "The graph key will be radius_[radius]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c75395-9df5-4770-86da-657075bc72f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "radius = 20\n",
    "\n",
    "# radius graph without cellmasks, using my predetermined X, Y locs\n",
    "config = GRAPH_BUILDER_DEFAULT_PARAMS['radius']\n",
    "config['builder_params']['radius'] = radius\n",
    "for spl in tqdm(samples):\n",
    "             sh.graph.build_graph(so, spl,\n",
    "                     key_added='radius_'+str(radius),\n",
    "                     builder_type='radius', mask_key=None, coordinate_keys=('X_loc', 'Y_loc'), config=config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d65c5c9b-bb6c-4855-b83c-ddaad4b3dac3",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Contact graph (with multiprocessing)\n",
    "Defines neighbours using direct cell-cell contact using cell mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68838a0-9558-4a01-acf6-5c896d558421",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# This select which samples to use - all by default\n",
    "samples = so.spl.index.tolist() #so.spl.index \n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Run the paralell processing\n",
    "if parallel==True:\n",
    "    \n",
    "    contact_graph_partial = partial(mikeimc_v2.contact_graph, so)\n",
    "    \n",
    "    with Pool(processes = num_cores) as pool:      \n",
    "        data = pool.map(contact_graph_partial, tqdm(samples))\n",
    "        print('Complete')\n",
    "    \n",
    "    for spl,grph in zip(samples,data):\n",
    "\n",
    "        if spl in so.G:\n",
    "            so.G[spl].update({'contact': grph})\n",
    "        else:\n",
    "            so.G[spl] = {'contact':grph}\n",
    "    \n",
    "    print('Added back into original SO!')\n",
    "    \n",
    "elif parallel==False:\n",
    "    \n",
    "    for spl in samples:\n",
    "        try:\n",
    "            sh.graph.build_graph(so, spl, builder_type='contact', mask_key='cellmasks')\n",
    "        except KeyboardInterrupt:\n",
    "            pass\n",
    "        except BaseException as err:\n",
    "            print(\"An exception occurred in calculating contact graph for \" + spl)\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e285a2d-44e1-4adf-819e-54375db6a14c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell-level scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d208e360-cdd6-4874-ab7d-3577b54619ef",
   "metadata": {},
   "source": [
    "In their original denition, all entropic scores described above do not take the spatial component into account when calculated on a whole tumor level. For this reason, ATHENA implements two avors of these scores: a global avor, in which the metrics are computed at a whole sample level using only the phenotype distribution, and a local avor, in which the scores are computed at a single-cell level, using also the graph structure. Specically, when computing local scores, ATHENA iterates over all cells, and for each cell, computes the local entropy within its neighborhood. In this way, highly diverse regions where cells from multiple diferent phenotypes coexist can be highlighted, and, instead of computing a single entropy value as for the global avor, a distribution of entropic values is returned.\n",
    "\n",
    "Cell-level scores quantify heterogeneity in a spatial manner, accounting for local effects, and return a value per single cell, saved in so.obs. To apply these scores to the data we use again .metrics but this time with local=True. Since these scores heavily depend on the tumor topology, the graph type and occasionally additional parameters also need to be provided."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34d150c1-da7a-40d1-91b4-371d0c466189",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Calculate richness, shannon and quadratic entropy (with multiprocessing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "510ea38e-1188-495c-bac7-0c067988d54b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "# Population identifiers\n",
    "obs = [f'{x}_id' for x in population_obs_list]\n",
    "\n",
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# Get a list of all samples\n",
    "all_samples = so.spl.index.values\n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Get a list of existing .obs\n",
    "exising_obs = so.obs[samples[0]].columns.tolist()\n",
    "\n",
    "# Get a list of graphs\n",
    "graph_keys = list(so.G[all_samples[0]].keys())\n",
    "#graph_keys = ['contact']\n",
    "\n",
    "\n",
    "# Setting up lists for either parallel or normal processing\n",
    "s_list=[]\n",
    "o_list=[]\n",
    "g_list=[]\n",
    "\n",
    "for g in graph_keys:\n",
    "    for s in all_samples:\n",
    "        for o in obs:    \n",
    "            o_list.append(o)\n",
    "            s_list.append(s)\n",
    "            g_list.append(g)\n",
    "\n",
    "pool_func_input=list(zip(s_list,o_list,g_list))\n",
    "\n",
    "# To keep the same 'so' object passed every time\n",
    "cell_metrics_partial = partial(mikeimc_v2.cell_metrics, so)\n",
    "\n",
    "# Parallel processing option\n",
    "if parallel==True:\n",
    "       \n",
    "    with Pool(processes = num_cores) as pool:\n",
    "        data = pool.starmap(cell_metrics_partial, tqdm(pool_func_input, total=len(pool_func_input)))\n",
    "        print('Complete')\n",
    "\n",
    "    data_iter = iter(data)\n",
    "    for s, _, _ in pool_func_input:        \n",
    "        so.obs[s] = pd.concat((so.obs[s], next(data_iter)), axis=1)\n",
    "\n",
    "    print('Added back into original SO!')\n",
    "        \n",
    "        \n",
    "# Non-parallel processing option        \n",
    "elif parallel==False:\n",
    "\n",
    "    for s, o, g in tqdm(func_input):\n",
    "\n",
    "        try:\n",
    "            sh.metrics.richness(so, s, o, local=True, graph_key=g)\n",
    "            sh.metrics.shannon(so, s, o, local=True, graph_key=g)\n",
    "            sh.metrics.quadratic_entropy(so, s, o, local=True, graph_key=g, metric='cosine')              \n",
    "        except KeyboardInterrupt:\n",
    "            pass  \n",
    "        except BaseException as err:\n",
    "            print(f'Error caclculating sample:{s}, graph:{g}, observation:{o}')\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")  \n",
    "\n",
    "# This will delete duplicate columns in the .obs\n",
    "#for i in all_samples:\n",
    "#    so.obs[i]=so.obs[i].loc[:,~so.obs[i].columns.duplicated()].copy()\n",
    "\n",
    "new_obs = so.obs[samples[0]].columns.tolist()\n",
    "\n",
    "# Create a list of cellmetrics obs to graph later\n",
    "cellmetrics_obs = [m for m in new_obs if m not in exising_obs] "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af24d55-9912-4539-8063-e0eef991b467",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell type interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0fb883b-8049-4f93-8650-678345a2780e",
   "metadata": {},
   "source": [
    "mode: it can be classical, histocat or proportion\n",
    "\n",
    "Classic: In the classic avor of the neighborhood analysis score, the average number of interactions between cells of the same phenotype is computed.\n",
    "\n",
    "HistoCAT: In the histoCAT avor, the global average of interaction score is only computed across cells that actually show this interaction.\n",
    "\n",
    "Proportional: Finally, the proportional avor uses interaction frequencies instead of counts, i.e., divides the counts of pairwise interactions by all interactions a given cell has. We propose this approach as an alternative that normalizes the scores with respect to varying cell density. In contrast to the classic or the histoCAT avor, the proportional avor is not inuenced by the number of cells in a sample and the score is bounded in the range of [0,1].\n",
    "\n",
    "graph_key can be knn, radius or contact\n",
    "\n",
    "prediction_type can be observation, pvalue or diff\n",
    "In addition to a p-value, we propose to compute the dierence between the observed proportion of interactions obs ij between cell type i ! j and the randomised proportion of interactions rand ij . This dierence can be asymmetrical (i ! j = 6 j ! i), is bounded in the range of [-1,1], and might be more suited as input for certain machine learning models."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "682a2e63-fa98-49d4-98df-e63839cd1bb5",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Calculate all potential interaction metrics (with multiprocessing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f29d9c7-daa4-48b9-b1bb-80bad99c3294",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use parallel processing?\n",
    "parallel = True\n",
    "\n",
    "# Number of cores to use for parallel processing\n",
    "num_cores = 4\n",
    "\n",
    "# Get a list of all samples\n",
    "all_samples = so.spl.index.values\n",
    "\n",
    "# Modes to run the analyses in\n",
    "calc_modes = ['proportion']#['proportion', 'histoCAT', 'classic']\n",
    "\n",
    "# Population identifiers\n",
    "pop_ids = [f'{x}_id' for x in population_obs_list]\n",
    "\n",
    "# Get a list of all graphs\n",
    "graph_keys = list(so.G[all_samples[0]].keys())\n",
    "\n",
    "\n",
    "import logging\n",
    "logging.getLogger().setLevel(logging.ERROR)  # set logger to logging.INFO if you want to see more progress information\n",
    "\n",
    "# Create blank lists\n",
    "m_list=[]\n",
    "o_list=[]\n",
    "s_list=[]\n",
    "g_list=[]\n",
    "\n",
    "# Creates several parallel lists with the order of inputs we will need to call in the pool function\n",
    "for m in calc_modes:\n",
    "    for o in pop_ids:\n",
    "        for s in all_samples:\n",
    "            for g in graph_keys:\n",
    "                m_list.append(m)\n",
    "                o_list.append(o)\n",
    "                s_list.append(s)\n",
    "                g_list.append(g)\n",
    "\n",
    "neigh_int_partial = partial(mikeimc_v2.neigh_int, so)\n",
    "\n",
    "# Create a 'list of lists' for the pool.starmap\n",
    "pool_func_input=list(zip(m_list,o_list,s_list,g_list))\n",
    "\n",
    "\n",
    "if parallel==True:\n",
    "          \n",
    "    with Pool(processes = num_cores) as pool:\n",
    "        data = pool.starmap(neigh_int_partial, tqdm(pool_func_input, total=len(pool_func_input)))\n",
    "        print('Multiprocessing complete')\n",
    "                      \n",
    "    data_iter = iter(data)\n",
    "    for m, o, s, g in pool_func_input:\n",
    "        key = f'{o}_{m}_diff_{g}'\n",
    "        sh.neighborhood.estimators.add2uns(so, next(data_iter), s, 'interactions', key) \n",
    "\n",
    "    print('Added back into original SO!')\n",
    "            \n",
    "elif parallel==False:\n",
    "    \n",
    "    for m, o, s, g in tqdm(func_input):\n",
    "        try:\n",
    "            sh.neigh.interactions(so, s, o, mode=m, prediction_type='diff', graph_key=g)    \n",
    "        except KeyboardInterrupt:\n",
    "            pass  \n",
    "        except BaseException as err:\n",
    "            print(f'Error caclculating sample:{s}, graph:{g}, observation:{o}, mode:{m}')\n",
    "            print(f\"Unexpected {err=}, {type(err)=}\")\n",
    "            \n",
    "logging.getLogger().setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbefaa6-9f47-4f3f-9608-8c0e6ffe1a5f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'> Spatial mapping of cells</font>\n",
    "Cell populations can be spatially plotted using a couple of different ways."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b87c7ce-ce29-44a7-8d5a-e1381c2b28f5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cells as dots (& specify which subpopulations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "103aa2a6-1932-47ce-8214-82dc951db3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify list of population_obs to plot - by default will do all\n",
    "plotting_population_obs_list = population_obs_list\n",
    "\n",
    "# Specify the subfolder in 'Mapped_populations' where images should be saved\n",
    "subfolder='Plotting'\n",
    "\n",
    "# Specify specific populations within population_obs to plot - if blank will plot all\n",
    "pops_to_plot=[]\n",
    "\n",
    "#######################################\n",
    "\n",
    "for pop_obs in plotting_population_obs_list:\n",
    "    for i in so.spl.index:\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10,10))\n",
    "\n",
    "        if pops_to_plot != []:\n",
    "            data = adata.obs[(adata.obs.ROI==i)&(adata.obs[pop_obs].isin(pops_to_plot))]\n",
    "        else:\n",
    "            data = adata.obs[adata.obs.ROI==i]\n",
    "\n",
    "        sb.scatterplot(data = data,\n",
    "                           x = 'X_loc',\n",
    "                           y='Y_loc',\n",
    "                           hue=pop_obs,\n",
    "                           ax=ax,\n",
    "                         palette = colour_maps[pop_obs],\n",
    "                        edgecolor=None, #or 'black' or 'white'\n",
    "                        s=100)\n",
    "\n",
    "        ax.invert_yaxis()\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "        ax.set_xlim(0,adata.uns['sample'].loc[i,'Size_x'])\n",
    "        ax.set_ylim(0,adata.uns['sample'].loc[i,'Size_y'])\n",
    "\n",
    "        ax.grid(False)\n",
    "        #ax.get_legend().remove()\n",
    "        ax.legend(bbox_to_anchor=(1.01, 1))                         \n",
    "\n",
    "        figure_dir = Path('Figures','Mapped_populations',subfolder)\n",
    "        os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "        fig.savefig(Path(figure_dir, f'{i}_{pop_obs}_{subfolder}.png'), bbox_inches='tight', dpi=300)\n",
    "\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bce3d12-b0d3-4a8c-b396-abeabc21a3ab",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cells as dots, all ROIs plotted in a grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436b54e8-0d97-4360-bfe5-cca20f3c010e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify list of population_obs to plot\n",
    "plotting_population_obs_list = population_obs_list\n",
    "\n",
    "# Specify the subfolder in 'Mapped_populations' where images should be saved\n",
    "subfolder='Grid'\n",
    "\n",
    "####################################\n",
    "\n",
    "for pop in plotting_population_obs_list:\n",
    "\n",
    "    sb.set_style(\"whitegrid\", {'axes.grid' : False})\n",
    "    \n",
    "    ax = sb.lmplot(data = adata.obs,\n",
    "                      x = 'X_loc',\n",
    "                      y='Y_loc',\n",
    "                      hue=pop,\n",
    "                      height = 4,\n",
    "                      palette=colour_maps[pop],\n",
    "                      col = 'ROI',\n",
    "                      col_wrap = 3,\n",
    "                      fit_reg = False,\n",
    "                      aspect=1,\n",
    "                      scatter_kws={\"s\": 5})\n",
    "\n",
    "    figure_dir = Path('Figures','Mapped_populations',subfolder)\n",
    "    os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "    ax.savefig(Path(figure_dir, f\"{pop}_gridplots.png\"))    \n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02cd1aa-85c7-451a-9f1d-a712b7aa3a4a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cell masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da406c1a-f1ac-43e7-a283-76684ce009e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Specify list of population_obs to plot\n",
    "plotting_population_obs_list = population_obs_list\n",
    "\n",
    "####################################\n",
    "\n",
    "figure_dir=Path('Figures','Mapped_populations','ATHENA_maps')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for obs in [f'{x}_id' for x in plotting_population_obs_list]:\n",
    "    for bgd in ['black','white']: \n",
    "        for i in so.spl.index:    \n",
    "            fig, ax = plt.subplots(figsize=(15,15), dpi=300)\n",
    "            sh.pl.spatial(so, i, obs, mode='mask', background_color=bgd, \n",
    "                          #save=Path(figure_dir, f'{bgd}_{obs}__{i}.svg'), \n",
    "                          show=False, ax=ax)\n",
    "            fig.savefig(Path(figure_dir, f'{bgd}_{obs}__{i}.svg'), bbox_inches='tight')\n",
    "            plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78f6b2f0-1945-4c36-a4fb-cd90e5423d4c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Cells mask, plotting marker expression over masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99be5761-3536-49a8-930a-ebac4f158228",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify list of population_obs to plot\n",
    "for marker in adata.var_names:\n",
    "\n",
    "    figure_dir=Path('Figures','Mapped_populations','Markers_Overlayed', marker)\n",
    "    os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "    for i in so.spl.index:    \n",
    "        fig, ax = plt.subplots(figsize=(15,15), dpi=300)\n",
    "        sh.pl.spatial(so, i, marker, \n",
    "                      mode='mask', \n",
    "                      background_color='white', \n",
    "                      show=False, ax=ax)\n",
    "        fig.savefig(Path(figure_dir, f'{marker}_{i}.png'), bbox_inches='tight')\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02239e0e-6355-43a1-a336-0e7b72f434ed",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Cell environment analysis</font>\n",
    "\n",
    "This will take a long time to run, but will do all markers and populations by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14726e7-3162-4822-88f3-981ae20f0408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select which markers to analyse\n",
    "markers = adata.var_names #Will do all, can specify specific markers\n",
    "\n",
    "####################\n",
    "\n",
    "master_list, q_table = mikeimc_v2.environmental_analysis_texture(adata,\n",
    "                           samples_list=so.spl.index, \n",
    "                           marker_list=markers, \n",
    "                           radius=20, \n",
    "                           return_quant_table=True,\n",
    "                           quantile=0.999)\n",
    "\n",
    "figure_dir=Path('Figures','Environmental_analysis')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "master_list.to_csv(Path(figure_dir,'Environmental_analysis.csv'))\n",
    "q_table.to_csv(Path(figure_dir,'Environmental_analysis_qtable.csv'))\n",
    "\n",
    "# Normalise data to 99.9th percentile               \n",
    "master_list_normalised = master_list.apply(lambda x: x.div(x.quantile(q=.999)).clip(upper=1), axis=0)               \n",
    "\n",
    "master_list_normalised.to_csv(Path(figure_dir,'Environmental_analysis_normalised.csv'))\n",
    "\n",
    "env_obs_merged = adata.obs.set_index('Master_Index').merge(master_list, left_index=True, right_index=True)\n",
    "env_obs_merged.to_csv(Path(figure_dir,'Env_obs_merged.csv'))\n",
    "\n",
    "normenv_obs_merged = adata.obs.set_index('Master_Index').merge(master_list_normalised, left_index=True, right_index=True)\n",
    "normenv_obs_merged.to_csv(Path(figure_dir,'NormEnv_obs_merged.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f24a4a-52dc-403c-9954-e22a72fbf3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify list of population_obs to plot\n",
    "plotting_population_obs_list = population_obs_list\n",
    "\n",
    "####################\n",
    "\n",
    "for population_obs in plotting_population_obs_list:\n",
    " \n",
    "    for label, results in zip(['Normalised','NonNormalised'], [normenv_obs_merged, env_obs_merged]):\n",
    "\n",
    "        #groupby = ['ROI', population_obs]\n",
    "        #results = env_obs_merged.groupby(groupby,observed=True).mean().reset_index()\n",
    "\n",
    "        figure_dir=Path('Figures','Environmental_analysis', 'Bargraphs', population_obs, label)\n",
    "        os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "        for i in master_list.columns:\n",
    "            plt.figure(figsize = (3,3))\n",
    "\n",
    "            ax = sb.barplot(data = results, \n",
    "                            y = i, \n",
    "                            x = population_obs, \n",
    "                            palette = colour_maps[population_obs],\n",
    "                            ci=68)\n",
    "\n",
    "            ax.set_xticklabels(ax.get_xticklabels(),rotation = 90, fontsize = 10)\n",
    "            #ax.legend(bbox_to_anchor=(1.01, 1))\n",
    "\n",
    "            fig = ax.get_figure()\n",
    "            fig.savefig(Path(figure_dir, (f'{i}_{label}.png')), bbox_inches='tight',dpi=400)\n",
    "            plt.close()      "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b766ccf3-6f18-42df-b361-654882371130",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Population abundance graphs</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1b272d1-884f-4b9a-bd74-4edd0d14a2f8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Categorise samples using UMAP for cell abundance\n",
    "This will use the abundance of the different populations to create a UMAP, then colour it by an adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cb786e5-94a3-4048-bd8b-83767c6de4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify which sample-level obs (e.g. Case or tumour type) should be used to colour the UMAP\n",
    "colour_by_obs='Case' \n",
    "\n",
    "##################\n",
    "\n",
    "figure_dir=Path('Figures','Abundance_graphs')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for pop_obs in population_obs_list:\n",
    "\n",
    "    mikeimc_v2.cellabundance_UMAP(adata,\n",
    "                                  ROI_id='ROI',\n",
    "                                  population=pop_obs,\n",
    "                                  colour_by=colour_by_obs,\n",
    "                                  annotate=True,\n",
    "                                  save=Path(figure_dir, f'{pop_obs}_roi_abundance_umap.svg'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6565eaab-7915-4b2b-b27f-06aaf6fb2fe9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Breakdown of populations in case/ROI\n",
    "Here, break down the myeloid and tumour populations per case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eac3cb66-af45-4e97-a9af-92e11cf54bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obs to use to group into, e.g. Case, ROI, etc\n",
    "group_by_obs = 'ROI'\n",
    "\n",
    "##################\n",
    "\n",
    "for pop_obs in population_obs_list:\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(10, 5))\n",
    "\n",
    "    tmp = pd.crosstab(adata.obs[group_by_obs],adata.obs[pop_obs], normalize='index')\n",
    "    \n",
    "    tmp.plot.bar(color=colour_maps[pop_obs],\n",
    "                 ax=ax,\n",
    "                 stacked=True)#.legend(bbox_to_anchor=(0.75, -0.2))#.legend(bbox_to_anchor=(1.1, 1))\n",
    "    \n",
    "    #ax.get_legend().remove()\n",
    "    ax.legend(bbox_to_anchor=(1.01, 1))    \n",
    "\n",
    "    figure_dir=Path('Figures','Abundance_graphs','Breakdown')\n",
    "    os.makedirs(figure_dir, exist_ok=True)\n",
    "    \n",
    "    fig.savefig(Path(figure_dir, f'{pop_obs}_{group_by_obs}.png'), bbox_inches='tight')\n",
    "    \n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f2f98d4-aa89-4293-bb4e-590af78e29fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Population abundances and MLM for comparing two groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d852b3d7-f020-43bb-8d8c-d33115dd5eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs='population'\n",
    "\n",
    "# Specify which sample-level obs (e.g. Case or tumour type) should be used to group by\n",
    "grouping_obs='HEClass_2class'\n",
    "\n",
    "# Obs that identifies each case\n",
    "Case_id='Case'\n",
    "\n",
    "#######################\n",
    "\n",
    "mikeimc_v2.mlm_pops(adata_test,\n",
    "                    x_axis=population_obs,\n",
    "                    grouping_obs=grouping_obs,\n",
    "                    Case_id=Case_id,\n",
    "                    fig_size=(3,3),\n",
    "                    save=Path('Figures','Abundance_graphs',f'{population_obs}_{grouping_obs}_roiavg_abundances.png'),\n",
    "                    log_scale=True,\n",
    "                    use_mm2=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8639f3-3371-41f5-9ca4-8c9bb2a7117a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Abundance clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66bd82a6-599f-4bfa-bcf9-41fe9e6d9074",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify which sample-level obs (e.g. Case or tumour type) should be used to group by\n",
    "grouping_obs='Case'\n",
    "\n",
    "##################\n",
    "\n",
    "figure_dir=Path('Figures','Abundance_graphs')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for population_obs in population_obs_list\n",
    "\n",
    "    mikeimc_v2.abundance_clustermap(adata,\n",
    "                             row_obs=population_obs,\n",
    "                             col_obs=grouping_obs,\n",
    "                             row_colourmap=colour_maps[population_obs],\n",
    "                             col_colourmap=None,\n",
    "                             row_cluster=True,\n",
    "                             col_cluster=True,\n",
    "                             log10=False,\n",
    "                             figsize=(5,10),\n",
    "                             save=Path(figure_dir, f'AbundanceClustermap_{population_obs}_{col_obs}.png'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da0e7340-eb3e-4dce-ada3-7b5e2fa8aefe",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>UMAPs with populations</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5450eb43-cd2b-4bdf-9654-4614dc32210e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Populations over UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b67f145-91be-466b-aba9-2cc7980f06f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_dir=Path('Figures','UMAPs')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for pop_obs in population_obs_list:\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(5,5))\n",
    "    sc.pl.umap(adata, \n",
    "               color=pop_obs, \n",
    "               #size=10,\n",
    "               palette=colour_maps[pop_obs],\n",
    "               ax=ax)\n",
    "    \n",
    "    fig.savefig(Path(figure_dir, f'{pop_obs}.png'), bbox_inches='tight', dpi=300) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e886e86-103e-4b00-9f06-f0c12f568790",
   "metadata": {},
   "source": [
    "## Markers in a grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1bfc6cd-b163-440f-96cb-54958145f09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_dir=Path('Figures','UMAPs','Markers')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "fig = sc.pl.umap(adata, \n",
    "                 color=adata.var_names, \n",
    "                 #size=3, \n",
    "                 ncols=4, \n",
    "                 color_map='viridis', \n",
    "                 return_fig=True)\n",
    "\n",
    "fig.savefig(Path(figure_dir, 'umap_allmarkers.png'), bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc215c5-b426-48cf-8d6a-18e33777476b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Individual markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3697c32c-0c8c-4e55-bf88-87e64a678f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_dir=Path('Figures','UMAPs','Markers')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for marker in adata.var_names:\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(5,5))\n",
    "    sc.pl.umap(adata, \n",
    "               color=marker, \n",
    "               #size=10,\n",
    "               palette=colour_maps[pop_obs],\n",
    "               ax=ax,\n",
    "              show=False)\n",
    "    \n",
    "    fig.savefig(Path(figure_dir, f'{marker}.png'), bbox_inches='tight', dpi=300)\n",
    "    plt.close(fig)              "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6a7361e-e214-4a33-92fc-274d5bf11b4e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Heatmaps</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55f4e404-e06e-474f-8a4b-8afb6fcc7e87",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Marker lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fb538a3-17f4-49ea-93ee-9c5d390077e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example marker list\n",
    "myeloid_markers = ['CD11b','CD11c','CD14','CD163','CD16','CD206','CD45','CD66b','CD68','Cd74','HLADR','P2RY12','TMEM119','VISTA','iba1','HIF1a','Ki67']\n",
    "\n",
    "marker_dict = {'Lymphocyte': ['CD3','Cd8a','GranzymeB'],\n",
    "                'Myeloid': ['iba1','P2RY12','TMEM119','CD68','CD11c','CD14','CD16','CD163','CD206','CD66b','VISTA','Cd74','HLADR'],\n",
    "                'Pan-immune': ['CD11b','CD45'],\n",
    "                'Neoplastic': ['MHCI','CD44','Cd109','Met','S100B','Ki67','Nestin','Olig2','Sox2','pERK','GLUT1','GFAP','HIF1a','Vimentin'],\n",
    "                'Vascular': ['CD31','Sma','Fibrinogen']}\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0129488e-172a-4ee5-b96e-75bcd8457d8b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e3b27dd-ef59-49d3-ac62-5a376403b259",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_dir=Path('Figures','Heatmaps')\n",
    "os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "for population_obs in population_obs_list:\n",
    "\n",
    "    fig = sc.pl.matrixplot(adata,\n",
    "                            var_names= adata.var_names, #adata.var_names will do all markers\n",
    "                            groupby=population_obs, \n",
    "                            vmax=0.6,\n",
    "                            dendrogram=True,\n",
    "                            return_fig=True,\n",
    "                            )\n",
    "\n",
    "    fig.savefig(Path(figure_dir, f'{population_obs}_heatmap.png'), bbox_inches='tight', dpi=300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d44077d-fc29-404c-a6df-71ceb3f1aa0f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>ATHENA interactions graphs</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98260d2c-97de-4509-bce2-c58d196fc13a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Interactions heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d100d75a-ee50-4f76-8639-5004e9525fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "figure_dir=Path('Figures','Interactions_heatmaps')\n",
    "figure_dir.mkdir(exist_ok=True)\n",
    "\n",
    "graph_types = list(so.G[all_samples[0]].keys())\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "\n",
    "    for graph in graph_types:\n",
    "        for obs in [f'{x}_id' for x in population_obs_list]:\n",
    "\n",
    "            mikeimc_v2.interactions_summary(so = so,\n",
    "                                            samples_list = so.spl.index.tolist(),\n",
    "                                            interaction_reference = f'{obs}_proportion_diff_{graph}',\n",
    "                                            population_dictionary = so.uns['cmap_labels'][obs],\n",
    "                                            calc_ttest_p_value=0.05,\n",
    "                                            save=Path(figure_dir,f'{obs}_proportion_diff_{graph}.png'),\n",
    "                                            test='ttest', #wilcoxon\n",
    "                                           )\n",
    "\n",
    "            plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21b6b477-4fec-4565-b611-740478668d6f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Interaction bar graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa715864-2c6c-41bf-9219-d185cb346b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = so.spl.index\n",
    "population_obs='leiden_1'\n",
    "target_population='0'\n",
    "\n",
    "###\n",
    "\n",
    "mikeimc_v2.interactions_bargraph(so,\n",
    "                      samples=samples,\n",
    "                      population_obs=population_obs,\n",
    "                      target_population=target_population\n",
    "                      var='diff', #Using 'score' will give the uncorrected interactions\n",
    "                      colourmap=colour_maps[population_obs])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42362f7a-ee59-4314-8b69-efb51ee80d36",
   "metadata": {
    "tags": []
   },
   "source": [
    "## UMAP clustering of interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5aeb62-7273-4dee-8a83-f45ff6fbc8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = list(so.spl.index)\n",
    "remove_samples=[]\n",
    "samples = [x for x in samples if x not in remove_samples]\n",
    "\n",
    "figure_dir=Path('Figures','Interactions_UMAPs')\n",
    "figure_dir.mkdir(exist_ok=True)\n",
    "\n",
    "\n",
    "for g in so.uns[samples[0]]['interactions'].keys():\n",
    "\n",
    "    mikeimc_v2.interactions_summary_UMAP(so=so,\n",
    "                                         samples_list=samples,\n",
    "                                         interaction_reference=g,\n",
    "                                         var='diff', #or score\n",
    "                                         category_columns=categorical_obs,\n",
    "                                         annotate='ROI',\n",
    "                                        save=Path(figure_dir, (g+'.png')))\n",
    "    \n",
    "    plt.close('all')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91749a66-19a9-41c3-832c-aaa92b379ba8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>ATHENA cell-metrics graphs - e.g. Quadratic Entropy</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61dde09b-c58c-44b9-9ff3-2750ec1850f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for population_obs in population_obs_list:\n",
    "    \n",
    "    figure_dir=Path('Figures','ATHENA_cellmetrics', population_obs)\n",
    "    os.makedirs(figure_dir, exist_ok=True)\n",
    "\n",
    "    # Cell average of each population in each ROI\n",
    "    df = mikeimc_v2.so_groupby_roimean(so,\n",
    "                                       samples=so.spl.index,\n",
    "                                       population_obs=population_obs)\n",
    "    \n",
    "    # Saves the raw data used for plotting\n",
    "    df.to_csv(Path(figure_dir, f'{population_obs}_cell_metrics_ROIavgOverPop.csv'))\n",
    "    \n",
    "    # Plots and saves graphs\n",
    "    for i in cellmetrics_obs:\n",
    "        plt.figure(figsize = (2,3))\n",
    "        ax = sb.barplot(data = df, \n",
    "                        y = i, \n",
    "                        x = population_obs, \n",
    "                        palette = colour_maps[population_obs],\n",
    "                        ci=68)\n",
    "\n",
    "        ax.set_xticklabels(ax.get_xticklabels(),rotation = 90, fontsize = 10)\n",
    "        #ax.legend(bbox_to_anchor=(1.01, 1))\n",
    "\n",
    "        fig = ax.get_figure()\n",
    "        fig.savefig(Path(figure_dir, (f'{population_obs}_{i}_.png')), bbox_inches='tight',dpi=400)\n",
    "        plt.close()                    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd17175-53df-4b82-9e7a-6fd39a9f10b5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Spatial distance analyses</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b491c44-3935-4574-932c-edc2b984b013",
   "metadata": {},
   "source": [
    "## Perform calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b27b6fd4-ef4b-4b3c-9001-86e03d539cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs='population'\n",
    "\n",
    "sc_results_df, summary_df = mikeimc_v2.calculate_cell_distances(adata,\n",
    "                                                                samples=so.spl.index,\n",
    "                                                                population_obs=population_obs,\n",
    "                                                                simulation_number=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1d88dd-d996-4a78-9434-ff8c17140774",
   "metadata": {},
   "source": [
    "## Clustermap of distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3553a5d3-7c0b-4b8c-9ef6-8dffd2ed4585",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats=mikeimc_v2.distances_clustermap(summary_df,\n",
    "                     population_obs=population_obs,\n",
    "                    colour_map=colour_maps[population_obs],\n",
    "                     vmax=100,\n",
    "                     vmin=-100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86d0fe43-008b-4ed9-97fe-11b6b66d9bb8",
   "metadata": {},
   "source": [
    "## View full stats results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db2a652-5de3-48a9-923c-8248d8bb127b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.print_full(stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4581fea8-4281-4d66-bce4-863f6f4d33d4",
   "metadata": {},
   "source": [
    "## Bargraph showing distance to specific population (distance from predicted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15bda1bf-ede6-447f-89e1-e8de12f89768",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs='population'\n",
    "target_population='0'\n",
    "\n",
    "mikeimc_v2.distances_bargraph(summary_df,\n",
    "                   population_obs=population_obs,\n",
    "                   target_population=target_population,\n",
    "                   colourmap=colour_maps[population_obs],\n",
    "                   metric='Distance_diff',\n",
    "                   )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ac9315c-1d2a-4e3b-aace-7fe9c64a24aa",
   "metadata": {},
   "source": [
    "## Bargraph showing distance to specific population (single-cell data, without ROI-averaging or distance from predicted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02ff6724-0bbc-47eb-820e-48f3644143b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs='leiden_1'\n",
    "target_population='0'\n",
    "\n",
    "mikeimc_v2.distances_bargraph(sc_results_df,\n",
    "                   population_obs=population_obs,\n",
    "                   target_population=target_population,\n",
    "                   colourmap=colour_maps[population_obs],\n",
    "                   metric='Distance',\n",
    "                   )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d947e86-38ca-4ce4-9793-081698bb6cdd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Ripley scores</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00a1d7f6-a9ce-4ff1-939b-4e0182d96cd9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Ripleys calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddf9460b-e9bc-45ba-83ee-a0ffc1587d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_obs = 'population'\n",
    "##################\n",
    "\n",
    "rip, sims = mikeimc_v2.ripley_roiavg_calc(adata,\n",
    "           so.spl.index,\n",
    "           population=population_obs,\n",
    "           mode='L',          \n",
    "           roi_obs='ROI')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9010e804-d0d6-482b-9d49-0857aa245c0b",
   "metadata": {},
   "source": [
    "## Ripley plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2765791d-076e-49f6-a0e9-f5068ef0114c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rip_summary = rip.groupby(['bins',population_obs]).mean()\n",
    "sims_summary = sims.groupby(['bins']).mean().reset_index()\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "sb.lineplot(\n",
    "    y=\"stats\",\n",
    "    x=\"bins\",\n",
    "    ci=\"sd\",\n",
    "    hue=population_obs,\n",
    "    palette = colour_maps[population_obs],\n",
    "    data=rip_summary,\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "# This plots the simulated data\n",
    "sb.lineplot(data=sims_summary, y=\"stats\", x=\"bins\", ci=\"sd\", alpha=0.5, color=\"black\", ax=ax)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "figure_dir=Path('Figures','Ripley')\n",
    "figure_dir.mkdir(exist_ok=True)\n",
    "\n",
    "fig.savefig(Path(figure_dir, f'Ripley_{population_obs}.png'), bbox_inches='tight')\n",
    "\n",
    "rip_summary.to_csv(Path(figure_dir, f'Ripley_{population_obs}.csv')) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b58a5a-7b14-47c2-9937-c85cd921ee4a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>Squidpy Network analyses</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72bf1f4e-b3c7-46c8-977a-c2bc787e49c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_copy=adata.copy()\n",
    "\n",
    "for population_obs in population_obs_list:\n",
    "\n",
    "    sq.gr.spatial_neighbors(adata_copy, library_key='ROI',coord_type='generic', n_neighs=6)\n",
    "    sq.gr.centrality_scores(adata_copy, population_obs)\n",
    "\n",
    "\n",
    "    data = adata_copy.uns[f'{population_obs}_centrality_scores'].reset_index()\n",
    "\n",
    "    fig, axs = plt.subplots(1, 3, figsize=(8, 3), dpi=300)\n",
    "    fig.suptitle(population_obs, fontsize=16)\n",
    "    fig.tight_layout()\n",
    "    plt.subplots_adjust(wspace = 0.4)\n",
    "\n",
    "    for count, i in enumerate(['degree_centrality','average_clustering','closeness_centrality']):\n",
    "\n",
    "        sb.barplot(data = data, \n",
    "                   x = \"index\", \n",
    "                   y = i, \n",
    "                   ci=68,\n",
    "                   ax=axs[count],\n",
    "                   palette=colour_maps[population_obs],\n",
    "                  order=data.sort_values(i)['index']\n",
    "                  )\n",
    "\n",
    "        axs[count].set_title(i)\n",
    "        axs[count].tick_params(axis='x', labelrotation = 90)\n",
    "        axs[count].tick_params(axis='y', labelsize=10)\n",
    "        plt.xticks(rotation=90)\n",
    "\n",
    "    figure_dir=Path('Figures','Squidpy_NetworkAnalyses')\n",
    "    figure_dir.mkdir(exist_ok=True)\n",
    "\n",
    "    fig.savefig(Path(figure_dir, f'{population_obs}_network_centrality_scores.png'), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcca26f0-5b46-46c5-8660-114a533c0060",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='green'>PAGA/Trajectory Analysis</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4def298c-ee53-473f-918c-c71b00223694",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### PAGA on ALL cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1612cb95-9b78-467b-8849-9cc9dee3570a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "population_obs = 'population'\n",
    "\n",
    "###########\n",
    "\n",
    "#Calculate PAGA\n",
    "sc.tl.paga(adata, groups=population_obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3e9d7d-831e-45e3-85f9-fea956e969d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Plot PAGA\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "sc.pl.paga(adata, \n",
    "           node_size_scale=3, \n",
    "           edge_width_scale=1, \n",
    "           save=f'_{population_obs}_paga.png', \n",
    "           threshold=0.1,\n",
    "           ax=ax,\n",
    "           node_size_power=1,\n",
    "           fontoutline=1.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f5f2511-f6b4-492a-9d8a-e65eb243161f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.paga_compare(adata,save=f'{population_obs}_pagacompare.png',**{'threshold':0.1})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1003ef1f-9c64-4b6c-9db7-5e783942e139",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Confusion matrix\n",
    "Compare distribution between two categorical labels of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8811ad94-9972-4a60-9b19-e8272b7634d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_1='population'\n",
    "pop_2='population_broad'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "cmtx = sc.metrics.confusion_matrix(pop_1, pop_2, adata.obs)\n",
    "sb.heatmap(cmtx, square=True)\n",
    "\n",
    "fig.savefig(Path('Figures', f'{pop_1}_{pop_2}_confusion.png'),bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "052ace4a-4c58-4ff5-8ea5-c4f2b0e558e4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# <font color='orange'>LISA Clustering</font>\n",
    "LISA clustering\n",
    "Paper: https://www.biorxiv.org/content/10.1101/2021.08.16.456469v1\n",
    "\n",
    "Web app: https://shiny.maths.usyd.edu.au/lisaClust/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c046c39-e02e-4a53-b450-7566842f9ea9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Exporting data for LISA analysis\n",
    "Export a simplified version of the .obs dataframe with the information we need to do the LISA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab916bfd-1f49-4a66-aa4b-c1be6d1b971a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Specify the cluster that has population information in the adata.obs\n",
    "cluster_id='population_broad'\n",
    "\n",
    "adata.obs[['ROI',cluster_id,'X_loc','Y_loc']].to_csv('lisaclust_export.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bc653ee-1070-45d0-b83a-0c9c1a654a45",
   "metadata": {},
   "source": [
    "#### <font color='red'>Upload the 'lisaclust_export.csv' to the web app above, and then download the results as 'LISAclust.csv' in this folder</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c498bc-d9e8-4699-bbb9-2cc8547220a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Importing data from LISA analysis from web app\n",
    "This will add in the LISA results to the adata.obs as a new column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea2cb210-19e0-4bab-a9d2-a5ff8a36ecc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.lisa_import(adata,LISA_file = 'LISAclust.csv',LISA_col_title = 'LISAclust')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c31e9d72-42cb-4202-8b7e-60447ae68491",
   "metadata": {},
   "source": [
    "#### Stacked bar graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e199fb2-e517-448e-ba11-7c58f9bec7a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in ['LISAclust']: \n",
    "\n",
    "    for x in ['population','population_broad','hierarchy']:\n",
    "        \n",
    "        tmp = pd.crosstab(adata.obs[i],adata.obs[x], normalize='index')\n",
    "        tmp.plot.bar(stacked=True, figsize=(3, 3)).legend(bbox_to_anchor=(1.02, 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37f06ea-e75e-4bfc-91a0-b19c8e9087c5",
   "metadata": {},
   "source": [
    "#### Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dee3a2c-2a35-43df-b0b4-f40afa4ed636",
   "metadata": {},
   "outputs": [],
   "source": [
    "LISA_clust_obs='LISAclust'\n",
    "population_id='population_broad'\n",
    "\n",
    "vmax=0.30\n",
    "vmin=0\n",
    "figsize=(5,5)\n",
    "\n",
    "fig, axs = plt.subplots(figsize=figsize)\n",
    "tmp = pd.crosstab(adata.obs[population_id], adata.obs[LISA_clust_obs], normalize='index')\n",
    "sb.heatmap(data=tmp, robust=True,linewidths=.5,square=True,cmap='viridis',vmax=vmax, vmin=vmin, ax=axs)\n",
    "fig.savefig('LISAclust_heatmap.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cc4420-7904-4b6a-919e-ba0a3fca23c8",
   "metadata": {},
   "source": [
    "#### UMAP based on cluster proportion in an ROI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0f7a01-d3e9-45f1-a910-496e69a64dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.cellabundance_UMAP(adata, 'ROI',population='LISAclust', colour_by='HEClass_2class', normalize='index', dim_red='UMAP')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "041168ca-08b1-4cb9-b35c-6e9180b1ebdf",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Edge vs core of lisa clusters\n",
    "Normalising over ROI, so it becomes proportion of cells in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b787f1d-1c5b-43b5-a07a-3450966cf400",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikeimc_v2.grouped_graph(adata,\n",
    "                         ROI_id='ROI',\n",
    "                         group_by_obs='HEClass_2class',\n",
    "                         x_axis='LISAclust',\n",
    "                         fig_size=(3,3),\n",
    "                         log_scale=False,\n",
    "                        display_tables=True,\n",
    "                         crosstab_norm='index'\n",
    "                        ) #If you change display_tables to True, will also do stats on the groups\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d980e13-3cbf-49db-9a3b-97f018ad2570",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Draw all voronoi and save them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f012a30-b6f9-4212-a3f7-1233ab6b8955",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from mikeimc_v2 import draw_voronoi_scatter\n",
    "\n",
    "for i in adata.obs['ROI'].unique().tolist():\n",
    "\n",
    "    spot = adata.obs[adata.obs['ROI']==i]\n",
    "\n",
    "    _ = draw_voronoi_scatter(spot=spot,\n",
    "                             c=[],\n",
    "                             voronoi_palette = sc.pl.palettes.vega_20_scanpy,\n",
    "                             X='X_loc',\n",
    "                             Y='Y_loc',\n",
    "                             voronoi_hue='LISAclust')\n",
    "    plt.savefig(str(i)+'_voronoi.svg')\n",
    "    plt.savefig(str(i)+'_voronoi.png')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5cfc15-dc7a-430e-8cd6-81b1e4da9ba5",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Voronoi, with a specific cell population drawn ontop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14fab5e-3bf9-475b-9cdb-84b0377a4aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define the ROI to look at\n",
    "ROI = adata.obs[adata.obs['ROI']=='1A1']\n",
    "\n",
    "#Specify which cell populations will be overlayed\n",
    "specific_cells=ROI[ROI.hierarchy=='Myeloid'].copy()\n",
    "\n",
    "#Make a new column with a number per category, will be used to colour cells that are added\n",
    "specific_cells['colour']=specific_cells['population_broad'].cat.codes\n",
    "\n",
    "_ = draw_voronoi_scatter(spot=ROI,\n",
    "                         c=specific_cells,\n",
    "                         voronoi_palette = sc.pl.palettes.vega_20_scanpy,\n",
    "                         X='X_loc',\n",
    "                         Y='Y_loc',\n",
    "                         voronoi_hue='LISAclust',\n",
    "                         scatter_hue='colour',\n",
    "                         scatter_palette=sc.pl.palettes.vega_20_scanpy,\n",
    "                         scatter_kwargs={'s':10},\n",
    "                         figsize=(5,5))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
