<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>SpatialBiologyToolkit package &mdash; SpatialBiologyToolkit 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=2709fde1"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Introduction" href="introduction.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SpatialBiologyToolkit
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">SpatialBiologyToolkit package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.correlate_outcomes">SpatialBiologyToolkit.correlate_outcomes module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.correlate_outcomes.adata_population_survival"><code class="docutils literal notranslate"><span class="pre">adata_population_survival()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.correlate_outcomes.create_pca_timeline"><code class="docutils literal notranslate"><span class="pre">create_pca_timeline()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.image_analysis">SpatialBiologyToolkit.image_analysis module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.image_analysis.map_pixel_values_to_colors"><code class="docutils literal notranslate"><span class="pre">map_pixel_values_to_colors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.image_analysis.save_labelled_image"><code class="docutils literal notranslate"><span class="pre">save_labelled_image()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.image_analysis.save_labelled_image_as_svg"><code class="docutils literal notranslate"><span class="pre">save_labelled_image_as_svg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.image_analysis.transform_images"><code class="docutils literal notranslate"><span class="pre">transform_images()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.imc_import_bodenmiller">SpatialBiologyToolkit.imc_import_bodenmiller module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_bodenmiller.import_bodenmiller"><code class="docutils literal notranslate"><span class="pre">import_bodenmiller()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_bodenmiller.reload_dfs"><code class="docutils literal notranslate"><span class="pre">reload_dfs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_bodenmiller.remove_ROIs_and_markers"><code class="docutils literal notranslate"><span class="pre">remove_ROIs_and_markers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_bodenmiller.setup_anndata"><code class="docutils literal notranslate"><span class="pre">setup_anndata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_bodenmiller.stacks_to_imagefolders"><code class="docutils literal notranslate"><span class="pre">stacks_to_imagefolders()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.imc_import_steinpose">SpatialBiologyToolkit.imc_import_steinpose module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_steinpose.cellpose_setup"><code class="docutils literal notranslate"><span class="pre">cellpose_setup()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_steinpose.normalise_markers"><code class="docutils literal notranslate"><span class="pre">normalise_markers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.imc_import_steinpose.split_images_to_channels"><code class="docutils literal notranslate"><span class="pre">split_images_to_channels()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.napari_imc_explorer">SpatialBiologyToolkit.napari_imc_explorer module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.napari_imc_explorer.napari_imc_explorer"><code class="docutils literal notranslate"><span class="pre">napari_imc_explorer()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.plotting">SpatialBiologyToolkit.plotting module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.bargraph"><code class="docutils literal notranslate"><span class="pre">bargraph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.cellabundance_UMAP"><code class="docutils literal notranslate"><span class="pre">cellabundance_UMAP()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.draw_voronoi_scatter"><code class="docutils literal notranslate"><span class="pre">draw_voronoi_scatter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.mlm_stats"><code class="docutils literal notranslate"><span class="pre">mlm_stats()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.obs_to_mask"><code class="docutils literal notranslate"><span class="pre">obs_to_mask()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.overlayed_heatmaps"><code class="docutils literal notranslate"><span class="pre">overlayed_heatmaps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.plotting.plot_colorbar"><code class="docutils literal notranslate"><span class="pre">plot_colorbar()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.population_identification">SpatialBiologyToolkit.population_identification module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.batch_neighbors"><code class="docutils literal notranslate"><span class="pre">batch_neighbors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.consensus"><code class="docutils literal notranslate"><span class="pre">consensus()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.create_remapping"><code class="docutils literal notranslate"><span class="pre">create_remapping()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.leiden"><code class="docutils literal notranslate"><span class="pre">leiden()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.population_summary"><code class="docutils literal notranslate"><span class="pre">population_summary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.prune_leiden_using_dendrogram"><code class="docutils literal notranslate"><span class="pre">prune_leiden_using_dendrogram()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.read_remapping"><code class="docutils literal notranslate"><span class="pre">read_remapping()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.recolor_population"><code class="docutils literal notranslate"><span class="pre">recolor_population()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.population_identification.transfer_populations"><code class="docutils literal notranslate"><span class="pre">transfer_populations()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.scrnaseq_to_imc_matching">SpatialBiologyToolkit.scrnaseq_to_imc_matching module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.data_mapping"><code class="docutils literal notranslate"><span class="pre">data_mapping()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.evaluate_model"><code class="docutils literal notranslate"><span class="pre">evaluate_model()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.predict_groups"><code class="docutils literal notranslate"><span class="pre">predict_groups()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.rank_feature_importance"><code class="docutils literal notranslate"><span class="pre">rank_feature_importance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.train_model"><code class="docutils literal notranslate"><span class="pre">train_model()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.spatial">SpatialBiologyToolkit.spatial module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.Neighborhood_Identification"><code class="docutils literal notranslate"><span class="pre">Neighborhood_Identification()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.analyse_environments"><code class="docutils literal notranslate"><span class="pre">analyse_environments()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.average_boundary_matrices"><code class="docutils literal notranslate"><span class="pre">average_boundary_matrices()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.calculate_boundary_matrix"><code class="docutils literal notranslate"><span class="pre">calculate_boundary_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.calculate_overlap_matrix"><code class="docutils literal notranslate"><span class="pre">calculate_overlap_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.count_label_pixels"><code class="docutils literal notranslate"><span class="pre">count_label_pixels()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.create_env_anndata"><code class="docutils literal notranslate"><span class="pre">create_env_anndata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.create_network_graphs"><code class="docutils literal notranslate"><span class="pre">create_network_graphs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.create_population_anndata"><code class="docutils literal notranslate"><span class="pre">create_population_anndata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.create_spoox_heatmaps"><code class="docutils literal notranslate"><span class="pre">create_spoox_heatmaps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.create_subregions"><code class="docutils literal notranslate"><span class="pre">create_subregions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.lisa_clustering_image"><code class="docutils literal notranslate"><span class="pre">lisa_clustering_image()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.measure_label_objects"><code class="docutils literal notranslate"><span class="pre">measure_label_objects()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.plot_subregion_interactions"><code class="docutils literal notranslate"><span class="pre">plot_subregion_interactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.run_spoox"><code class="docutils literal notranslate"><span class="pre">run_spoox()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.spatial.squidpy_subregion_interactions"><code class="docutils literal notranslate"><span class="pre">squidpy_subregion_interactions()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.utils">SpatialBiologyToolkit.utils module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.adlog"><code class="docutils literal notranslate"><span class="pre">adlog()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.analyze_masks"><code class="docutils literal notranslate"><span class="pre">analyze_masks()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.check_file_names_and_counts"><code class="docutils literal notranslate"><span class="pre">check_file_names_and_counts()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.compare_lists"><code class="docutils literal notranslate"><span class="pre">compare_lists()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.create_celltable"><code class="docutils literal notranslate"><span class="pre">create_celltable()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.extract_pixel_environment"><code class="docutils literal notranslate"><span class="pre">extract_pixel_environment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.extract_single_cell"><code class="docutils literal notranslate"><span class="pre">extract_single_cell()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.get_module_path"><code class="docutils literal notranslate"><span class="pre">get_module_path()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.pip_freeze_to_dataframe"><code class="docutils literal notranslate"><span class="pre">pip_freeze_to_dataframe()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.print_full"><code class="docutils literal notranslate"><span class="pre">print_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.remove_leading_string"><code class="docutils literal notranslate"><span class="pre">remove_leading_string()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.remove_tiff_extensions"><code class="docutils literal notranslate"><span class="pre">remove_tiff_extensions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.rename_folders"><code class="docutils literal notranslate"><span class="pre">rename_folders()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.rename_tiff_files"><code class="docutils literal notranslate"><span class="pre">rename_tiff_files()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.rename_tiff_files_according_to_dict"><code class="docutils literal notranslate"><span class="pre">rename_tiff_files_according_to_dict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.save_unique_tiff_names_to_csv"><code class="docutils literal notranslate"><span class="pre">save_unique_tiff_names_to_csv()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.utils.subset"><code class="docutils literal notranslate"><span class="pre">subset()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit.visium">SpatialBiologyToolkit.visium module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#SpatialBiologyToolkit.visium.create_visium_mask"><code class="docutils literal notranslate"><span class="pre">create_visium_mask()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-SpatialBiologyToolkit">Module contents</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SpatialBiologyToolkit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">SpatialBiologyToolkit package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/SpatialBiologyToolkit.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="spatialbiologytoolkit-package">
<h1>SpatialBiologyToolkit package<a class="headerlink" href="#spatialbiologytoolkit-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-SpatialBiologyToolkit.correlate_outcomes">
<span id="spatialbiologytoolkit-correlate-outcomes-module"></span><h2>SpatialBiologyToolkit.correlate_outcomes module<a class="headerlink" href="#module-SpatialBiologyToolkit.correlate_outcomes" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.correlate_outcomes.adata_population_survival">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.correlate_outcomes.</span></span><span class="sig-name descname"><span class="pre">adata_population_survival</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_input</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeline</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Case'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">survival_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Survival_diagnosis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spatial_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crosstab_normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quartiles</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_quarts_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linear</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cox</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_cell_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Survival_Figures'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/correlate_outcomes.html#adata_population_survival"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.correlate_outcomes.adata_population_survival" title="Link to this definition"></a></dt>
<dd><p>Analyze and visualize the relationship between cell populations and survival data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_input</strong> (<em>ad.AnnData</em>) – Input annotated data matrix.</p></li>
<li><p><strong>timeline</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Precomputed timeline DataFrame. If not provided, it will be computed.</p></li>
<li><p><strong>case_obs</strong> (<em>str</em>) – Column in adata_input.obs representing case IDs.</p></li>
<li><p><strong>survival_obs</strong> (<em>str</em>) – Column in adata_input.obs representing survival time.</p></li>
<li><p><strong>cell_obs</strong> (<em>str</em>) – Column in adata_input.obs representing cell clusters.</p></li>
<li><p><strong>crosstab_normalize</strong> (<em>str</em>) – Normalization method for crosstabulation.</p></li>
<li><p><strong>quartiles</strong> (<em>int</em>) – Number of quartiles to use.</p></li>
<li><p><strong>top_quarts_only</strong> (<em>bool</em>) – Whether to only consider the top and bottom quartiles.</p></li>
<li><p><strong>kmf</strong> (<em>bool</em>) – Whether to plot Kaplan-Meier survival curves.</p></li>
<li><p><strong>linear</strong> (<em>bool</em>) – Whether to perform linear regression analysis.</p></li>
<li><p><strong>cox</strong> (<em>bool</em>) – Whether to perform Cox proportional hazards regression.</p></li>
<li><p><strong>xlim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the x-axis in plots.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix for saved file names.</p></li>
<li><p><strong>scale_cell_fraction</strong> (<em>float</em><em>, </em><em>optional</em>) – Factor to scale cell fractions.</p></li>
<li><p><strong>drop_columns</strong> (<em>list</em>) – List of columns to drop from the timeline DataFrame.</p></li>
<li><p><strong>save_directory</strong> (<em>str</em>) – Directory to save output figures.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing timeline DataFrame, cox p-values DataFrame, ols p-values DataFrame, and cox summary DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.correlate_outcomes.create_pca_timeline">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.correlate_outcomes.</span></span><span class="sig-name descname"><span class="pre">create_pca_timeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_input</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Case'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">survival_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Survival_diagnosis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spatial_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crosstab_normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_loadings</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/correlate_outcomes.html#create_pca_timeline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.correlate_outcomes.create_pca_timeline" title="Link to this definition"></a></dt>
<dd><p>Create a PCA timeline from cell population data and survival information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_input</strong> (<em>ad.AnnData</em>) – Input annotated data matrix.</p></li>
<li><p><strong>n_components</strong> (<em>int</em>) – Number of PCA components.</p></li>
<li><p><strong>case_obs</strong> (<em>str</em>) – Column in adata_input.obs representing case IDs.</p></li>
<li><p><strong>survival_obs</strong> (<em>str</em>) – Column in adata_input.obs representing survival time.</p></li>
<li><p><strong>cell_obs</strong> (<em>str</em>) – Column in adata_input.obs representing cell clusters.</p></li>
<li><p><strong>crosstab_normalize</strong> (<em>str</em>) – Normalization method for crosstabulation.</p></li>
<li><p><strong>show_loadings</strong> (<em>bool</em>) – Whether to show PCA loadings.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Timeline DataFrame with PCA components and survival data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.image_analysis">
<span id="spatialbiologytoolkit-image-analysis-module"></span><h2>SpatialBiologyToolkit.image_analysis module<a class="headerlink" href="#module-SpatialBiologyToolkit.image_analysis" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.image_analysis.map_pixel_values_to_colors">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.image_analysis.</span></span><span class="sig-name descname"><span class="pre">map_pixel_values_to_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_norm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/image_analysis.html#map_pixel_values_to_colors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.image_analysis.map_pixel_values_to_colors" title="Link to this definition"></a></dt>
<dd><p>Map pixel values in an image to hex colors using a specified matplotlib colormap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>image</strong> (<em>np.ndarray</em>) – 2D numpy array of pixel values.</p></li>
<li><p><strong>cmap_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the colormap (default is ‘viridis’).</p></li>
<li><p><strong>min_val</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum value for color scaling. If None, it is set to the minimum value in the image.</p></li>
<li><p><strong>max_val</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for color scaling. If None, it is set to the maximum value in the image.</p></li>
<li><p><strong>quantile</strong> (<em>float</em><em>, </em><em>optional</em>) – If specified, use this quantile to set the max value for color scaling.</p></li>
<li><p><strong>return_norm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return the normalization function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary mapping unique pixel values to hex color codes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.image_analysis.save_labelled_image">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.image_analysis.</span></span><span class="sig-name descname"><span class="pre">save_labelled_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label_image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_mapping</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_axes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_ticks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/image_analysis.html#save_labelled_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.image_analysis.save_labelled_image" title="Link to this definition"></a></dt>
<dd><p>Convert a label image to a color image using a given dictionary of pixel values to HEX colors and save the image using matplotlib.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>label_image</strong> (<em>np.ndarray</em>) – The input label image with pixel values corresponding to the dictionary keys.</p></li>
<li><p><strong>color_mapping</strong> (<em>dict</em>) – A dictionary mapping pixel values (int) to HEX colors (str).</p></li>
<li><p><strong>save_path</strong> (<em>str</em>) – The file path to save the output color image.</p></li>
<li><p><strong>background_color</strong> (<em>str</em><em>, </em><em>optional</em>) – The background color of the saved image. If None, the background will be transparent.</p></li>
<li><p><strong>hide_axes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, hides the axes completely (default is False).</p></li>
<li><p><strong>hide_ticks</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, hides the ticks and labels on the axes (default is True).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.image_analysis.save_labelled_image_as_svg">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.image_analysis.</span></span><span class="sig-name descname"><span class="pre">save_labelled_image_as_svg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_mapping</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_zero</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/image_analysis.html#save_labelled_image_as_svg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.image_analysis.save_labelled_image_as_svg" title="Link to this definition"></a></dt>
<dd><p>Save a labelled image as an SVG file with optimized paths for contiguous pixels of the same color.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>image</strong> (<em>np.ndarray</em>) – A 2D numpy array where each value corresponds to a label.</p></li>
<li><p><strong>color_mapping</strong> (<em>dict</em>) – A dictionary mapping pixel values to colors (hex code or color names).</p></li>
<li><p><strong>save_path</strong> (<em>str</em>) – The name of the output SVG file.</p></li>
<li><p><strong>exclude_zero</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to exclude pixels with a value of 0.</p></li>
<li><p><strong>smoothing_factor</strong> (<em>float</em><em>, </em><em>optional</em>) – Smoothing factor for spline interpolation. Set to 0 for no smoothing.</p></li>
<li><p><strong>background_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Background color for the SVG (default is None).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.image_analysis.transform_images">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.image_analysis.</span></span><span class="sig-name descname"><span class="pre">transform_images</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transformations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transformations_defaults</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_dataframe_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Transformations.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_file_extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.tif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file_extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.tif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_image_as_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_image_in_filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matching_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resize_shape</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_axis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_axis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels_axis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel_names_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['red',</span> <span class="pre">'green',</span> <span class="pre">'blue']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">note_skipped_images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/image_analysis.html#transform_images"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.image_analysis.transform_images" title="Link to this definition"></a></dt>
<dd><p>Process and transform images stored in a directory according to specified transformations,
including optional resizing based on either predefined or dynamically determined dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_directory</strong> (<em>str</em>) – Directory containing input images.</p></li>
<li><p><strong>output_directory</strong> (<em>str</em>) – Directory to save transformed images.</p></li>
<li><p><strong>input_filter</strong> (<em>str</em>) – String that must be in the name of files in the input directory, but will not be considered when matching image names.</p></li>
<li><p><strong>transformations</strong> (<em>list</em>) – List of transformations to apply (e.g., [‘flip_y’, ‘flip_x’, ‘resize’, ‘split_channels’]).</p></li>
<li><p><strong>transformations_defaults</strong> (<em>dict</em>) – Default values for each transformation, including optional resize dimensions.</p></li>
<li><p><strong>image_dataframe_file</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to a CSV file or a DataFrame containing transformation settings.</p></li>
<li><p><strong>input_file_extension</strong> (<em>str</em>) – File extension of input images.</p></li>
<li><p><strong>output_file_extension</strong> (<em>str</em>) – File extension for output images.</p></li>
<li><p><strong>source_image_as_directory</strong> (<em>bool</em>) – If True, create a subdirectory for each image in the output directory.</p></li>
<li><p><strong>source_image_in_filename</strong> (<em>bool</em>) – If True, include the source image name as a prefix in output filenames.</p></li>
<li><p><strong>matching_directory</strong> (<em>str</em>) – Directory containing images used for matching dimensions.</p></li>
<li><p><strong>output_file_prefix</strong> (<em>str</em>) – Prefix for output file names.</p></li>
<li><p><strong>resize_shape</strong> (<em>tuple</em>) – Default dimensions to resize images to (used if matching_directory is not provided).</p></li>
<li><p><strong>y_axis</strong> (<em>int</em>) – Y axis in image array.</p></li>
<li><p><strong>x_axis</strong> (<em>int</em>) – X axis in image array.</p></li>
<li><p><strong>channels_axis</strong> (<em>int</em>) – Channels axis in image array (if input images are multichannel).</p></li>
<li><p><strong>channel_names_list</strong> (<em>list</em>) – List of channel names for split channels transformation.</p></li>
<li><p><strong>note_skipped_images</strong> (<em>bool</em>) – If True, log a warning for images that are skipped.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.imc_import_bodenmiller">
<span id="spatialbiologytoolkit-imc-import-bodenmiller-module"></span><h2>SpatialBiologyToolkit.imc_import_bodenmiller module<a class="headerlink" href="#module-SpatialBiologyToolkit.imc_import_bodenmiller" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_bodenmiller.import_bodenmiller">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_bodenmiller.</span></span><span class="sig-name descname"><span class="pre">import_bodenmiller</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpout'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'panel.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cell.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'image.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'images'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_filename_fullstack</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'FileName_FullStack'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_filename_cellmask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'FileName_cellmask'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_roi_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Metadata_description'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_mcdfile_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Metadata_acname'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Width_FullStack'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file_height</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Height_FullStack'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">acquisition_metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'acquisition_metadata.csv'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_bodenmiller.html#import_bodenmiller"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_bodenmiller.import_bodenmiller" title="Link to this definition"></a></dt>
<dd><p>Import data from the Bodenmiller pipeline and return sample and panel dataframes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> (<em>str</em>) – Directory containing the Bodenmiller pipeline outputs.</p></li>
<li><p><strong>panel_file</strong> (<em>str</em>) – Filename of the panel file.</p></li>
<li><p><strong>cell_file</strong> (<em>str</em>) – Filename of the cell file.</p></li>
<li><p><strong>image_file</strong> (<em>str</em>) – Filename of the image file.</p></li>
<li><p><strong>masks_dir</strong> (<em>str</em>) – Directory containing mask files.</p></li>
<li><p><strong>images_dir</strong> (<em>str</em>) – Directory containing image files.</p></li>
<li><p><strong>image_file_filename_fullstack</strong> (<em>str</em>) – Column name for the full stack filename in the image file.</p></li>
<li><p><strong>image_file_filename_cellmask</strong> (<em>str</em>) – Column name for the cell mask filename in the image file.</p></li>
<li><p><strong>image_file_roi_name</strong> (<em>str</em>) – Column name for the ROI name in the image file.</p></li>
<li><p><strong>image_file_mcdfile_name</strong> (<em>str</em>) – Column name for the MCD file name in the image file.</p></li>
<li><p><strong>image_file_width</strong> (<em>str</em>) – Column name for the image width in the image file.</p></li>
<li><p><strong>image_file_height</strong> (<em>str</em>) – Column name for the image height in the image file.</p></li>
<li><p><strong>acquisition_metadata</strong> (<em>str</em>) – Filename of the acquisition metadata file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the sample dataframe and panel dataframe.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_bodenmiller.reload_dfs">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_bodenmiller.</span></span><span class="sig-name descname"><span class="pre">reload_dfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sample_df.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'panel_df.csv'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_bodenmiller.html#reload_dfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_bodenmiller.reload_dfs" title="Link to this definition"></a></dt>
<dd><p>Reload sample and panel dataframes from disk if they are supplied as paths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_df</strong> (<em>str</em>) – Path to the sample dataframe file.</p></li>
<li><p><strong>panel_df</strong> (<em>str</em>) – Path to the panel dataframe file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the reloaded sample dataframe and panel dataframe.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_bodenmiller.remove_ROIs_and_markers">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_bodenmiller.</span></span><span class="sig-name descname"><span class="pre">remove_ROIs_and_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ROI_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ROIs_to_remove</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Markers_to_remove</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['DNA1',</span> <span class="pre">'DNA3']</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_bodenmiller.html#remove_ROIs_and_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_bodenmiller.remove_ROIs_and_markers" title="Link to this definition"></a></dt>
<dd><p>Remove unused or failed ROIs and/or markers from the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>ROI_obs</strong> (<em>str</em>) – Observation field containing the ROI names.</p></li>
<li><p><strong>ROIs_to_remove</strong> (<em>list</em>) – List of ROIs to remove.</p></li>
<li><p><strong>Markers_to_remove</strong> (<em>list</em>) – List of markers to remove.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object with the specified ROIs and markers removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_bodenmiller.setup_anndata">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_bodenmiller.</span></span><span class="sig-name descname"><span class="pre">setup_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cell.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpout'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sample_df.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'panel_df.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_df_x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Location_Center_X'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_df_y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Location_Center_Y'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_df_ROIcol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dictionary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'dictionary.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_cat_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_df_extra_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_normalisation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'q99.9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_df_target_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Target'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_table_format</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'bodenmmiller'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_normalised_markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_bodenmiller.html#setup_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_bodenmiller.setup_anndata" title="Link to this definition"></a></dt>
<dd><p>Create an AnnData object using the cell table .csv file, and panel_df and sample_df files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_df</strong> (<em>str</em>) – Path to the cell dataframe file.</p></li>
<li><p><strong>directory</strong> (<em>str</em>) – Directory containing the files.</p></li>
<li><p><strong>sample_df</strong> (<em>str</em>) – Path to the sample dataframe file.</p></li>
<li><p><strong>panel_df</strong> (<em>str</em>) – Path to the panel dataframe file.</p></li>
<li><p><strong>cell_df_x</strong> (<em>str</em>) – Column name for the x-coordinates in the cell dataframe.</p></li>
<li><p><strong>cell_df_y</strong> (<em>str</em>) – Column name for the y-coordinates in the cell dataframe.</p></li>
<li><p><strong>cell_df_ROIcol</strong> (<em>str</em>) – Column name for the ROI in the cell dataframe.</p></li>
<li><p><strong>dictionary</strong> (<em>str</em>) – Path to the dictionary file.</p></li>
<li><p><strong>non_cat_obs</strong> (<em>list</em>) – List of observation columns not to be converted to categorical.</p></li>
<li><p><strong>cell_df_extra_columns</strong> (<em>list</em>) – List of extra columns from the cell dataframe to be included.</p></li>
<li><p><strong>marker_normalisation</strong> (<em>str</em>) – Normalisation method for the markers.</p></li>
<li><p><strong>panel_df_target_col</strong> (<em>str</em>) – Column name for the target in the panel dataframe.</p></li>
<li><p><strong>cell_table_format</strong> (<em>str</em>) – Format of the cell table (‘bodenmmiller’ or ‘cleaned’).</p></li>
<li><p><strong>return_normalised_markers</strong> (<em>bool</em>) – Whether to return the normalised markers along with the AnnData object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object containing the processed data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_bodenmiller.stacks_to_imagefolders">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_bodenmiller.</span></span><span class="sig-name descname"><span class="pre">stacks_to_imagefolders</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bodenmiller_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpout'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sample_df.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'panel_df.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unstacked_output_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'images'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_output_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_df_filename_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'FileName_FullStack'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_df_mask_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'FileName_cellmask'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_df_target_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Target'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_bodenmiller.html#stacks_to_imagefolders"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_bodenmiller.stacks_to_imagefolders" title="Link to this definition"></a></dt>
<dd><p>Create mask and image folders using the Bodenmiller outputs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bodenmiller_folder</strong> (<em>str</em>) – Directory containing the Bodenmiller outputs.</p></li>
<li><p><strong>sample_df</strong> (<em>str</em>) – Path to the sample dataframe file.</p></li>
<li><p><strong>panel_df</strong> (<em>str</em>) – Path to the panel dataframe file.</p></li>
<li><p><strong>unstacked_output_folder</strong> (<em>str</em>) – Directory to save the unstacked images.</p></li>
<li><p><strong>masks_output_folder</strong> (<em>str</em>) – Directory to save the renamed masks.</p></li>
<li><p><strong>sample_df_filename_col</strong> (<em>str</em>) – Column name for the filename in the sample dataframe.</p></li>
<li><p><strong>sample_df_mask_col</strong> (<em>str</em>) – Column name for the mask filename in the sample dataframe.</p></li>
<li><p><strong>panel_df_target_col</strong> (<em>str</em>) – Column name for the target in the panel dataframe.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.imc_import_steinpose">
<span id="spatialbiologytoolkit-imc-import-steinpose-module"></span><h2>SpatialBiologyToolkit.imc_import_steinpose module<a class="headerlink" href="#module-SpatialBiologyToolkit.imc_import_steinpose" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_steinpose.cellpose_setup">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_steinpose.</span></span><span class="sig-name descname"><span class="pre">cellpose_setup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cellpose_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intensities_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'intensities'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'img'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_channel_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Images'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionprops_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'regionprops'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionsprops_metrics</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['area']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dictionary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'dictionary.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalisation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['q0.999']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_package_versions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_steinpose.html#cellpose_setup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_steinpose.cellpose_setup" title="Link to this definition"></a></dt>
<dd><p>Takes a Steinpose directory structure and returns an AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cellpose_folder</strong> (<em>str</em>) – Directory containing the Steinpose outputs.</p></li>
<li><p><strong>intensities_folder</strong> (<em>str</em>) – Directory containing the intensity .csv files.</p></li>
<li><p><strong>masks_folder</strong> (<em>str</em>) – Directory containing the mask files.</p></li>
<li><p><strong>images_folder</strong> (<em>str</em>) – Directory containing the image files.</p></li>
<li><p><strong>image_channel_folder</strong> (<em>str</em>) – Directory to save the split channel images.</p></li>
<li><p><strong>regionprops_folder</strong> (<em>str</em>) – Directory containing the regionprops .csv files.</p></li>
<li><p><strong>regionsprops_metrics</strong> (<em>list</em>) – List of regionprops metrics to include.</p></li>
<li><p><strong>dictionary</strong> (<em>str</em>) – Path to the dictionary file.</p></li>
<li><p><strong>normalisation</strong> (<em>list</em>) – List of normalisation methods.</p></li>
<li><p><strong>store_package_versions</strong> (<em>bool</em>) – Whether to store the package versions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object containing the processed data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_steinpose.normalise_markers">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_steinpose.</span></span><span class="sig-name descname"><span class="pre">normalise_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_steinpose.html#normalise_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_steinpose.normalise_markers" title="Link to this definition"></a></dt>
<dd><p>Normalises a markers DataFrame with a list of methods, in sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>markers</strong> (<em>pd.DataFrame</em>) – DataFrame containing marker intensities (cells in rows, markers in columns).</p></li>
<li><p><strong>method_list</strong> (<em>list</em>) – List of normalisation methods.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Normalised DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.imc_import_steinpose.split_images_to_channels">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.imc_import_steinpose.</span></span><span class="sig-name descname"><span class="pre">split_images_to_channels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/imc_import_steinpose.html#split_images_to_channels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.imc_import_steinpose.split_images_to_channels" title="Link to this definition"></a></dt>
<dd><p>Split images outputted from Steinpose into individual channels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_dir</strong> (<em>str</em>) – Directory containing the input images.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Directory to save the split channel images.</p></li>
<li><p><strong>channel_names</strong> (<em>list</em>) – List of channel names.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.napari_imc_explorer">
<span id="spatialbiologytoolkit-napari-imc-explorer-module"></span><h2>SpatialBiologyToolkit.napari_imc_explorer module<a class="headerlink" href="#module-SpatialBiologyToolkit.napari_imc_explorer" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.napari_imc_explorer.napari_imc_explorer">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.napari_imc_explorer.</span></span><span class="sig-name descname"><span class="pre">napari_imc_explorer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">masks_folder:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'Masks',</span> <span class="pre">image_folders:</span> <span class="pre">list</span> <span class="pre">=</span> <span class="pre">['Images'],</span> <span class="pre">HE_folder:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'HE',</span> <span class="pre">roi_obs:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'ROI',</span> <span class="pre">adata:</span> <span class="pre">~anndata._core.anndata.AnnData</span> <span class="pre">=</span> <span class="pre">AnnData</span> <span class="pre">object</span> <span class="pre">with</span> <span class="pre">n_obs</span> <span class="pre">×</span> <span class="pre">n_vars</span> <span class="pre">=</span> <span class="pre">0</span> <span class="pre">×</span> <span class="pre">0,</span> <span class="pre">check_masks:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">False</span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Viewer</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/napari_imc_explorer.html#napari_imc_explorer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.napari_imc_explorer.napari_imc_explorer" title="Link to this definition"></a></dt>
<dd><p>Start an interactive Napari viewer for exploring IMC data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>masks_folder</strong> (<em>str</em>) – Directory containing a mask for each ROI, each file named after the ROI. Masks should be uint16 .tif files.</p></li>
<li><p><strong>image_folders</strong> (<em>list</em>) – Directories containing subdirectories, each named after ROIs in the AnnData. Images are named after channels (adata.var_names), uint16 .tiff files.</p></li>
<li><p><strong>HE_folder</strong> (<em>str</em>) – Directory containing a color (RGB) .tif file, named after each ROI.</p></li>
<li><p><strong>roi_obs</strong> (<em>str</em>) – Column in adata.obs indicating the ROI.</p></li>
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object as created from the pipeline.</p></li>
<li><p><strong>check_masks</strong> (<em>bool</em>) – If True, will check that all the masks match the number of cells in the AnnData object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Napari viewer object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>napari.Viewer</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.plotting">
<span id="spatialbiologytoolkit-plotting-module"></span><h2>SpatialBiologyToolkit.plotting module<a class="headerlink" href="#module-SpatialBiologyToolkit.plotting" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.bargraph">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">bargraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hue_order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specify_populations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_over</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence_interval</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">68</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crosstab_normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells_per_mm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_grid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_figure</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotate_x_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">90</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case_col_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Case'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ROI_col_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Figure</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#bargraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.bargraph" title="Link to this definition"></a></dt>
<dd><p>Plot bar graphs for population abundances or measured values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>AnnData</em><em>, </em><em>DataFrame</em><em>, or </em><em>str</em>) – The input data.</p></li>
<li><p><strong>pop_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column identifying a population.</p></li>
<li><p><strong>value_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column identifying the measured values.</p></li>
<li><p><strong>hue</strong> (<em>str</em><em>, </em><em>optional</em>) – Column to subgroup the graph.</p></li>
<li><p><strong>hue_order</strong> (<em>list</em><em>, </em><em>optional</em>) – Order of hues.</p></li>
<li><p><strong>specify_populations</strong> (<em>list</em><em>, </em><em>optional</em>) – List of specific populations to plot.</p></li>
<li><p><strong>levels</strong> (<em>list</em><em>, </em><em>optional</em>) – Levels at which the data is structured.</p></li>
<li><p><strong>mean_over</strong> (<em>list</em><em>, </em><em>optional</em>) – Levels at which to calculate a mean.</p></li>
<li><p><strong>confidence_interval</strong> (<em>int</em><em>, </em><em>optional</em>) – Confidence interval for error bars.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure.</p></li>
<li><p><strong>crosstab_normalize</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether, and how, to normalize crosstab results.</p></li>
<li><p><strong>cells_per_mm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Normalize values by mm².</p></li>
<li><p><strong>palette</strong> (<em>dict</em><em>, </em><em>optional</em>) – Color palette.</p></li>
<li><p><strong>hide_grid</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to hide the grid.</p></li>
<li><p><strong>legend</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display the legend.</p></li>
<li><p><strong>save_data</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the raw data.</p></li>
<li><p><strong>save_figure</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the figure.</p></li>
<li><p><strong>return_data</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return the data.</p></li>
<li><p><strong>rotate_x_labels</strong> (<em>int</em><em>, </em><em>optional</em>) – Angle of x labels.</p></li>
<li><p><strong>case_col_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the case column.</p></li>
<li><p><strong>ROI_col_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the ROI column.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The data used to create the figure or the figure itself.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame or Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.cellabundance_UMAP">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">cellabundance_UMAP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ROI_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colour_by</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim_red</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Axes</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(3,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Figure</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Axes</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#cellabundance_UMAP"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.cellabundance_UMAP" title="Link to this definition"></a></dt>
<dd><p>Visualize cell population abundance within ROIs using UMAP, PCA, or tSNE with Seaborn’s scatterplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The AnnData object containing single-cell data.</p></li>
<li><p><strong>ROI_id</strong> (<em>str</em>) – Column in adata.obs that contains ROI identifiers.</p></li>
<li><p><strong>population</strong> (<em>str</em>) – Column in adata.obs that contains cell population identifiers.</p></li>
<li><p><strong>colour_by</strong> (<em>str</em><em>, </em><em>optional</em>) – Column to color points by. If None, points are colored by ROI_id.</p></li>
<li><p><strong>annotate</strong> (<em>bool</em><em>, </em><em>optional</em>) – Annotate points with ROI identifiers.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the plot.</p></li>
<li><p><strong>normalize</strong> (<em>bool</em><em>, </em><em>optional</em>) – Normalize cell counts across ROIs.</p></li>
<li><p><strong>dim_red</strong> (<em>str</em><em>, </em><em>optional</em>) – Dimensionality reduction technique to use (‘UMAP’, ‘PCA’, ‘tSNE’).</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – Axes on which to draw the plot. If None, a new figure and axes object are created.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Matplotlib colormap to use.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure to create. Ignored if ‘ax’ is not None.</p></li>
<li><p><strong>point_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the points in the scatter plot.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The figure or axes object containing the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>plt.Figure or plt.Axes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.draw_voronoi_scatter">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">draw_voronoi_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">voronoi_palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[(0.00784313725490196,</span> <span class="pre">0.24313725490196078,</span> <span class="pre">1.0),</span> <span class="pre">(1.0,</span> <span class="pre">0.48627450980392156,</span> <span class="pre">0.0),</span> <span class="pre">(0.10196078431372549,</span> <span class="pre">0.788235294117647,</span> <span class="pre">0.2196078431372549),</span> <span class="pre">(0.9098039215686274,</span> <span class="pre">0.0,</span> <span class="pre">0.043137254901960784),</span> <span class="pre">(0.5450980392156862,</span> <span class="pre">0.16862745098039217,</span> <span class="pre">0.8862745098039215),</span> <span class="pre">(0.6235294117647059,</span> <span class="pre">0.2823529411764706,</span> <span class="pre">0.0),</span> <span class="pre">(0.9450980392156862,</span> <span class="pre">0.2980392156862745,</span> <span class="pre">0.7568627450980392),</span> <span class="pre">(0.6392156862745098,</span> <span class="pre">0.6392156862745098,</span> <span class="pre">0.6392156862745098),</span> <span class="pre">(1.0,</span> <span class="pre">0.7686274509803922,</span> <span class="pre">0.0),</span> <span class="pre">(0.0,</span> <span class="pre">0.8431372549019608,</span> <span class="pre">1.0)]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'voronoi'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X:X'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Y:Y'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">voronoi_hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'neighborhood10'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ClusterName'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">voronoi_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#draw_voronoi_scatter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.draw_voronoi_scatter" title="Link to this definition"></a></dt>
<dd><p>Plot Voronoi of a region and overlay the location of specific cell types.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spot</strong> (<em>pd.DataFrame</em>) – Cells that are used for Voronoi diagram.</p></li>
<li><p><strong>c</strong> (<em>pd.DataFrame</em>) – Cells that are plotted over Voronoi.</p></li>
<li><p><strong>voronoi_palette</strong> (<em>list</em><em>, </em><em>optional</em>) – Color palette used for coloring neighborhoods. Default is sb.color_palette(‘bright’).</p></li>
<li><p><strong>scatter_palette</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – Color palette for scatter plot. Default is ‘voronoi’.</p></li>
<li><p><strong>X</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name used for X locations. Default is ‘X:X’.</p></li>
<li><p><strong>Y</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name used for Y locations. Default is ‘Y:Y’.</p></li>
<li><p><strong>voronoi_hue</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name used for neighborhood allocation. Default is ‘neighborhood10’.</p></li>
<li><p><strong>scatter_hue</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name used for scatter plot. Default is ‘ClusterName’.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of figure. Default is (5, 5).</p></li>
<li><p><strong>voronoi_kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Arguments passed to plot_voronoi function. Default is {}.</p></li>
<li><p><strong>scatter_kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Arguments passed to plt.scatter(). Default is {}.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sizes of each Voronoi region.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.mlm_stats">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">mlm_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Case'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Cells'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'holm-sidak'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">average_cases</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_t_values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_t_tests</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#mlm_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.mlm_stats" title="Link to this definition"></a></dt>
<dd><p>Conduct mixed linear model analysis and optional t-test on data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – The input data.</p></li>
<li><p><strong>pop_col</strong> (<em>str</em>) – Column identifying different populations.</p></li>
<li><p><strong>group_col</strong> (<em>str</em>) – Column identifying 2 groups.</p></li>
<li><p><strong>case_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column identifying cases.</p></li>
<li><p><strong>value_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column containing the values to be analyzed.</p></li>
<li><p><strong>roi_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column identifying ROIs.</p></li>
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em>) – Method to correct for multiple comparisons.</p></li>
<li><p><strong>average_cases</strong> (<em>bool</em><em>, </em><em>optional</em>) – Average results over cases before performing t-test.</p></li>
<li><p><strong>show_t_values</strong> (<em>bool</em><em>, </em><em>optional</em>) – Include t-values in the output.</p></li>
<li><p><strong>run_t_tests</strong> (<em>bool</em><em>, </em><em>optional</em>) – Run t-tests.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>P-values of the mixed linear model analysis and t-tests.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.obs_to_mask">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">obs_to_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cat_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cat_colour_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cat_obs_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quant_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quant_colour_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_colormap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_ext</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_axes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_ticks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">svg_smoothing_factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#obs_to_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.obs_to_mask" title="Link to this definition"></a></dt>
<dd><p>Map values from an AnnData object to a mask and generate a color map, with options to save the resulting image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>roi</strong> (<em>str</em>) – Region of interest identifier.</p></li>
<li><p><strong>roi_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in adata.obs that contains ROI information (default is ‘ROI’).</p></li>
<li><p><strong>cat_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Categorical observation to map.</p></li>
<li><p><strong>cat_colour_map</strong> (<em>str</em><em>, </em><em>list</em><em> or </em><em>dict</em><em>, </em><em>optional</em>) – Colormap for categorical data (default is ‘tab20’).</p></li>
<li><p><strong>cat_obs_groups</strong> (<em>list</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Specific groups to include.</p></li>
<li><p><strong>quant_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Quantitative observation to map.</p></li>
<li><p><strong>quant_colour_map</strong> (<em>str</em><em>, </em><em>optional</em>) – Colormap for quantitative data (default is ‘viridis’).</p></li>
<li><p><strong>adata_colormap</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use colormap from AnnData if available (default is True).</p></li>
<li><p><strong>masks_folder</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder containing mask files (default is ‘Masks’).</p></li>
<li><p><strong>masks_ext</strong> (<em>str</em><em>, </em><em>optional</em>) – File extension for mask files (default is ‘tif’).</p></li>
<li><p><strong>min_val</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum value for quantitative color scaling.</p></li>
<li><p><strong>max_val</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for quantitative color scaling.</p></li>
<li><p><strong>quantile</strong> (<em>float</em><em>, </em><em>optional</em>) – Quantile to set the maximum value for quantitative color scaling.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the output image. Supports PNG and SVG formats.</p></li>
<li><p><strong>background_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Background color of the saved image. If None, the background will be transparent for PNG files.</p></li>
<li><p><strong>hide_axes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, hides the axes completely in the saved image (default is False).</p></li>
<li><p><strong>hide_ticks</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, hides the ticks and labels on the axes in the saved image (default is True).</p></li>
<li><p><strong>svg_smoothing_factor</strong> (<em>int</em><em>, </em><em>optional</em>) – Smoothing factor for SVG output (default is 0).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapped mask, pixel colormap, and category dictionary (if applicable).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.overlayed_heatmaps">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">overlayed_heatmaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dfs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmaps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Reds',</span> <span class="pre">'Greens',</span> <span class="pre">'Greens',</span> <span class="pre">'Purples',</span> <span class="pre">'Oranges',</span> <span class="pre">'Greys']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'horizontal'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(10,</span> <span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_linewidth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar_plot_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{'pad':</span> <span class="pre">0.25,</span> <span class="pre">'size':</span> <span class="pre">'2%',</span> <span class="pre">'spacing':</span> <span class="pre">0.25}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">heatmap_kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#overlayed_heatmaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.overlayed_heatmaps" title="Link to this definition"></a></dt>
<dd><p>Overlay multiple heatmaps on the same plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dfs</strong> (<em>list</em><em> of </em><em>pd.DataFrame</em>) – List of DataFrames to plot for ‘horizontal’ mode; exactly two for ‘diagonal’ mode. DataFrames must be the same size.</p></li>
<li><p><strong>cmaps</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of colormaps for each DataFrame.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – ‘horizontal’ for horizontal bars; ‘diagonal’ for a diagonal split of the second DataFrame.</p></li>
<li><p><strong>vmin_list</strong> (<em>list</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – List of vmin values. Default is to calculate automatically.</p></li>
<li><p><strong>vmax_list</strong> (<em>list</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – List of vmax values. Default is to calculate automatically.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Tuple for the figure size.</p></li>
<li><p><strong>grid_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color of the manually plotted grid lines.</p></li>
<li><p><strong>grid_linewidth</strong> (<em>int</em><em>, </em><em>optional</em>) – Width of the manually plotted grid lines.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save the output image. If None, the image is not saved.</p></li>
<li><p><strong>colorbar_plot_list</strong> (<em>list</em><em> of </em><em>bool</em><em>, </em><em>optional</em>) – By default all will be plotted, can alternatively provide a list of True/False entries to indicate which should be plotted.</p></li>
<li><p><strong>colorbar_params</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary to customize the size and placement of colour bars (size, pad, spacing).</p></li>
<li><p><strong>colorbar_labels</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of labels to associate with colorbars.</p></li>
<li><p><strong>**heatmap_kwargs</strong> – Additional keyword arguments for sns.heatmap.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.plotting.plot_colorbar">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_colorbar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aspect</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_ticks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Figure</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/plotting.html#plot_colorbar"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.plotting.plot_colorbar" title="Link to this definition"></a></dt>
<dd><p>Plot a colorbar based on the provided array and customization parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) – Input array to generate the colorbar. If the array is 1-dimensional,
it will be repeated to form a 2-dimensional array. If not provided,
a default array of shape (100, 100) filled with <cite>vmax</cite> will be used.</p></li>
<li><p><strong>vmin</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum data value that corresponds to the lower limit of the color scale.
Defaults to None, in which case the minimum value in <cite>array</cite> is used.</p></li>
<li><p><strong>vmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum data value that corresponds to the upper limit of the color scale.
Defaults to None, in which case the maximum value in <cite>array</cite> is used.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Colormap used for the colorbar. Defaults to ‘viridis’.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the colorbar image. If provided, the colorbar will be saved
to this file path, and the function will not return any value. Defaults to None.</p></li>
<li><p><strong>aspect</strong> (<em>int</em><em>, </em><em>optional</em>) – Aspect ratio of the colorbar. Defaults to 10.</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Font size for the colorbar tick labels. Defaults to None, in which case the
default font size is used.</p></li>
<li><p><strong>hide_ticks</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the colorbar ticks will be hidden. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The generated figure containing the colorbar, if <cite>save</cite> is not provided.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.population_identification">
<span id="spatialbiologytoolkit-population-identification-module"></span><h2>SpatialBiologyToolkit.population_identification module<a class="headerlink" href="#module-SpatialBiologyToolkit.population_identification" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.batch_neighbors">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">batch_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'bbknn'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_correction_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Case'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_for_pca</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#batch_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.batch_neighbors" title="Link to this definition"></a></dt>
<dd><p>Perform batch correction and preprocessing on an AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>correction_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Method for batch correction. Options are ‘bbknn’, ‘harmony’, ‘both’, or None.
Default is ‘bbknn’.</p></li>
<li><p><strong>batch_correction_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Observation key for batch correction.
Default is ‘Case’.</p></li>
<li><p><strong>n_for_pca</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of principal components to use. If None, it defaults to one less than the number of markers.
Default is None.</p></li>
<li><p><strong>save</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the results.
Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If <cite>n_for_pca</cite> is not specified, it is set to one less than the number of markers in <cite>adata.var_names</cite>.</p></li>
<li><p>The function performs PCA followed by the specified batch correction method.</p></li>
<li><p>The function logs progress and results using <cite>adlog</cite>.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.consensus">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[3,</span> <span class="pre">5,</span> <span class="pre">10]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_runs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_range</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#consensus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.consensus" title="Link to this definition"></a></dt>
<dd><p>Run consensus clustering to cluster cells in an AnnData object.</p>
<p>This function requires the user to perform dimensionality
reduction using PCA (<cite>scanpy.tl.pca</cite>) first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>n_clusters</strong> (<em>list</em><em>, </em><em>optional</em>) – Number of clusters. Default is [3, 5, 10].</p></li>
<li><p><strong>n_runs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of realizations to perform for the consensus. Default is 25.</p></li>
<li><p><strong>d_range</strong> (<em>list</em><em>, </em><em>optional</em>) – Number of PCs. Default is 25, or the number of PCs in the
AnnData object, whichever is lower. Can accept a list
(e.g. [15, 20, 25]).</p></li>
<li><p><strong>save</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the results. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If <cite>n_clusters</cite> is not a list, it will be converted to a list.</p></li>
<li><p>The function logs progress and results using <cite>adlog</cite>.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.create_remapping">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">create_remapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['population',</span> <span class="pre">'population_broad',</span> <span class="pre">'hierarchy']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#create_remapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.create_remapping" title="Link to this definition"></a></dt>
<dd><p>Creates a mapping file for renaming populations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>obs_column</strong> (<em>str</em>) – Observation column containing the populations to be remapped.</p></li>
<li><p><strong>groups</strong> (<em>list</em><em>, </em><em>optional</em>) – List of group names for the mapping. Default is [‘population’, ‘population_broad’, ‘hierarchy’].</p></li>
<li><p><strong>file</strong> (<em>str</em><em>, </em><em>optional</em>) – File name to save the mapping file. If not provided, a default name based on <cite>obs_column</cite> is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.leiden">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">leiden</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_obs_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to_existing_leiden</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">existing_leiden_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#leiden"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.leiden" title="Link to this definition"></a></dt>
<dd><p>Perform Leiden clustering on an AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em>) – Resolution parameter for the Leiden algorithm.
Default is 0.3.</p></li>
<li><p><strong>leiden_obs_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the observation field to store the Leiden clustering results.
Default is ‘leiden_&lt;resolution&gt;’.</p></li>
<li><p><strong>restrict_to_existing_leiden</strong> (<em>str</em><em>, </em><em>optional</em>) – Restrict clustering to specific existing Leiden groups.
Default is None.</p></li>
<li><p><strong>existing_leiden_groups</strong> (<em>list</em><em>, </em><em>optional</em>) – Specific groups to restrict clustering to. If not a list, it will be converted to a list.
Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If <cite>leiden_obs_name</cite> is not specified, it defaults to ‘leiden_&lt;resolution&gt;’.</p></li>
<li><p>If <cite>existing_leiden_groups</cite> is not a list, it will be converted to a list.</p></li>
<li><p>If <cite>restrict_to_existing_leiden</cite> is specified, clustering is restricted to the specified groups.</p></li>
<li><p>The function logs progress and results using <cite>adlog</cite>.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.population_summary">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">population_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categorical_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heatmap_vmax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_log_scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_point_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">display_tables</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#population_summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.population_summary" title="Link to this definition"></a></dt>
<dd><p>Produces graphs to summarize clustering performance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>groupby_obs</strong> (<em>str</em>) – Observation field to group by.</p></li>
<li><p><strong>markers</strong> (<em>list</em><em>, </em><em>optional</em>) – List of markers to use. If not specified, all markers are used.</p></li>
<li><p><strong>categorical_obs</strong> (<em>list</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – List of categorical observations. Default is an empty list.</p></li>
<li><p><strong>heatmap_vmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for the heatmap.</p></li>
<li><p><strong>graph_log_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use logarithmic scale for graphs. Default is True.</p></li>
<li><p><strong>umap_point_size</strong> (<em>float</em><em>, </em><em>optional</em>) – Size of points in UMAP plot.</p></li>
<li><p><strong>display_tables</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display tables. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If <cite>markers</cite> is not specified, all markers in <cite>adata.var_names</cite> are used.</p></li>
<li><p>Logs and visualizations are generated using <cite>scanpy</cite> plotting functions.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.prune_leiden_using_dendrogram">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">prune_leiden_using_dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'leiden_merged'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_leiden</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_pop_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#prune_leiden_using_dendrogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.prune_leiden_using_dendrogram" title="Link to this definition"></a></dt>
<dd><p>Use the results of a dendrogram to reduce the number of Leiden populations by merging them into larger and more robust populations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object.</p></li>
<li><p><strong>leiden_obs</strong> (<em>str</em>) – Observation identifying the Leiden clusters to use.</p></li>
<li><p><strong>new_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – New observation to be created in the AnnData. Default is ‘leiden_merged’.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode for merging clusters. Options are ‘max’, ‘size’, or ‘percent’. Default is ‘max’.</p></li>
<li><p><strong>max_leiden</strong> (<em>int</em><em>, </em><em>optional</em>) – Largest Leiden population number to keep when mode is ‘max’. Default is None.</p></li>
<li><p><strong>minimum_pop_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum population size to keep when mode is ‘size’ or ‘percent’. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary mapping old clusters to new clusters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.read_remapping">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">read_remapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#read_remapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.read_remapping" title="Link to this definition"></a></dt>
<dd><p>Reads a mapping file for renaming populations and adds the details of the population to <cite>adata.uns</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>obs_column</strong> (<em>str</em>) – Observation column containing the populations to be remapped.</p></li>
<li><p><strong>file</strong> (<em>str</em><em>, </em><em>optional</em>) – File name to read the mapping from. If not provided, a default name based on <cite>obs_column</cite> is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing the remapping information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.recolor_population">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">recolor_population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#recolor_population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.recolor_population" title="Link to this definition"></a></dt>
<dd><p>Allow the user to recolor populations in an AnnData object using a GUI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>population_obs</strong> (<em>str</em>) – Observation field containing the populations to be recolored.</p></li>
<li><p><strong>save</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the AnnData object before and after recoloring. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.population_identification.transfer_populations">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.population_identification.</span></span><span class="sig-name descname"><span class="pre">transfer_populations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_source_populations_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_target_populations_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">common_cell_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Master_Index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/population_identification.html#transfer_populations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.population_identification.transfer_populations" title="Link to this definition"></a></dt>
<dd><p>Transfer populations from a source AnnData object to a target AnnData object based on a common cell index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_source</strong> (<em>AnnData</em>) – Source AnnData object containing the populations to be transferred.</p></li>
<li><p><strong>adata_source_populations_obs</strong> (<em>str</em>) – Observation field in the source AnnData object that contains the population information.</p></li>
<li><p><strong>adata_target</strong> (<em>AnnData</em>) – Target AnnData object where the populations will be transferred to.</p></li>
<li><p><strong>adata_target_populations_obs</strong> (<em>str</em>) – Observation field in the target AnnData object where the population information will be stored.</p></li>
<li><p><strong>common_cell_index</strong> (<em>str</em><em>, </em><em>optional</em>) – Common cell index used to match cells between the source and target AnnData objects.
Default is ‘Master_Index’.</p></li>
<li><p><strong>pop_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to add to the population names in the target AnnData object.
Default is an empty string.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.scrnaseq_to_imc_matching">
<span id="spatialbiologytoolkit-scrnaseq-to-imc-matching-module"></span><h2>SpatialBiologyToolkit.scrnaseq_to_imc_matching module<a class="headerlink" href="#module-SpatialBiologyToolkit.scrnaseq_to_imc_matching" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.scrnaseq_to_imc_matching.data_mapping">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.scrnaseq_to_imc_matching.</span></span><span class="sig-name descname"><span class="pre">data_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_sc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_imc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_protein_mapping</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_protein_mapping_usecol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_protein_mapping_sc_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'snRNAseq'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_protein_mapping_imc_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'IMC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_pop_exclude</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'population'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imc_pop_exclude</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imc_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'population'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neurons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'adam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">32</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/scrnaseq_to_imc_matching.html#data_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.data_mapping" title="Link to this definition"></a></dt>
<dd><p>Train neural network models to map single-cell data to IMC populations and predict groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_sc</strong> (<em>ad.AnnData</em>) – Single-cell AnnData object.</p></li>
<li><p><strong>adata_imc</strong> (<em>ad.AnnData</em>) – IMC AnnData object.</p></li>
<li><p><strong>gene_protein_mapping</strong> (<em>pd.DataFrame</em>) – DataFrame mapping genes to proteins.</p></li>
<li><p><strong>gene_protein_mapping_usecol</strong> (<em>str</em>) – Column in gene_protein_mapping indicating which mappings to use.</p></li>
<li><p><strong>gene_protein_mapping_sc_col</strong> (<em>str</em>) – Column name for single-cell gene mappings.</p></li>
<li><p><strong>gene_protein_mapping_imc_col</strong> (<em>str</em>) – Column name for IMC protein mappings.</p></li>
<li><p><strong>sc_pop_exclude</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of single-cell populations to exclude.</p></li>
<li><p><strong>sc_obs</strong> (<em>str</em>) – Observation column for single-cell populations.</p></li>
<li><p><strong>imc_pop_exclude</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of IMC populations to exclude.</p></li>
<li><p><strong>imc_obs</strong> (<em>str</em>) – Observation column for IMC populations.</p></li>
<li><p><strong>scale</strong> (<em>str</em>) – Scaling method (‘ss’ for StandardScaler, ‘mm’ for MinMaxScaler).</p></li>
<li><p><strong>confidence_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Confidence threshold for predictions.</p></li>
<li><p><strong>neurons</strong> (<em>int</em>) – Number of neurons in hidden layers.</p></li>
<li><p><strong>optimizer</strong> (<em>str</em>) – Optimizer for training the model.</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – Batch size for training.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.scrnaseq_to_imc_matching.evaluate_model">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.scrnaseq_to_imc_matching.</span></span><span class="sig-name descname"><span class="pre">evaluate_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequential</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/scrnaseq_to_imc_matching.html#evaluate_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.evaluate_model" title="Link to this definition"></a></dt>
<dd><p>Evaluate the accuracy of the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>Sequential</em>) – Trained neural network model.</p></li>
<li><p><strong>X</strong> (<em>np.ndarray</em>) – Input features.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Target labels.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Accuracy of the model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.scrnaseq_to_imc_matching.predict_groups">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.scrnaseq_to_imc_matching.</span></span><span class="sig-name descname"><span class="pre">predict_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequential</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codes_to_category</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">below_limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/scrnaseq_to_imc_matching.html#predict_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.predict_groups" title="Link to this definition"></a></dt>
<dd><p>Predict cell populations using the trained model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>Sequential</em>) – Trained neural network model.</p></li>
<li><p><strong>X</strong> (<em>np.ndarray</em>) – Input features for prediction.</p></li>
<li><p><strong>codes_to_category</strong> (<em>Dict</em><em>[</em><em>int</em><em>, </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Mapping of codes to categories.</p></li>
<li><p><strong>scale</strong> (<em>str</em>) – Scaling method (‘ss’ for StandardScaler, ‘mm’ for MinMaxScaler).</p></li>
<li><p><strong>confidence_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for prediction confidence.</p></li>
<li><p><strong>below_limit</strong> (<em>int</em><em>, </em><em>optional</em>) – Value to assign if confidence is below threshold.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Predicted classes and their confidence scores.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[np.ndarray, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.scrnaseq_to_imc_matching.rank_feature_importance">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.scrnaseq_to_imc_matching.</span></span><span class="sig-name descname"><span class="pre">rank_feature_importance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequential</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/scrnaseq_to_imc_matching.html#rank_feature_importance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.rank_feature_importance" title="Link to this definition"></a></dt>
<dd><p>Rank feature importance using permutation importance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>Sequential</em>) – Trained neural network model.</p></li>
<li><p><strong>X</strong> (<em>np.ndarray</em>) – Input features.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Target labels.</p></li>
<li><p><strong>feature_names</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of feature names.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing feature names and their importance scores.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.scrnaseq_to_imc_matching.train_model">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.scrnaseq_to_imc_matching.</span></span><span class="sig-name descname"><span class="pre">train_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neurons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">16</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'adam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">32</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Sequential</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Sequential</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/scrnaseq_to_imc_matching.html#train_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.scrnaseq_to_imc_matching.train_model" title="Link to this definition"></a></dt>
<dd><p>Train a neural network model to classify cell populations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em>) – Input features.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Target labels.</p></li>
<li><p><strong>scale</strong> (<em>str</em>) – Scaling method (‘ss’ for StandardScaler, ‘mm’ for MinMaxScaler).</p></li>
<li><p><strong>adata</strong> (<em>ad.AnnData</em><em>, </em><em>optional</em>) – Annotated data matrix.</p></li>
<li><p><strong>obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Observation column in AnnData.</p></li>
<li><p><strong>neurons</strong> (<em>int</em>) – Number of neurons in hidden layers.</p></li>
<li><p><strong>optimizer</strong> (<em>str</em>) – Optimizer for training the model.</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – Batch size for training.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Trained model and optionally a mapping of codes to categories.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Union[Sequential, Tuple[Sequential, Dict[int, str]]]</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.spatial">
<span id="spatialbiologytoolkit-spatial-module"></span><h2>SpatialBiologyToolkit.spatial module<a class="headerlink" href="#module-SpatialBiologyToolkit.spatial" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.Neighborhood_Identification">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">Neighborhood_Identification</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[20]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Y_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['abundancy',</span> <span class="pre">'connectivity']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connect_suffix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_raw</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connectivity_modes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['conn_all',</span> <span class="pre">'conn_self']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#Neighborhood_Identification"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.Neighborhood_Identification" title="Link to this definition"></a></dt>
<dd><p>Identify neighborhoods in spatial data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em><em>, </em><em>AnnData</em><em>, or </em><em>str</em>) – Input data. Can be a DataFrame, AnnData object, or path to a CSV file.</p></li>
<li><p><strong>cluster_col</strong> (<em>str</em>) – Column defining the populations.</p></li>
<li><p><strong>ks</strong> (<em>list</em><em>, </em><em>optional</em>) – List of window sizes to try. Default is [20].</p></li>
<li><p><strong>keep_cols</strong> (<em>str</em><em>, </em><em>optional</em>) – Columns to keep in the output metadata. Default is ‘all’.</p></li>
<li><p><strong>radius</strong> (<em>float</em><em>, </em><em>optional</em>) – Radius at which cells are considered connected/interacting. Default is 20.</p></li>
<li><p><strong>X</strong> (<em>str</em><em>, </em><em>optional</em>) – Column defining X location. Default is ‘X_loc’.</p></li>
<li><p><strong>Y</strong> (<em>str</em><em>, </em><em>optional</em>) – Column defining Y location. Default is ‘Y_loc’.</p></li>
<li><p><strong>reg</strong> (<em>str</em><em>, </em><em>optional</em>) – Column defining each separate ROI. Default is ‘ROI’.</p></li>
<li><p><strong>modes</strong> (<em>list</em><em>, </em><em>optional</em>) – List of modes for neighborhood identification. Default is [‘abundancy’, ‘connectivity’].</p></li>
<li><p><strong>connect_suffix</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, add suffix to connectivity column names. Default is True.</p></li>
<li><p><strong>return_raw</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return raw data. Default is False.</p></li>
<li><p><strong>bootstrap</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bootstrap iterations. Default is 75.</p></li>
<li><p><strong>connectivity_modes</strong> (<em>list</em><em>, </em><em>optional</em>) – List of connectivity modes to calculate. Default is [‘conn_all’, ‘conn_self’].</p></li>
<li><p><strong>reset_index</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reset the index of the input data. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of modalities if return_raw is True, else AnnData object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.analyse_environments">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">analyse_environments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'summary'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_cores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folder_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'images'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_loc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_loc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Y_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_index_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Master_Index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_quant_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert_value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#analyse_environments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.analyse_environments" title="Link to this definition"></a></dt>
<dd><p>Analyzes the environmental texture around cells in given images, extracting specified parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>ad.AnnData</em>) – Anndata object where the cell locations are stored.</p></li>
<li><p><strong>samples_list</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of sample identifiers to include in the analysis.</p></li>
<li><p><strong>marker_list</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of markers to analyze.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Operation mode, ‘summary’ by default.</p></li>
<li><p><strong>radius</strong> (<em>int</em><em>, </em><em>optional</em>) – Radius of the square area around the cell for analysis.</p></li>
<li><p><strong>num_cores</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cores to use for multiprocessing.</p></li>
<li><p><strong>folder_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory containing the image files.</p></li>
<li><p><strong>roi_id</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in <cite>adata</cite> indicating the ROI.</p></li>
<li><p><strong>x_loc</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in <cite>adata</cite> specifying the x-coordinate of the cell.</p></li>
<li><p><strong>y_loc</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in <cite>adata</cite> specifying the y-coordinate of the cell.</p></li>
<li><p><strong>cell_index_id</strong> (<em>str</em><em>, </em><em>optional</em>) – Identifier for cells in the dataframe.</p></li>
<li><p><strong>quantile</strong> (<em>Optional</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) – Quantile to determine the maximum intensity for image scaling. If None, no normalization is applied.
The quantile is calculated over all the images for that marker.</p></li>
<li><p><strong>parameters</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – List of strings specifying which features to calculate. Features can include ‘Mean’, ‘Median’, ‘Std’,
‘Kurtosis’, ‘Skew’, and texture features like ‘Contrast’, ‘Dissimilarity’, ‘Homogeneity’, ‘Energy’,
‘Correlation’, ‘ASM’, as well as quantiles (e.g., ‘Quantile_0.5’). Default parameter list is:
[‘Mean’, ‘Std’, ‘Quantile_0.1’, ‘Quantile_0.5’, ‘Quantile_0.9’].</p></li>
<li><p><strong>return_quant_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, returns a table of quantile values.</p></li>
<li><p><strong>invert_value</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – Pixel values for images will be subtracted from this number, if given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Depending on <cite>return_quant_table</cite>, returns either a DataFrame of results or a tuple containing
the DataFrame and a quantile table.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Union[pd.DataFrame, Tuple[pd.DataFrame, pd.DataFrame]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.average_boundary_matrices">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">average_boundary_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#average_boundary_matrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.average_boundary_matrices" title="Link to this definition"></a></dt>
<dd><p>Average a list of boundary matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dataframes</strong> (<em>List</em><em>[</em><em>pd.DataFrame</em><em>]</em>) – List of DataFrames to average.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The averaged DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.calculate_boundary_matrix">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">calculate_boundary_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labeled_image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_to_all_boundaries</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#calculate_boundary_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.calculate_boundary_matrix" title="Link to this definition"></a></dt>
<dd><p>Calculate a boundary matrix showing the number of boundary pixels shared between each pair of labels in a labeled image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labeled_image</strong> (<em>np.ndarray</em>) – A 2D labeled image where each unique integer represents a different label.</p></li>
<li><p><strong>normalize</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, normalize the boundary counts by the total number of non-zero pixels.</p></li>
<li><p><strong>normalize_to_all_boundaries</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, normalize the boundary counts by the total number of boundary pixels.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame where rows and columns represent unique labels, and values represent the number of boundary pixels shared.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.calculate_overlap_matrix">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">calculate_overlap_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label_image1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_image2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_zeros</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#calculate_overlap_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.calculate_overlap_matrix" title="Link to this definition"></a></dt>
<dd><p>Calculate the overlap matrix between two label images.</p>
<p>This function calculates the overlap between the unique labels in two label images of equal size.
It can ignore zero values and normalize the overlap counts by the total number of overlapping pixels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>label_image1</strong> (<em>np.ndarray</em>) – The first labeled image.</p></li>
<li><p><strong>label_image2</strong> (<em>np.ndarray</em>) – The second labeled image.</p></li>
<li><p><strong>ignore_zeros</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, ignores the zero values in both images during calculation.</p></li>
<li><p><strong>normalize</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, normalizes the overlap counts by the total number of overlapping pixels.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame where the rows represent unique labels from the first image,
the columns represent unique labels from the second image, and the values
represent the number of overlapping pixels between each pair of labels.
If normalization is applied, the values are normalized accordingly.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input images do not have the same dimensions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.count_label_pixels">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">count_label_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labeled_image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#count_label_pixels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.count_label_pixels" title="Link to this definition"></a></dt>
<dd><p>Count the number of pixels for each unique label in a labeled image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>labeled_image</strong> (<em>np.ndarray</em>) – A 2D labeled image where each unique integer represents a different label.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame with labels as the index and their respective pixel counts.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.create_env_anndata">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">create_env_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">master_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_anndata_for_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_to_transfer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Master_Index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_quantile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.99</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_unmeasured_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#create_env_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.create_env_anndata" title="Link to this definition"></a></dt>
<dd><p>Create an AnnData object for environmental analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>master_list</strong> (<em>pd.DataFrame</em>) – DataFrame containing the measurements.</p></li>
<li><p><strong>source_anndata_for_obs</strong> (<em>Optional</em><em>[</em><em>ad.AnnData</em><em>]</em><em>, </em><em>optional</em>) – Source AnnData object to extract .obs data from.</p></li>
<li><p><strong>obs_to_transfer</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – List of observation column names to transfer from source AnnData.</p></li>
<li><p><strong>cell_index</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name to be used as index in the resulting AnnData.</p></li>
<li><p><strong>norm_quantile</strong> (<em>float</em><em>, </em><em>optional</em>) – Quantile for normalization.</p></li>
<li><p><strong>drop_unmeasured_cells</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, drop cells that couldn’t be measured (sum=0).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The created AnnData object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.create_network_graphs">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">create_network_graphs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spoox_figures'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Reds'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'MH_FDR_less':</span> <span class="pre">0.05,</span> <span class="pre">'MH_SES':</span> <span class="pre">(0,</span> <span class="pre">100),</span> <span class="pre">'gr20</span> <span class="pre">PCF</span> <span class="pre">lower_greater':</span> <span class="pre">1,</span> <span class="pre">'gr20_greater':</span> <span class="pre">1}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_1_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_2_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_weight_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gr20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'%contacts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_area_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'circular'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center_cell_population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_centre</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_outline</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_bbox_to_anchor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.35,</span> <span class="pre">0.5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_showtitle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.png'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#create_network_graphs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.create_network_graphs" title="Link to this definition"></a></dt>
<dd><p>Generates a adjaceny network graph that summarises how populations interact using the SpOOx output</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> – A pandas DataFrame containing the cell interaction data output from the SpOOx pipeline.</p></li>
<li><p><strong>output_folder</strong> – The folder where the graphs will be saved. Defaults to ‘spoox_figures’.</p></li>
<li><p><strong>fig_size</strong> – The size of the figure. Defaults to (10,10).</p></li>
<li><p><strong>edge_color_map</strong> – The colormap to use for the edge colors. Defaults to ‘Reds’.</p></li>
<li><p><strong>edge_color_map_min/max</strong> – If defined, will set the min and max values on the edge colour.</p></li>
<li><p><strong>node_color_map</strong> – A dictionary mapping cell types to colors. If None, a default color map is generated. Defaults to None.</p></li>
<li><p><strong>filters</strong> – A dictionary with the filters to apply on the SpOOx output. By default it is the following:
gr20 &gt; 1
gr20 lower bound of 95% CI &gt; 1 (ie, statistically significant)
Morueta-Holme false discovery rate &lt; 0.05 (ie, statistically significant)
Morueta-Holme standard effect size from 0 to 100 (ie, positive associations only)</p></li>
<li><p><strong>cell_type_1_list</strong> (<em>list</em><em>, </em><em>strs</em>) – Populations to filter to in cell type 1.</p></li>
<li><p><strong>cell_type_2_list</strong> (<em>list</em><em>, </em><em>strs</em>) – Populations to filter to in cell type 2.</p></li>
<li><p><strong>edge_weight_column</strong> – The column from the dataframe that defines the edge weights. Defaults to ‘gr20’.</p></li>
<li><p><strong>edge_color_column</strong> – The column from the dataframe that defines the edge colors. Defaults to ‘%contacts’.</p></li>
<li><p><strong>edge_scale</strong> – A value to scale the edge weights by. Defaults to 1.</p></li>
<li><p><strong>node_scale</strong> – A value to scale the node sizes by. Defaults to 1.</p></li>
<li><p><strong>node_area_scale</strong> – Scale node sizes so that the area correlates with population abundance, rather than node diameter. Default to True.</p></li>
<li><p><strong>layout_scale</strong> – A scale factor for the node layout. A larger value spreads the nodes further apart. Defaults to 1.</p></li>
<li><p><strong>layout_type</strong> – The type of layout to use for the graph. Options are ‘circular’, ‘kamada_kawai’, or ‘spring’. Defaults to ‘circular’.</p></li>
<li><p><strong>center_cell_population</strong> – If specified, only nodes immediately connected to this node are visualised. Defaults to None.</p></li>
<li><p><strong>force_centre</strong> – If True, force the center_cell_population node to be at the centre of the graph.</p></li>
<li><p><strong>draw_labels</strong> – Whether to draw labels on the nodes. Defaults to True.</p></li>
<li><p><strong>node_outline</strong> – Whether to draw a black outline around the nodes. Defaults to True.</p></li>
<li><p><strong>add_legend</strong> – Whether to add a legend to the graph. Defaults to True.</p></li>
<li><p><strong>legend_bbox_to_anchor</strong> – The location to place the legend in bbox_to_anchor format. Defaults to (1.35, 0.5).</p></li>
<li><p><strong>figure_box</strong> – Whether to show a bounding box around the figure. Defaults to True.</p></li>
<li><p><strong>figure_padding</strong> – The padding around a figure, adjust if nodes are overlapping the edge. Defaults to 0.1.</p></li>
<li><p><strong>figure_showtitle</strong> – Whether to show a title over each figure. Defaults to True.</p></li>
<li><p><strong>save_extension</strong> – Extension for saving file. Defaults to .png</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Output:</dt><dd><p>Saved one graph per state/tissue type detailed in the ‘state’ colun</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.create_population_anndata">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">create_population_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crosstab_normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#create_population_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.create_population_anndata" title="Link to this definition"></a></dt>
<dd><p>Create an AnnData object summarizing populations based on specified observations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>ad.AnnData</em>) – AnnData object.</p></li>
<li><p><strong>pop_obs</strong> (<em>str</em>) – Obs that identifies the population.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em>) – If supplied, Obs which identifies the level at which to summarize the populations, i.e., the subregion column.
If None, populations will be summarized at the ROI level.</p></li>
<li><p><strong>roi_obs</strong> (<em>str</em><em>, </em><em>optional</em>) – Obs that identifies ROIs. Default is ‘ROI’.</p></li>
<li><p><strong>crosstab_normalize</strong> (<em>Union</em><em>[</em><em>bool</em><em>, </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Whether to normalize counts. ‘index’ will normalize within each ROI/subregion,
whereas ‘columns’ will normalize within each population. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An AnnData object summarizing the populations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.create_spoox_heatmaps">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">create_spoox_heatmaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_input</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_mh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spoox_figures'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.png'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_1_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_2_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_signficance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specify_row_order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specify_col_order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_ticks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#create_spoox_heatmaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.create_spoox_heatmaps" title="Link to this definition"></a></dt>
<dd><p>Creates heatmaps from the SpOOx sumary data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_input</strong> (<em>pd.DataFrame</em>) – Pandas dataframe of loaded SpOOx summary data.</p></li>
<li><p><strong>percentile</strong> (<em>float</em>) – Percentile to determine vmax for heatmap color scaling. Default is 95.</p></li>
<li><p><strong>sig_threshold</strong> (<em>float</em>) – Threshold for significance. Default is 0.05.</p></li>
<li><p><strong>cluster_mh</strong> (<em>bool</em>) – If True, the Morueta-Holme column is clustered. Default is True.</p></li>
<li><p><strong>save_folder</strong> (<em>str</em>) – Folder to save the generated heatmaps. Default is ‘spoox_figures’.</p></li>
<li><p><strong>save_extension</strong> (<em>str</em>) – File extension for the saved heatmaps. Default is ‘.png’.</p></li>
<li><p><strong>figsize</strong> (<em>int</em>) – Size of the figure for the heatmaps. Default is 10.</p></li>
<li><p><strong>cell_type_1_list</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Populations to filter to in cell type 1 (rows).</p></li>
<li><p><strong>cell_type_2_list</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Populations to filter to in cell type 1 (columns).</p></li>
<li><p><strong>annotate_signficance</strong> (<em>bool</em>) – Whether to annotate significant values or not. Default is True.</p></li>
<li><p><strong>specify_row_order</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Specify a row order. Default is None.</p></li>
<li><p><strong>specify_col_order</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Specify a column order. Default is None.</p></li>
<li><p><strong>cmap_ticks</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>float</em><em>]</em><em>]</em>) – List of where ticks should appear on the color map. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function saves heatmap figures in the specified folder.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.create_subregions">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">create_subregions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Y_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_noise_as_nan</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_col_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'subregion_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique_label_col_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'subregion'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#create_subregions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.create_subregions" title="Link to this definition"></a></dt>
<dd><p>Apply DBSCAN clustering to spatial data within specified ROI and group columns,
adding unique labels for spatially-separated clusters of labels, which we call subregions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>Union</em><em>[</em><em>pd.DataFrame</em><em>, </em><em>AnnData</em><em>]</em>) – The input DataFrame or an AnnData object (AnnData.obs is used as the DataFrame).</p></li>
<li><p><strong>group_col</strong> (<em>str</em>) – Column name that defines cells belonging to different groups (e.g., the manual labels).</p></li>
<li><p><strong>x_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name for the x spatial coordinates. Default is ‘X_loc’.</p></li>
<li><p><strong>y_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name for the y spatial coordinates. Default is ‘Y_loc’.</p></li>
<li><p><strong>roi_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name for the region of interest (ROI). Default is ‘ROI’.</p></li>
<li><p><strong>max_distance</strong> (<em>float</em><em>, </em><em>optional</em>) – The maximum distance between two cells for them to be considered as in the same subregion. Default is 100.</p></li>
<li><p><strong>min_cells</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of cells in a neighborhood to be established as a subregion. Default is 10.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, modifies the DataFrame in place; otherwise, returns a modified copy. Default is True.</p></li>
<li><p><strong>return_noise_as_nan</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, cells identified as noise by DBSCAN are returned as NaN in the new columns. Default is True.</p></li>
<li><p><strong>cluster_col_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Custom name for the DBSCAN cluster column. Default is ‘subregion_cluster’.</p></li>
<li><p><strong>unique_label_col_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Custom name for the unique label column. Default is ‘subregion’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace=False, returns a new DataFrame with the clustering results. Otherwise, modifies the input DataFrame in place.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[pd.DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.lisa_clustering_image">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">lisa_clustering_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#lisa_clustering_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.lisa_clustering_image" title="Link to this definition"></a></dt>
<dd><p>Perform LISA clustering on two images where each pixel represents a different feature,
and returns a binary clustering result in the shape of the original images. Optionally,
images can be downscaled to reduce computation and memory requirements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>image1</strong> (<em>np.ndarray</em>) – 2D numpy array representing the first feature image.</p></li>
<li><p><strong>image2</strong> (<em>np.ndarray</em>) – 2D numpy array representing the second feature image.</p></li>
<li><p><strong>scale_factor</strong> (<em>float</em><em>, </em><em>optional</em>) – Scaling factor for downsizing the images. Default is 1.0 (no scaling).
Values less than 1.0 will downscale the images.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>2D numpy array of clustering results where:</p>
<ul class="simple">
<li><dl class="simple">
<dt>1 represents ‘High’ clustering status, indicating both conditions are met:</dt><dd><ul>
<li><p>the standardized value of the feature at the location is positive</p></li>
<li><p>the weighted sum of the neighboring values for the second feature is positive</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>0 represents ‘Low’ clustering status, indicating either or both conditions are not met.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input images do not have the same dimensions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.measure_label_objects">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">measure_label_objects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labeled_image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'count'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#measure_label_objects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.measure_label_objects" title="Link to this definition"></a></dt>
<dd><p>Measure specified properties of labeled objects in a labeled image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labeled_image</strong> (<em>np.ndarray</em>) – A 2D array where each unique integer represents a different label.</p></li>
<li><p><strong>metric</strong> (<em>str</em><em>, </em><em>optional</em>) – The property to measure for each labeled object. If ‘count’, counts the number of objects for each label.
Otherwise, any property available in skimage.measure.regionprops can be specified.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame with labels as the index and the specified metric as the column.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.plot_subregion_interactions">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">plot_subregion_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interactions_results</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_counts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(10,</span> <span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path_counts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'subregions_counts.svg'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path_zscore</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'subregions_zscore.svg'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">heatmap_kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#plot_subregion_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.plot_subregion_interactions" title="Link to this definition"></a></dt>
<dd><p>Plot neighborhood enrichment results for subregions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interactions_results</strong> (<em>dict</em>) – Dictionary containing neighborhood enrichment results.</p></li>
<li><p><strong>log_counts</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to log-transform the count data. Default is True.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Default is (10, 10).</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Colormap for heatmaps. Default is ‘viridis’.</p></li>
<li><p><strong>save_path_counts</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save the count heatmap. Default is ‘subregions_counts.svg’.</p></li>
<li><p><strong>save_path_zscore</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save the z-score heatmap. Default is ‘subregions_zscore.svg’.</p></li>
<li><p><strong>**heatmap_kwargs</strong> – Additional keyword arguments for heatmap plotting.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.run_spoox">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">run_spoox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specify_functions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spoox_output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spooxout'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spoox_output_summary_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spooxout_summary'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'stats.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Master_Index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ROI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xloc_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">yloc_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Y_loc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_source_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks_destination_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spoox_mask'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_analysis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">analyse_samples_together</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#run_spoox"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.run_spoox" title="Link to this definition"></a></dt>
<dd><p>Run the SpOOx analysis pipeline from an AnnData object.
More information can be found here: <a class="reference external" href="https://github.com/Taylor-CCB-Group/SpOOx/tree/main/src/spatialstats">https://github.com/Taylor-CCB-Group/SpOOx/tree/main/src/spatialstats</a></p>
<p>Any errors in running functions will be saved in errors.csv</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>Union</em><em>[</em><em>ad.AnnData</em><em>, </em><em>str</em><em>]</em>) – AnnData object, or path (string) to a saved h5ad file.</p></li>
<li><p><strong>population_obs</strong> (<em>str</em>) – The .obs that defines the population for each cell.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – If specifed, should be a .obs that identifies different groups in the data.
In the summary step, it will then compare.</p></li>
<li><p><strong>samples</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Specify a list of samples, if None will process all samples.</p></li>
<li><p><strong>specify_functions</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – By default will run the follow functions from the Spoox pipeline: paircorrelationfunction morueta-holme networkstatistics.
This is a complete list that will be run if ‘all’ is used:
- paircorrelationfunction
- localclusteringheatmaps
- celllocationmap
- contourplots
- quadratcounts
- quadratcelldistributions
- morueta-holme
- networkstatistics.</p></li>
<li><p><strong>run_analysis</strong> (<em>bool</em>) – Whether or not to run the analyis, or just create the spatialstats file.</p></li>
<li><p><strong>analyse_samples_together</strong> (<em>bool</em>) – Whether to analyse all samples together.</p></li>
<li><p><strong>summary</strong> (<em>bool</em>) – Whether to run summary script.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Creates two folders with outputs of SpOOx pipeline.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.spatial.squidpy_subregion_interactions">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.spatial.</span></span><span class="sig-name descname"><span class="pre">squidpy_subregion_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subregion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(0,</span> <span class="pre">20)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_permutations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/spatial.html#squidpy_subregion_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.spatial.squidpy_subregion_interactions" title="Link to this definition"></a></dt>
<dd><p>Calculate neighborhood enrichment for subregions using Squidpy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>population_obs</strong> (<em>str</em>) – Observation key for population annotation.</p></li>
<li><p><strong>subregion</strong> (<em>str</em>) – Key for the subregion annotation.</p></li>
<li><p><strong>radius</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radius for spatial neighbors. Default is (0, 20).</p></li>
<li><p><strong>n_permutations</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of permutations for neighborhood enrichment. Default is 1000.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing neighborhood enrichment results for each subregion.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.utils">
<span id="spatialbiologytoolkit-utils-module"></span><h2>SpatialBiologyToolkit.utils module<a class="headerlink" href="#module-SpatialBiologyToolkit.utils" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.adlog">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">adlog</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ModuleType</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'adata_temp.h5ad'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#adlog"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.adlog" title="Link to this definition"></a></dt>
<dd><p>This function saves a log in the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>entry</strong> (<em>str</em><em>, </em><em>optional</em>) – Log entry description.</p></li>
<li><p><strong>module</strong> (<em>Union</em><em>[</em><em>ModuleType</em><em>, </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Module to log.</p></li>
<li><p><strong>module_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the module.</p></li>
<li><p><strong>module_version</strong> (<em>str</em><em>, </em><em>optional</em>) – Version of the module.</p></li>
<li><p><strong>save</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, saves a temporary file (default is False).</p></li>
<li><p><strong>temp_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path for the temporary file (default is ‘adata_temp.h5ad’).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.analyze_masks">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">analyze_masks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.tiff'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#analyze_masks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.analyze_masks" title="Link to this definition"></a></dt>
<dd><p>Analyze mask files in a given directory, returning a DataFrame with information about each mask.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory containing the image files.</p></li>
<li><p><strong>file_extension</strong> (<em>str</em>) – File extension of the image files (default is ‘.tiff’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with columns ‘ROI’, ‘x_size’, ‘y_size’, ‘Cell_count’, and ‘Mask_path’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.check_file_names_and_counts">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">check_file_names_and_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_folders</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#check_file_names_and_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.check_file_names_and_counts" title="Link to this definition"></a></dt>
<dd><p>Check that the file names (or folder names) in a given directory match the names in a column of a DataFrame,
and that all the files (or folders) have the same number of items in each.
Also provides lists of names that are present in the folder but not in the DataFrame, and vice versa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory containing the files or folders.</p></li>
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the file or folder names in one of its columns.</p></li>
<li><p><strong>column_name</strong> (<em>str</em>) – The name of the column in the DataFrame containing the file or folder names.</p></li>
<li><p><strong>check_folders</strong> (<em>bool</em>) – If True, check folders instead of files.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>True if the conditions are met, False otherwise.</p></li>
<li><p>Set of names present in the folder but not in the DataFrame.</p></li>
<li><p>Set of names present in the DataFrame but not in the folder.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool, set, set</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.compare_lists">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">compare_lists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">L1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">L2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">L1_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">L2_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_error</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#compare_lists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.compare_lists" title="Link to this definition"></a></dt>
<dd><p>Function for thoroughly comparing lists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>L1</strong> (<em>list</em>) – First list to compare.</p></li>
<li><p><strong>L2</strong> (<em>list</em>) – Second list to compare.</p></li>
<li><p><strong>L1_name</strong> (<em>str</em>) – Name of the first list.</p></li>
<li><p><strong>L2_name</strong> (<em>str</em>) – Name of the second list.</p></li>
<li><p><strong>return_error</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, raise a TypeError if the lists do not match (default is True).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.create_celltable">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">create_celltable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Celltables'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'celltable.csv'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#create_celltable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.create_celltable" title="Link to this definition"></a></dt>
<dd><p>Concatenate all CSV files in the specified directory into a single DataFrame,
make a copy of the index with the heading ‘Master_Index’, rename ‘Label’ to ‘ObjectNumber’,
and save the final DataFrame as ‘celltable.csv’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_directory</strong> (<em>str</em>) – Directory containing the CSV files.</p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – Path to save the final concatenated DataFrame.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The concatenated DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.extract_pixel_environment">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">extract_pixel_environment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pixel_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pixel_masks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'celltables_pixel'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#extract_pixel_environment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.extract_pixel_environment" title="Link to this definition"></a></dt>
<dd><p>Extract the most common pixel environment that a cell is in within a labelled image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame with columns ‘ROI’, ‘Mask path’ containing mask file paths.</p></li>
<li><p><strong>pixel_folder</strong> (<em>str</em>) – Directory containing pixel mask images.</p></li>
<li><p><strong>save_directory</strong> (<em>str</em>) – Directory to save the resulting CSV files.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.extract_single_cell">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">extract_single_cell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images_folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#extract_single_cell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.extract_single_cell" title="Link to this definition"></a></dt>
<dd><p>Analyze images to extract single cell data and save the results in separate CSV files for each ROI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame with columns ‘ROI’, ‘Mask path’ containing mask file paths.</p></li>
<li><p><strong>images_folder</strong> (<em>str</em>) – Directory containing folders of images named after each ROI.</p></li>
<li><p><strong>save_directory</strong> (<em>str</em>) – Directory to save the resulting CSV files.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.get_module_path">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">get_module_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">module_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#get_module_path"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.get_module_path" title="Link to this definition"></a></dt>
<dd><p>Get the file path of a Python module or package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>module_name</strong> (<em>str</em>) – The name of the module or package.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the module file or package directory.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.pip_freeze_to_dataframe">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">pip_freeze_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#pip_freeze_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.pip_freeze_to_dataframe" title="Link to this definition"></a></dt>
<dd><p>Capture the output of <cite>pip freeze</cite> and convert it to a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing package names and versions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.print_full">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">print_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#print_full"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.print_full" title="Link to this definition"></a></dt>
<dd><p>Prints a full pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<em>pd.DataFrame</em>) – DataFrame to be printed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.remove_leading_string">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">remove_leading_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ending_string</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.tif'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#remove_leading_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.remove_leading_string" title="Link to this definition"></a></dt>
<dd><p>Rename files in a given directory if they end with a specified string.
The specified string is removed from the file title.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory to search for .tif files.</p></li>
<li><p><strong>ending_string</strong> (<em>str</em>) – The string to check for at the end of the file names.</p></li>
<li><p><strong>extension</strong> (<em>str</em><em>, </em><em>optional</em>) – File extension (default is “.tif”).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.remove_tiff_extensions">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">remove_tiff_extensions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">series</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Series</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#remove_tiff_extensions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.remove_tiff_extensions" title="Link to this definition"></a></dt>
<dd><p>Remove .tiff or .tif extensions from the file names in a pandas Series.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>series</strong> (<em>pd.Series</em>) – The Series containing the file names.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with the file extensions removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.rename_folders">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">rename_folders</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rename_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#rename_folders"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.rename_folders" title="Link to this definition"></a></dt>
<dd><p>Rename folders in a given directory based on a provided dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory containing folders to be renamed.</p></li>
<li><p><strong>rename_dict</strong> (<em>dict</em>) – Dictionary with keys as original folder names and values as new folder names.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.rename_tiff_files">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">rename_tiff_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#rename_tiff_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.rename_tiff_files" title="Link to this definition"></a></dt>
<dd><p>Rename .tiff files in a given directory and all its subdirectories.
Each .tiff file is renamed by taking the last item from its name split by ‘_’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory to search for .tiff files.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.rename_tiff_files_according_to_dict">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">rename_tiff_files_according_to_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rename_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#rename_tiff_files_according_to_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.rename_tiff_files_according_to_dict" title="Link to this definition"></a></dt>
<dd><p>Rename .tiff files in a given directory and all its subdirectories according to a provided dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory to search for .tiff files.</p></li>
<li><p><strong>rename_dict</strong> (<em>dict</em>) – Dictionary with keys as original file names (without extension) and values as new file names.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.save_unique_tiff_names_to_csv">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">save_unique_tiff_names_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csv_file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#save_unique_tiff_names_to_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.save_unique_tiff_names_to_csv" title="Link to this definition"></a></dt>
<dd><p>Save all the unique names of .tiff files from a directory and its subdirectories to a CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory_path</strong> (<em>str</em>) – Path to the directory to search for .tiff files.</p></li>
<li><p><strong>csv_file_path</strong> (<em>str</em>) – Path to save the CSV file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.utils.subset">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.utils.</span></span><span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/utils.html#subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.utils.subset" title="Link to this definition"></a></dt>
<dd><p>Returns a filtered AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>obs_column</strong> (<em>str</em>) – Column in .obs to filter on.</p></li>
<li><p><strong>obs_values</strong> (<em>list</em>) – Values to include if include_list=True, or exclude if include_list=False.</p></li>
<li><p><strong>include_list</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, includes the values in the filter, else excludes them (default is True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Filtered AnnData object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit.visium">
<span id="spatialbiologytoolkit-visium-module"></span><h2>SpatialBiologyToolkit.visium module<a class="headerlink" href="#module-SpatialBiologyToolkit.visium" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="SpatialBiologyToolkit.visium.create_visium_mask">
<span class="sig-prename descclassname"><span class="pre">SpatialBiologyToolkit.visium.</span></span><span class="sig-name descname"><span class="pre">create_visium_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">folder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_mapping</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/SpatialBiologyToolkit/visium.html#create_visium_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatialBiologyToolkit.visium.create_visium_mask" title="Link to this definition"></a></dt>
<dd><p>Create masks for Visium spots based upon a ‘spatial’ directory output from Spaceranger.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>folder</strong> (<em>str</em>) – The ‘spatial’ folder which is the output of Spaceranger.</p></li>
<li><p><strong>show_mapping</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, displays the mask over the tissue image (default is True).</p></li>
<li><p><strong>Output</strong></p></li>
<li><p><strong>------</strong></p></li>
<li><p><strong>folder.</strong> (<em>Saves a 16-bit mask file in the specified</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-SpatialBiologyToolkit">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-SpatialBiologyToolkit" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="introduction.html" class="btn btn-neutral float-left" title="Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Michael Haley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>